<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>GenomicTuples: Classes and Methods • GenomicTuples</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">GenomicTuples</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/GenomicTuplesIntroduction.html">GenomicTuples: Classes and Methods</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>GenomicTuples: Classes and Methods</h1>
                        <h4 class="author">Peter Hickey</h4>
            
            <h4 class="date">Modified: 12 February 2015. Compiled: 30 Mar 2018</h4>
          </div>

    
    
<div class="contents">
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>The <em><a href="http://bioconductor.org/packages/GenomicTuples">GenomicTuples</a></em> R package defines general purpose containers for storing <em>genomic tuples</em>. It aims to provide functionality for tuples of genomic co-ordinates that are analogous to those available for genomic ranges in the <em><a href="http://bioconductor.org/packages/GenomicRanges">GenomicRanges</a></em> Bioconductor package.</p>
<p>As you will see, the functionality of the <em><a href="http://bioconductor.org/packages/GenomicTuples">GenomicTuples</a></em> package is based almost entirely on the wonderful <em><a href="http://bioconductor.org/packages/GenomicRanges">GenomicRanges</a></em> package. Therefore, I have tried to keep the user interface as similar as possible. This vignette is also heavily based on the vignette “An Introduction to Genomic Ranges Classes”, which is included with the <em><a href="http://bioconductor.org/packages/GenomicRanges">GenomicRanges</a></em> package<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>. While not essential, familiarity with the <em><a href="http://bioconductor.org/packages/GenomicRanges">GenomicRanges</a></em> will be of benefit in understanding the <em><a href="http://bioconductor.org/packages/GenomicTuples">GenomicTuples</a></em> package.</p>
<div id="what-is-a-genomic-tuple" class="section level2">
<h2 class="hasAnchor">
<a href="#what-is-a-genomic-tuple" class="anchor"></a>What is a genomic tuple?</h2>
<p>A genomic tuple is defined by a <em>sequence name</em> (<code>seqnames</code>), a <em>strand</em> (<code>strand</code>) and a <em>tuple</em> (<code>tuples</code>). All positions in a genomic tuple must be on the same strand and sorted in ascending order. Each tuple has an associated <code>size</code>, which is a positive integer. For example, <code>chr1:+:{34, 39, 60}</code> is a 3-tuple (<code>size</code> = 3) of the positions <code>chr1:34</code>, <code>chr1:39</code> and <code>chr1:60</code> on the <code>+</code> strand.</p>
<p>When referring to genomic tuples of a general (fixed) <code>size</code>, I will abbreviate these to <span class="math inline">\(m\)</span>-tuples, where <span class="math inline">\(m\)</span> = <code>size</code>. I will refer to the first position as <span class="math inline">\(pos_{1}\)</span> (<code>pos1</code>), the second as <span class="math inline">\(pos_{2}\)</span> (<code>pos2</code>), <span class="math inline">\(\ldots{}\)</span>, and the final position as <span class="math inline">\(pos_{m}\)</span> (<code>posm</code>).</p>
<p>The difference between a genomic tuple and a genomic range can be thought of as the difference between a set and an interval. For example, the genomic tuple <code>chr10:-:{800, 900}</code> only includes the positions <code>chr10:-:800</code> and <code>chr10:-:900</code> whereas the genomic range <code>chr10:-:[800, 900]</code> includes the positions <code>chr10:-:800</code>, <code>chr10:-:801</code>, <code>chr10:-:802</code>, <span class="math inline">\(\ldots{}\)</span>, <code>chr10:-:900</code>.</p>
</div>
<div id="when-might-you-need-a-genomic-tuple" class="section level2">
<h2 class="hasAnchor">
<a href="#when-might-you-need-a-genomic-tuple" class="anchor"></a>When might you need a genomic tuple?</h2>
<p>In short, whenever the co-ordinates of your genomic data are better defined by a set than by an interval.</p>
<p>The original use case for the <em>GTuples</em> class was to store the genomic co-ordinates of “methylation patterns”. I am currently developing these ideas in a separate R package, <em><a href="https://github.com/PeteHaitch/MethylationTuples">MethylationTuples</a></em>, which makes heavy use of the <em>GTuples</em> class. Other genomic data, such as long reads containing multiple variants, may also be better conceptualised as genomic tuples rather than as genomic ranges and therefore may benefit from the <em><a href="http://bioconductor.org/packages/GenomicTuples">GenomicTuples</a></em> infrastructure.</p>
</div>
</div>
<div id="gtuples" class="section level1">
<h1 class="hasAnchor">
<a href="#gtuples" class="anchor"></a><em>GTuples</em>
</h1>
<p>The <em>GTuples</em> class represents a collection of genomic tuples, where each tuple has the same <code>size</code>. These objects can be created by using the <code>GTuples</code> constructor function. For example, the following code creates a <em>GTuples</em> object with 10 genomic tuples:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">library</span>(GenomicTuples)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="co">#&gt; Loading required package: GenomicRanges</span></a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="co">#&gt; Loading required package: stats4</span></a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="co">#&gt; Loading required package: BiocGenerics</span></a>
<a class="sourceLine" id="cb1-5" data-line-number="5"><span class="co">#&gt; Loading required package: parallel</span></a>
<a class="sourceLine" id="cb1-6" data-line-number="6"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-7" data-line-number="7"><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></a>
<a class="sourceLine" id="cb1-8" data-line-number="8"><span class="co">#&gt; The following objects are masked from 'package:parallel':</span></a>
<a class="sourceLine" id="cb1-9" data-line-number="9"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-10" data-line-number="10"><span class="co">#&gt;     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,</span></a>
<a class="sourceLine" id="cb1-11" data-line-number="11"><span class="co">#&gt;     clusterExport, clusterMap, parApply, parCapply, parLapply,</span></a>
<a class="sourceLine" id="cb1-12" data-line-number="12"><span class="co">#&gt;     parLapplyLB, parRapply, parSapply, parSapplyLB</span></a>
<a class="sourceLine" id="cb1-13" data-line-number="13"><span class="co">#&gt; The following objects are masked from 'package:stats':</span></a>
<a class="sourceLine" id="cb1-14" data-line-number="14"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-15" data-line-number="15"><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></a>
<a class="sourceLine" id="cb1-16" data-line-number="16"><span class="co">#&gt; The following objects are masked from 'package:base':</span></a>
<a class="sourceLine" id="cb1-17" data-line-number="17"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-18" data-line-number="18"><span class="co">#&gt;     anyDuplicated, append, as.data.frame, basename, cbind,</span></a>
<a class="sourceLine" id="cb1-19" data-line-number="19"><span class="co">#&gt;     colMeans, colnames, colSums, dirname, do.call, duplicated,</span></a>
<a class="sourceLine" id="cb1-20" data-line-number="20"><span class="co">#&gt;     eval, evalq, Filter, Find, get, grep, grepl, intersect,</span></a>
<a class="sourceLine" id="cb1-21" data-line-number="21"><span class="co">#&gt;     is.unsorted, lapply, lengths, Map, mapply, match, mget, order,</span></a>
<a class="sourceLine" id="cb1-22" data-line-number="22"><span class="co">#&gt;     paste, pmax, pmax.int, pmin, pmin.int, Position, rank, rbind,</span></a>
<a class="sourceLine" id="cb1-23" data-line-number="23"><span class="co">#&gt;     Reduce, rowMeans, rownames, rowSums, sapply, setdiff, sort,</span></a>
<a class="sourceLine" id="cb1-24" data-line-number="24"><span class="co">#&gt;     table, tapply, union, unique, unsplit, which, which.max,</span></a>
<a class="sourceLine" id="cb1-25" data-line-number="25"><span class="co">#&gt;     which.min</span></a>
<a class="sourceLine" id="cb1-26" data-line-number="26"><span class="co">#&gt; Loading required package: S4Vectors</span></a>
<a class="sourceLine" id="cb1-27" data-line-number="27"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-28" data-line-number="28"><span class="co">#&gt; Attaching package: 'S4Vectors'</span></a>
<a class="sourceLine" id="cb1-29" data-line-number="29"><span class="co">#&gt; The following object is masked from 'package:base':</span></a>
<a class="sourceLine" id="cb1-30" data-line-number="30"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-31" data-line-number="31"><span class="co">#&gt;     expand.grid</span></a>
<a class="sourceLine" id="cb1-32" data-line-number="32"><span class="co">#&gt; Loading required package: IRanges</span></a>
<a class="sourceLine" id="cb1-33" data-line-number="33"><span class="co">#&gt; Loading required package: GenomeInfoDb</span></a></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1">seqinfo &lt;-<span class="st"> </span><span class="kw">Seqinfo</span>(<span class="kw">paste0</span>(<span class="st">"chr"</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>), <span class="kw">c</span>(<span class="dv">1000</span>, <span class="dv">2000</span>, <span class="dv">1500</span>), <span class="ot">NA</span>, <span class="st">"mock1"</span>)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2">gt3 &lt;-<span class="st"> </span><span class="kw">GTuples</span>(<span class="dt">seqnames =</span> <span class="kw">Rle</span>(<span class="kw">c</span>(<span class="st">"chr1"</span>, <span class="st">"chr2"</span>, <span class="st">"chr1"</span>, <span class="st">"chr3"</span>),</a>
<a class="sourceLine" id="cb2-3" data-line-number="3">                              <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">4</span>)),</a>
<a class="sourceLine" id="cb2-4" data-line-number="4">               <span class="dt">tuples =</span> <span class="kw">matrix</span>(<span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">2</span><span class="op">:</span><span class="dv">11</span>, <span class="dv">3</span><span class="op">:</span><span class="dv">12</span>), <span class="dt">ncol =</span> <span class="dv">3</span>),</a>
<a class="sourceLine" id="cb2-5" data-line-number="5">               <span class="dt">strand =</span> <span class="kw">Rle</span>(<span class="kw">strand</span>(<span class="kw">c</span>(<span class="st">"-"</span>, <span class="st">"+"</span>, <span class="st">"*"</span>, <span class="st">"+"</span>, <span class="st">"-"</span>)),</a>
<a class="sourceLine" id="cb2-6" data-line-number="6">                            <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">2</span>)),</a>
<a class="sourceLine" id="cb2-7" data-line-number="7">               <span class="dt">score =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">GC =</span> <span class="kw">seq</span>(<span class="dv">1</span>, <span class="dv">0</span>, <span class="dt">length =</span> <span class="dv">10</span>), <span class="dt">seqinfo =</span> seqinfo)</a>
<a class="sourceLine" id="cb2-8" data-line-number="8"><span class="kw">names</span>(gt3) &lt;-<span class="st"> </span>letters[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</a>
<a class="sourceLine" id="cb2-9" data-line-number="9">gt3</a>
<a class="sourceLine" id="cb2-10" data-line-number="10"><span class="co">#&gt; GTuples object with 10 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb2-11" data-line-number="11"><span class="co">#&gt;     seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb2-12" data-line-number="12"><span class="co">#&gt;   a     chr1    1    2    3      - |     1                 1</span></a>
<a class="sourceLine" id="cb2-13" data-line-number="13"><span class="co">#&gt;   b     chr2    2    3    4      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb2-14" data-line-number="14"><span class="co">#&gt;   c     chr2    3    4    5      + |     3 0.777777777777778</span></a>
<a class="sourceLine" id="cb2-15" data-line-number="15"><span class="co">#&gt;   d     chr2    4    5    6      * |     4 0.666666666666667</span></a>
<a class="sourceLine" id="cb2-16" data-line-number="16"><span class="co">#&gt;   e     chr1    5    6    7      * |     5 0.555555555555556</span></a>
<a class="sourceLine" id="cb2-17" data-line-number="17"><span class="co">#&gt;   f     chr1    6    7    8      + |     6 0.444444444444444</span></a>
<a class="sourceLine" id="cb2-18" data-line-number="18"><span class="co">#&gt;   g     chr3    7    8    9      + |     7 0.333333333333333</span></a>
<a class="sourceLine" id="cb2-19" data-line-number="19"><span class="co">#&gt;   h     chr3    8    9   10      + |     8 0.222222222222222</span></a>
<a class="sourceLine" id="cb2-20" data-line-number="20"><span class="co">#&gt;   i     chr3    9   10   11      - |     9 0.111111111111111</span></a>
<a class="sourceLine" id="cb2-21" data-line-number="21"><span class="co">#&gt;   j     chr3   10   11   12      - |    10                 0</span></a>
<a class="sourceLine" id="cb2-22" data-line-number="22"><span class="co">#&gt;   ---</span></a>
<a class="sourceLine" id="cb2-23" data-line-number="23"><span class="co">#&gt;   seqinfo: 3 sequences from mock1 genome</span></a></code></pre></div>
<p>The output of the <em>GTuples</em> <code>show</code> method is very similar to that of the <code>show</code> method for <em>GenomicRanges::GRanges</em> objects. Namely, it separates the information into a left and right hand region that are separated by <code>|</code> symbols. The genomic coordinates (<code>seqnames</code>, <code>tuples</code>, and <code>strand</code>) are located on the left-hand side and the metadata columns (annotation) are located on the right. For this example, the metadata is comprised of <code>score</code> and <code>GC</code> information, but almost anything can be stored in the metadata portion of a <em>GTuples</em> object.</p>
<p>The main difference between a <em>GTuples</em> object and <em>GenomicRanges::GRanges</em> object is that the former uses <em>tuples</em> while the latter uses <em>ranges</em> in the genomic coordinates.</p>
<p>For even more information on the <em>GTuples</em> class, be sure to consult the documentation:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">?GTuples</a></code></pre></div>
<div id="gtuples-methods" class="section level2">
<h2 class="hasAnchor">
<a href="#gtuples-methods" class="anchor"></a><em>GTuples</em> methods</h2>
<p>Most methods defined for <em>GenomicRanges::GRanges</em> are also defined for <em>GTuples</em>. Those that are not yet defined, which are those that make sense for ranges but generally not for tuples, return error messages.</p>
<p>If you require a method that is not defined for <em>GTuples</em> but is defined for <em>GenomicRanges::GRanges</em>, then this can be achieved by first coercing the <em>GTuples</em> object to a <em>GenomicRanges::GRanges</em> object; <strong>Warning: coercing a <em>GTuples</em> object to a <em>GenomicRanges::GRanges</em> is generally a destructive operation</strong>.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="kw">as</span>(gt3, <span class="st">"GRanges"</span>)</a>
<a class="sourceLine" id="cb4-2" data-line-number="2"><span class="co">#&gt; GRanges object with 10 ranges and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb4-3" data-line-number="3"><span class="co">#&gt;     seqnames    ranges strand |     score                GC</span></a>
<a class="sourceLine" id="cb4-4" data-line-number="4"><span class="co">#&gt;        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;         &lt;numeric&gt;</span></a>
<a class="sourceLine" id="cb4-5" data-line-number="5"><span class="co">#&gt;   a     chr1       1-3      - |         1                 1</span></a>
<a class="sourceLine" id="cb4-6" data-line-number="6"><span class="co">#&gt;   b     chr2       2-4      + |         2 0.888888888888889</span></a>
<a class="sourceLine" id="cb4-7" data-line-number="7"><span class="co">#&gt;   c     chr2       3-5      + |         3 0.777777777777778</span></a>
<a class="sourceLine" id="cb4-8" data-line-number="8"><span class="co">#&gt;   d     chr2       4-6      * |         4 0.666666666666667</span></a>
<a class="sourceLine" id="cb4-9" data-line-number="9"><span class="co">#&gt;   e     chr1       5-7      * |         5 0.555555555555556</span></a>
<a class="sourceLine" id="cb4-10" data-line-number="10"><span class="co">#&gt;   f     chr1       6-8      + |         6 0.444444444444444</span></a>
<a class="sourceLine" id="cb4-11" data-line-number="11"><span class="co">#&gt;   g     chr3       7-9      + |         7 0.333333333333333</span></a>
<a class="sourceLine" id="cb4-12" data-line-number="12"><span class="co">#&gt;   h     chr3      8-10      + |         8 0.222222222222222</span></a>
<a class="sourceLine" id="cb4-13" data-line-number="13"><span class="co">#&gt;   i     chr3      9-11      - |         9 0.111111111111111</span></a>
<a class="sourceLine" id="cb4-14" data-line-number="14"><span class="co">#&gt;   j     chr3     10-12      - |        10                 0</span></a>
<a class="sourceLine" id="cb4-15" data-line-number="15"><span class="co">#&gt;   -------</span></a>
<a class="sourceLine" id="cb4-16" data-line-number="16"><span class="co">#&gt;   seqinfo: 3 sequences from mock1 genome</span></a></code></pre></div>
<div id="basic-gtuples-accessors" class="section level3">
<h3 class="hasAnchor">
<a href="#basic-gtuples-accessors" class="anchor"></a>Basic <em>GTuples</em> accessors</h3>
<p>The components of the genomic coordinates within a <em>GTuples</em> object can be extracted using the <code>seqnames</code>, <code>tuples</code>, and <code>strand</code> accessor functions. <strong>Warning: The <code>tuples</code> accessor should be used in place of the <code>ranges</code> accessor. While the <code>ranges</code> method is well-defined, namely it accesses <span class="math inline">\(pos_{1}\)</span> and <span class="math inline">\(pos_{m}\)</span> of the object, this is not generally what is desired or required.</strong></p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw">seqnames</span>(gt3)</a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="co">#&gt; factor-Rle of length 10 with 4 runs</span></a>
<a class="sourceLine" id="cb5-3" data-line-number="3"><span class="co">#&gt;   Lengths:    1    3    2    4</span></a>
<a class="sourceLine" id="cb5-4" data-line-number="4"><span class="co">#&gt;   Values : chr1 chr2 chr1 chr3</span></a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="co">#&gt; Levels(3): chr1 chr2 chr3</span></a>
<a class="sourceLine" id="cb5-6" data-line-number="6"><span class="kw"><a href="../reference/GTuples-class.html">tuples</a></span>(gt3)</a>
<a class="sourceLine" id="cb5-7" data-line-number="7"><span class="co">#&gt;       pos1 pos2 pos3</span></a>
<a class="sourceLine" id="cb5-8" data-line-number="8"><span class="co">#&gt;  [1,]    1    2    3</span></a>
<a class="sourceLine" id="cb5-9" data-line-number="9"><span class="co">#&gt;  [2,]    2    3    4</span></a>
<a class="sourceLine" id="cb5-10" data-line-number="10"><span class="co">#&gt;  [3,]    3    4    5</span></a>
<a class="sourceLine" id="cb5-11" data-line-number="11"><span class="co">#&gt;  [4,]    4    5    6</span></a>
<a class="sourceLine" id="cb5-12" data-line-number="12"><span class="co">#&gt;  [5,]    5    6    7</span></a>
<a class="sourceLine" id="cb5-13" data-line-number="13"><span class="co">#&gt;  [6,]    6    7    8</span></a>
<a class="sourceLine" id="cb5-14" data-line-number="14"><span class="co">#&gt;  [7,]    7    8    9</span></a>
<a class="sourceLine" id="cb5-15" data-line-number="15"><span class="co">#&gt;  [8,]    8    9   10</span></a>
<a class="sourceLine" id="cb5-16" data-line-number="16"><span class="co">#&gt;  [9,]    9   10   11</span></a>
<a class="sourceLine" id="cb5-17" data-line-number="17"><span class="co">#&gt; [10,]   10   11   12</span></a>
<a class="sourceLine" id="cb5-18" data-line-number="18"><span class="kw">strand</span>(gt3)</a>
<a class="sourceLine" id="cb5-19" data-line-number="19"><span class="co">#&gt; factor-Rle of length 10 with 5 runs</span></a>
<a class="sourceLine" id="cb5-20" data-line-number="20"><span class="co">#&gt;   Lengths: 1 2 2 3 2</span></a>
<a class="sourceLine" id="cb5-21" data-line-number="21"><span class="co">#&gt;   Values : - + * + -</span></a>
<a class="sourceLine" id="cb5-22" data-line-number="22"><span class="co">#&gt; Levels(3): + - *</span></a></code></pre></div>
<p>Stored annotations for these coordinates can be extracted as a <em>DataFrame</em> object using the <code>mcols</code> accessor:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw">mcols</span>(gt3)</a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="co">#&gt; DataFrame with 10 rows and 2 columns</span></a>
<a class="sourceLine" id="cb6-3" data-line-number="3"><span class="co">#&gt;        score                GC</span></a>
<a class="sourceLine" id="cb6-4" data-line-number="4"><span class="co">#&gt;    &lt;integer&gt;         &lt;numeric&gt;</span></a>
<a class="sourceLine" id="cb6-5" data-line-number="5"><span class="co">#&gt; 1          1                 1</span></a>
<a class="sourceLine" id="cb6-6" data-line-number="6"><span class="co">#&gt; 2          2 0.888888888888889</span></a>
<a class="sourceLine" id="cb6-7" data-line-number="7"><span class="co">#&gt; 3          3 0.777777777777778</span></a>
<a class="sourceLine" id="cb6-8" data-line-number="8"><span class="co">#&gt; 4          4 0.666666666666667</span></a>
<a class="sourceLine" id="cb6-9" data-line-number="9"><span class="co">#&gt; 5          5 0.555555555555556</span></a>
<a class="sourceLine" id="cb6-10" data-line-number="10"><span class="co">#&gt; 6          6 0.444444444444444</span></a>
<a class="sourceLine" id="cb6-11" data-line-number="11"><span class="co">#&gt; 7          7 0.333333333333333</span></a>
<a class="sourceLine" id="cb6-12" data-line-number="12"><span class="co">#&gt; 8          8 0.222222222222222</span></a>
<a class="sourceLine" id="cb6-13" data-line-number="13"><span class="co">#&gt; 9          9 0.111111111111111</span></a>
<a class="sourceLine" id="cb6-14" data-line-number="14"><span class="co">#&gt; 10        10                 0</span></a></code></pre></div>
<p><em>Seqinfo</em> can be extracted using the <code>seqinfo</code> accessor:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="kw">seqinfo</span>(gt3)</a>
<a class="sourceLine" id="cb7-2" data-line-number="2"><span class="co">#&gt; Seqinfo object with 3 sequences from mock1 genome:</span></a>
<a class="sourceLine" id="cb7-3" data-line-number="3"><span class="co">#&gt;   seqnames seqlengths isCircular genome</span></a>
<a class="sourceLine" id="cb7-4" data-line-number="4"><span class="co">#&gt;   chr1           1000         NA  mock1</span></a>
<a class="sourceLine" id="cb7-5" data-line-number="5"><span class="co">#&gt;   chr2           2000         NA  mock1</span></a>
<a class="sourceLine" id="cb7-6" data-line-number="6"><span class="co">#&gt;   chr3           1500         NA  mock1</span></a></code></pre></div>
<p>Methods for accessing the length and names are also defined:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="kw">length</span>(gt3)</a>
<a class="sourceLine" id="cb8-2" data-line-number="2"><span class="co">#&gt; [1] 10</span></a>
<a class="sourceLine" id="cb8-3" data-line-number="3"><span class="kw">names</span>(gt3)</a>
<a class="sourceLine" id="cb8-4" data-line-number="4"><span class="co">#&gt;  [1] "a" "b" "c" "d" "e" "f" "g" "h" "i" "j"</span></a></code></pre></div>
</div>
<div id="splitting-and-combining-gtuples-objects" class="section level3">
<h3 class="hasAnchor">
<a href="#splitting-and-combining-gtuples-objects" class="anchor"></a>Splitting and combining <em>GTuples</em> objects}</h3>
<p><em>GTuples</em> objects can be divided into groups using the <code>split</code> method. This produces a <em>GTuplesList</em> object, a class that will be discussed in detail in the next section:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">sp &lt;-<span class="st"> </span><span class="kw">split</span>(gt3, <span class="kw">rep</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>, <span class="dt">each=</span><span class="dv">5</span>))</a>
<a class="sourceLine" id="cb9-2" data-line-number="2">sp</a>
<a class="sourceLine" id="cb9-3" data-line-number="3"><span class="co">#&gt; GTuplesList object of length 2:</span></a>
<a class="sourceLine" id="cb9-4" data-line-number="4"><span class="co">#&gt; $1 </span></a>
<a class="sourceLine" id="cb9-5" data-line-number="5"><span class="co">#&gt; GTuples object with 5 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb9-6" data-line-number="6"><span class="co">#&gt;     seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb9-7" data-line-number="7"><span class="co">#&gt;   a     chr1    1    2    3      - |     1                 1</span></a>
<a class="sourceLine" id="cb9-8" data-line-number="8"><span class="co">#&gt;   b     chr2    2    3    4      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb9-9" data-line-number="9"><span class="co">#&gt;   c     chr2    3    4    5      + |     3 0.777777777777778</span></a>
<a class="sourceLine" id="cb9-10" data-line-number="10"><span class="co">#&gt;   d     chr2    4    5    6      * |     4 0.666666666666667</span></a>
<a class="sourceLine" id="cb9-11" data-line-number="11"><span class="co">#&gt;   e     chr1    5    6    7      * |     5 0.555555555555556</span></a>
<a class="sourceLine" id="cb9-12" data-line-number="12"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb9-13" data-line-number="13"><span class="co">#&gt; $2 </span></a>
<a class="sourceLine" id="cb9-14" data-line-number="14"><span class="co">#&gt; GTuples object with 5 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb9-15" data-line-number="15"><span class="co">#&gt;     seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb9-16" data-line-number="16"><span class="co">#&gt;   f     chr1    6    7    8      + |     6 0.444444444444444</span></a>
<a class="sourceLine" id="cb9-17" data-line-number="17"><span class="co">#&gt;   g     chr3    7    8    9      + |     7 0.333333333333333</span></a>
<a class="sourceLine" id="cb9-18" data-line-number="18"><span class="co">#&gt;   h     chr3    8    9   10      + |     8 0.222222222222222</span></a>
<a class="sourceLine" id="cb9-19" data-line-number="19"><span class="co">#&gt;   i     chr3    9   10   11      - |     9 0.111111111111111</span></a>
<a class="sourceLine" id="cb9-20" data-line-number="20"><span class="co">#&gt;   j     chr3   10   11   12      - |    10                 0</span></a>
<a class="sourceLine" id="cb9-21" data-line-number="21"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb9-22" data-line-number="22"><span class="co">#&gt; -------</span></a>
<a class="sourceLine" id="cb9-23" data-line-number="23"><span class="co">#&gt; seqinfo: 3 sequences from mock1 genome</span></a></code></pre></div>
<p>If you then grab the components of this <code>GenomicTuplesList</code>, they can also be combined by using the <code>c</code> and <code>append</code> methods:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="kw">c</span>(sp[[<span class="dv">1</span>]], sp[[<span class="dv">2</span>]])</a>
<a class="sourceLine" id="cb10-2" data-line-number="2"><span class="co">#&gt; GTuples object with 10 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb10-3" data-line-number="3"><span class="co">#&gt;     seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb10-4" data-line-number="4"><span class="co">#&gt;   a     chr1    1    2    3      - |     1                 1</span></a>
<a class="sourceLine" id="cb10-5" data-line-number="5"><span class="co">#&gt;   b     chr2    2    3    4      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb10-6" data-line-number="6"><span class="co">#&gt;   c     chr2    3    4    5      + |     3 0.777777777777778</span></a>
<a class="sourceLine" id="cb10-7" data-line-number="7"><span class="co">#&gt;   d     chr2    4    5    6      * |     4 0.666666666666667</span></a>
<a class="sourceLine" id="cb10-8" data-line-number="8"><span class="co">#&gt;   e     chr1    5    6    7      * |     5 0.555555555555556</span></a>
<a class="sourceLine" id="cb10-9" data-line-number="9"><span class="co">#&gt;   f     chr1    6    7    8      + |     6 0.444444444444444</span></a>
<a class="sourceLine" id="cb10-10" data-line-number="10"><span class="co">#&gt;   g     chr3    7    8    9      + |     7 0.333333333333333</span></a>
<a class="sourceLine" id="cb10-11" data-line-number="11"><span class="co">#&gt;   h     chr3    8    9   10      + |     8 0.222222222222222</span></a>
<a class="sourceLine" id="cb10-12" data-line-number="12"><span class="co">#&gt;   i     chr3    9   10   11      - |     9 0.111111111111111</span></a>
<a class="sourceLine" id="cb10-13" data-line-number="13"><span class="co">#&gt;   j     chr3   10   11   12      - |    10                 0</span></a>
<a class="sourceLine" id="cb10-14" data-line-number="14"><span class="co">#&gt;   ---</span></a>
<a class="sourceLine" id="cb10-15" data-line-number="15"><span class="co">#&gt;   seqinfo: 3 sequences from mock1 genome</span></a></code></pre></div>
</div>
<div id="subsetting-gtuples-objects" class="section level3">
<h3 class="hasAnchor">
<a href="#subsetting-gtuples-objects" class="anchor"></a>Subsetting <em>GTuples</em> objects</h3>
<p>The expected subsetting operations are also available for <em>GTuples</em> objects:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1">gt3[<span class="dv">2</span><span class="op">:</span><span class="dv">3</span>]</a>
<a class="sourceLine" id="cb11-2" data-line-number="2"><span class="co">#&gt; GTuples object with 2 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb11-3" data-line-number="3"><span class="co">#&gt;     seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb11-4" data-line-number="4"><span class="co">#&gt;   b     chr2    2    3    4      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb11-5" data-line-number="5"><span class="co">#&gt;   c     chr2    3    4    5      + |     3 0.777777777777778</span></a>
<a class="sourceLine" id="cb11-6" data-line-number="6"><span class="co">#&gt;   ---</span></a>
<a class="sourceLine" id="cb11-7" data-line-number="7"><span class="co">#&gt;   seqinfo: 3 sequences from mock1 genome</span></a></code></pre></div>
<p>A second argument to the <code>[</code> subset operator can be used to specify which metadata columns to extract from the <em>GTuples</em> object. For example:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1">gt3[<span class="dv">2</span><span class="op">:</span><span class="dv">3</span>, <span class="st">"GC"</span>]</a>
<a class="sourceLine" id="cb12-2" data-line-number="2"><span class="co">#&gt; GTuples object with 2 x 3-tuples and 1 metadata column:</span></a>
<a class="sourceLine" id="cb12-3" data-line-number="3"><span class="co">#&gt;     seqnames pos1 pos2 pos3 strand |                GC</span></a>
<a class="sourceLine" id="cb12-4" data-line-number="4"><span class="co">#&gt;   b     chr2    2    3    4      + | 0.888888888888889</span></a>
<a class="sourceLine" id="cb12-5" data-line-number="5"><span class="co">#&gt;   c     chr2    3    4    5      + | 0.777777777777778</span></a>
<a class="sourceLine" id="cb12-6" data-line-number="6"><span class="co">#&gt;   ---</span></a>
<a class="sourceLine" id="cb12-7" data-line-number="7"><span class="co">#&gt;   seqinfo: 3 sequences from mock1 genome</span></a></code></pre></div>
<p>You can also assign into elements of the <em>GTuples</em> object. Here is an example where the 2nd row of a <em>GTuples</em> object is replaced with the 1st row of <code>gt3</code>:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">gt3_mod &lt;-<span class="st"> </span>gt3</a>
<a class="sourceLine" id="cb13-2" data-line-number="2">gt3_mod[<span class="dv">2</span>] &lt;-<span class="st"> </span>gt3[<span class="dv">1</span>]</a>
<a class="sourceLine" id="cb13-3" data-line-number="3"><span class="kw">head</span>(gt3_mod, <span class="dt">n =</span> <span class="dv">3</span>)</a>
<a class="sourceLine" id="cb13-4" data-line-number="4"><span class="co">#&gt; GTuples object with 3 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb13-5" data-line-number="5"><span class="co">#&gt;     seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb13-6" data-line-number="6"><span class="co">#&gt;   a     chr1    1    2    3      - |     1                 1</span></a>
<a class="sourceLine" id="cb13-7" data-line-number="7"><span class="co">#&gt;   b     chr1    1    2    3      - |     1                 1</span></a>
<a class="sourceLine" id="cb13-8" data-line-number="8"><span class="co">#&gt;   c     chr2    3    4    5      + |     3 0.777777777777778</span></a>
<a class="sourceLine" id="cb13-9" data-line-number="9"><span class="co">#&gt;   ---</span></a>
<a class="sourceLine" id="cb13-10" data-line-number="10"><span class="co">#&gt;   seqinfo: 3 sequences from mock1 genome</span></a></code></pre></div>
<p>There are also methods to repeat, reverse, or select specific portions of <em>GTuples</em> objects:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="kw">rep</span>(gt3[<span class="dv">2</span>], <span class="dt">times =</span> <span class="dv">3</span>)</a>
<a class="sourceLine" id="cb14-2" data-line-number="2"><span class="co">#&gt; GTuples object with 3 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb14-3" data-line-number="3"><span class="co">#&gt;     seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb14-4" data-line-number="4"><span class="co">#&gt;   b     chr2    2    3    4      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb14-5" data-line-number="5"><span class="co">#&gt;   b     chr2    2    3    4      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb14-6" data-line-number="6"><span class="co">#&gt;   b     chr2    2    3    4      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb14-7" data-line-number="7"><span class="co">#&gt;   ---</span></a>
<a class="sourceLine" id="cb14-8" data-line-number="8"><span class="co">#&gt;   seqinfo: 3 sequences from mock1 genome</span></a>
<a class="sourceLine" id="cb14-9" data-line-number="9"><span class="kw">rev</span>(gt3)</a>
<a class="sourceLine" id="cb14-10" data-line-number="10"><span class="co">#&gt; GTuples object with 10 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb14-11" data-line-number="11"><span class="co">#&gt;     seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb14-12" data-line-number="12"><span class="co">#&gt;   j     chr3   10   11   12      - |    10                 0</span></a>
<a class="sourceLine" id="cb14-13" data-line-number="13"><span class="co">#&gt;   i     chr3    9   10   11      - |     9 0.111111111111111</span></a>
<a class="sourceLine" id="cb14-14" data-line-number="14"><span class="co">#&gt;   h     chr3    8    9   10      + |     8 0.222222222222222</span></a>
<a class="sourceLine" id="cb14-15" data-line-number="15"><span class="co">#&gt;   g     chr3    7    8    9      + |     7 0.333333333333333</span></a>
<a class="sourceLine" id="cb14-16" data-line-number="16"><span class="co">#&gt;   f     chr1    6    7    8      + |     6 0.444444444444444</span></a>
<a class="sourceLine" id="cb14-17" data-line-number="17"><span class="co">#&gt;   e     chr1    5    6    7      * |     5 0.555555555555556</span></a>
<a class="sourceLine" id="cb14-18" data-line-number="18"><span class="co">#&gt;   d     chr2    4    5    6      * |     4 0.666666666666667</span></a>
<a class="sourceLine" id="cb14-19" data-line-number="19"><span class="co">#&gt;   c     chr2    3    4    5      + |     3 0.777777777777778</span></a>
<a class="sourceLine" id="cb14-20" data-line-number="20"><span class="co">#&gt;   b     chr2    2    3    4      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb14-21" data-line-number="21"><span class="co">#&gt;   a     chr1    1    2    3      - |     1                 1</span></a>
<a class="sourceLine" id="cb14-22" data-line-number="22"><span class="co">#&gt;   ---</span></a>
<a class="sourceLine" id="cb14-23" data-line-number="23"><span class="co">#&gt;   seqinfo: 3 sequences from mock1 genome</span></a>
<a class="sourceLine" id="cb14-24" data-line-number="24"><span class="kw">head</span>(gt3, <span class="dt">n =</span> <span class="dv">2</span>)</a>
<a class="sourceLine" id="cb14-25" data-line-number="25"><span class="co">#&gt; GTuples object with 2 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb14-26" data-line-number="26"><span class="co">#&gt;     seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb14-27" data-line-number="27"><span class="co">#&gt;   a     chr1    1    2    3      - |     1                 1</span></a>
<a class="sourceLine" id="cb14-28" data-line-number="28"><span class="co">#&gt;   b     chr2    2    3    4      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb14-29" data-line-number="29"><span class="co">#&gt;   ---</span></a>
<a class="sourceLine" id="cb14-30" data-line-number="30"><span class="co">#&gt;   seqinfo: 3 sequences from mock1 genome</span></a>
<a class="sourceLine" id="cb14-31" data-line-number="31"><span class="kw">tail</span>(gt3, <span class="dt">n =</span> <span class="dv">2</span>)</a>
<a class="sourceLine" id="cb14-32" data-line-number="32"><span class="co">#&gt; GTuples object with 2 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb14-33" data-line-number="33"><span class="co">#&gt;     seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb14-34" data-line-number="34"><span class="co">#&gt;   i     chr3    9   10   11      - |     9 0.111111111111111</span></a>
<a class="sourceLine" id="cb14-35" data-line-number="35"><span class="co">#&gt;   j     chr3   10   11   12      - |    10                 0</span></a>
<a class="sourceLine" id="cb14-36" data-line-number="36"><span class="co">#&gt;   ---</span></a>
<a class="sourceLine" id="cb14-37" data-line-number="37"><span class="co">#&gt;   seqinfo: 3 sequences from mock1 genome</span></a>
<a class="sourceLine" id="cb14-38" data-line-number="38"><span class="kw">window</span>(gt3, <span class="dt">start =</span> <span class="dv">2</span>, <span class="dt">end =</span> <span class="dv">4</span>)</a>
<a class="sourceLine" id="cb14-39" data-line-number="39"><span class="co">#&gt; GTuples object with 3 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb14-40" data-line-number="40"><span class="co">#&gt;     seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb14-41" data-line-number="41"><span class="co">#&gt;   b     chr2    2    3    4      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb14-42" data-line-number="42"><span class="co">#&gt;   c     chr2    3    4    5      + |     3 0.777777777777778</span></a>
<a class="sourceLine" id="cb14-43" data-line-number="43"><span class="co">#&gt;   d     chr2    4    5    6      * |     4 0.666666666666667</span></a>
<a class="sourceLine" id="cb14-44" data-line-number="44"><span class="co">#&gt;   ---</span></a>
<a class="sourceLine" id="cb14-45" data-line-number="45"><span class="co">#&gt;   seqinfo: 3 sequences from mock1 genome</span></a></code></pre></div>
</div>
<div id="basic-tuple-operations-for-gtuples-objects" class="section level3">
<h3 class="hasAnchor">
<a href="#basic-tuple-operations-for-gtuples-objects" class="anchor"></a>Basic tuple operations for <em>GTuples</em> objects</h3>
<p>Basic tuple characteristics of <em>GTuples</em> objects can be extracted using the <code>start</code>, <code>end</code>, and <code>tuples</code> methods. <strong>Warning: While the <code>width</code> method is well-defined, namely as <span class="math inline">\(pos_{m} - pos_{1} + 1\)</span>, this may not be what is required. Instead, please see the <code>IPD</code> method that will be discussed in the next section</strong>.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1"><span class="kw">start</span>(gt3)</a>
<a class="sourceLine" id="cb15-2" data-line-number="2"><span class="co">#&gt;  [1]  1  2  3  4  5  6  7  8  9 10</span></a>
<a class="sourceLine" id="cb15-3" data-line-number="3"><span class="kw">end</span>(gt3)</a>
<a class="sourceLine" id="cb15-4" data-line-number="4"><span class="co">#&gt;  [1]  3  4  5  6  7  8  9 10 11 12</span></a>
<a class="sourceLine" id="cb15-5" data-line-number="5"><span class="kw"><a href="../reference/GTuples-class.html">tuples</a></span>(gt3)</a>
<a class="sourceLine" id="cb15-6" data-line-number="6"><span class="co">#&gt;       pos1 pos2 pos3</span></a>
<a class="sourceLine" id="cb15-7" data-line-number="7"><span class="co">#&gt;  [1,]    1    2    3</span></a>
<a class="sourceLine" id="cb15-8" data-line-number="8"><span class="co">#&gt;  [2,]    2    3    4</span></a>
<a class="sourceLine" id="cb15-9" data-line-number="9"><span class="co">#&gt;  [3,]    3    4    5</span></a>
<a class="sourceLine" id="cb15-10" data-line-number="10"><span class="co">#&gt;  [4,]    4    5    6</span></a>
<a class="sourceLine" id="cb15-11" data-line-number="11"><span class="co">#&gt;  [5,]    5    6    7</span></a>
<a class="sourceLine" id="cb15-12" data-line-number="12"><span class="co">#&gt;  [6,]    6    7    8</span></a>
<a class="sourceLine" id="cb15-13" data-line-number="13"><span class="co">#&gt;  [7,]    7    8    9</span></a>
<a class="sourceLine" id="cb15-14" data-line-number="14"><span class="co">#&gt;  [8,]    8    9   10</span></a>
<a class="sourceLine" id="cb15-15" data-line-number="15"><span class="co">#&gt;  [9,]    9   10   11</span></a>
<a class="sourceLine" id="cb15-16" data-line-number="16"><span class="co">#&gt; [10,]   10   11   12</span></a></code></pre></div>
<div id="intra-tuple-operations" class="section level4">
<h4 class="hasAnchor">
<a href="#intra-tuple-operations" class="anchor"></a>Intra-tuple operations</h4>
<p>Most of the intra-range methods defined for <em>GenomicRanges::GRanges</em> objects are not currently defined via extension for <em>GTuples</em> objects due to the differences between <em>ranges</em> and <em>tuples</em>. Those not currently defined, and which return an error message, are:</p>
<ul>
<li><code>narrow</code></li>
<li><code>flank</code></li>
<li><code>promoters</code></li>
<li><code>resize</code></li>
<li><code>Ops</code></li>
</ul>
<p>I am happy to add these methods if appropriate, so please contact me if you have suggestions for good definitions.</p>
<p>Both the <code>trim</code> and <code>shift</code> methods are well-defined, although the former is somewhat limited since it will return an error if the <em>internal positions</em> exceed the <code>seqlengths</code>:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1"><span class="kw"><a href="../reference/intra-tuple-methods.html">shift</a></span>(gt3, <span class="dv">500</span>)</a>
<a class="sourceLine" id="cb16-2" data-line-number="2"><span class="co">#&gt; GTuples object with 10 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb16-3" data-line-number="3"><span class="co">#&gt;     seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb16-4" data-line-number="4"><span class="co">#&gt;   a     chr1  501  502  503      - |     1                 1</span></a>
<a class="sourceLine" id="cb16-5" data-line-number="5"><span class="co">#&gt;   b     chr2  502  503  504      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb16-6" data-line-number="6"><span class="co">#&gt;   c     chr2  503  504  505      + |     3 0.777777777777778</span></a>
<a class="sourceLine" id="cb16-7" data-line-number="7"><span class="co">#&gt;   d     chr2  504  505  506      * |     4 0.666666666666667</span></a>
<a class="sourceLine" id="cb16-8" data-line-number="8"><span class="co">#&gt;   e     chr1  505  506  507      * |     5 0.555555555555556</span></a>
<a class="sourceLine" id="cb16-9" data-line-number="9"><span class="co">#&gt;   f     chr1  506  507  508      + |     6 0.444444444444444</span></a>
<a class="sourceLine" id="cb16-10" data-line-number="10"><span class="co">#&gt;   g     chr3  507  508  509      + |     7 0.333333333333333</span></a>
<a class="sourceLine" id="cb16-11" data-line-number="11"><span class="co">#&gt;   h     chr3  508  509  510      + |     8 0.222222222222222</span></a>
<a class="sourceLine" id="cb16-12" data-line-number="12"><span class="co">#&gt;   i     chr3  509  510  511      - |     9 0.111111111111111</span></a>
<a class="sourceLine" id="cb16-13" data-line-number="13"><span class="co">#&gt;   j     chr3  510  511  512      - |    10                 0</span></a>
<a class="sourceLine" id="cb16-14" data-line-number="14"><span class="co">#&gt;   ---</span></a>
<a class="sourceLine" id="cb16-15" data-line-number="15"><span class="co">#&gt;   seqinfo: 3 sequences from mock1 genome</span></a></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="co"># Raises warning due to tuple being outside of seqlength</span></a>
<a class="sourceLine" id="cb17-2" data-line-number="2">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/intra-tuple-methods.html">shift</a></span>(gt3[<span class="dv">1</span>], <span class="dv">999</span>)</a>
<a class="sourceLine" id="cb17-3" data-line-number="3"><span class="co">#&gt; Warning in valid.GenomicRanges.seqinfo(x, suggest.trim = TRUE): GRanges object contains 1 out-of-bound range located on sequence</span></a>
<a class="sourceLine" id="cb17-4" data-line-number="4"><span class="co">#&gt;   chr1. Note that ranges located on a sequence whose length is</span></a>
<a class="sourceLine" id="cb17-5" data-line-number="5"><span class="co">#&gt;   unknown (NA) or on a circular sequence are not considered</span></a>
<a class="sourceLine" id="cb17-6" data-line-number="6"><span class="co">#&gt;   out-of-bound (use seqlengths() and isCircular() to get the lengths</span></a>
<a class="sourceLine" id="cb17-7" data-line-number="7"><span class="co">#&gt;   and circularity flags of the underlying sequences). You can use</span></a>
<a class="sourceLine" id="cb17-8" data-line-number="8"><span class="co">#&gt;   trim() to trim these ranges. See ?`trim,GenomicRanges-method` for</span></a>
<a class="sourceLine" id="cb17-9" data-line-number="9"><span class="co">#&gt;   more information.</span></a>
<a class="sourceLine" id="cb17-10" data-line-number="10">x</a>
<a class="sourceLine" id="cb17-11" data-line-number="11"><span class="co">#&gt; GTuples object with 1 x 3-tuple and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb17-12" data-line-number="12"><span class="co">#&gt;     seqnames pos1 pos2 pos3 strand | score GC</span></a>
<a class="sourceLine" id="cb17-13" data-line-number="13"><span class="co">#&gt;   a     chr1 1000 1001 1002      - |     1  1</span></a>
<a class="sourceLine" id="cb17-14" data-line-number="14"><span class="co">#&gt;   ---</span></a>
<a class="sourceLine" id="cb17-15" data-line-number="15"><span class="co">#&gt;   seqinfo: 3 sequences from mock1 genome</span></a>
<a class="sourceLine" id="cb17-16" data-line-number="16"></a>
<a class="sourceLine" id="cb17-17" data-line-number="17"><span class="co"># Returns an error because internal position exceeds sequence length, resulting </span></a>
<a class="sourceLine" id="cb17-18" data-line-number="18"><span class="co"># in a malformed tuple when trimmed.</span></a>
<a class="sourceLine" id="cb17-19" data-line-number="19"><span class="kw">trim</span>(x)</a>
<a class="sourceLine" id="cb17-20" data-line-number="20"><span class="co">#&gt; Error in validObject(object): invalid class "GTuples" object: positions in each tuple must be sorted in strictly increasing order, i.e. 'pos1' &lt; ... &lt; 'pos3'</span></a></code></pre></div>
</div>
<div id="inter-tuple-operations" class="section level4">
<h4 class="hasAnchor">
<a href="#inter-tuple-operations" class="anchor"></a>Inter-tuple operations</h4>
<p>None of the inter-range methods defined for <em>GenomicRanges::GRanges</em> objects are currently defined via extension for <em>GTuples</em> objects due to the differences between <em>ranges</em> and <em>tuples</em>. Those not currently defined, and which return an error message, are:</p>
<ul>
<li><code>range</code></li>
<li><code>reduce</code></li>
<li><code>gaps</code></li>
<li><code>disjoin</code></li>
<li><code>isDisjoint</code></li>
<li><code>disjointBins</code></li>
</ul>
<p>I am happy to add these methods if appropriate, so please contact me if you have suggestions for good definitions.</p>
</div>
<div id="interval-set-operations-for-gtuples-objects" class="section level4">
<h4 class="hasAnchor">
<a href="#interval-set-operations-for-gtuples-objects" class="anchor"></a>Interval set operations for <em>GTuples</em> objects</h4>
<p>None of the interval set operations defined for <em>GenomicRanges::GRanges</em> objects are currently defined via extension for <em>GTuples</em> objects due to the differences between ranges and tuples. Those not currently defined, and which return an error message, are:</p>
<ul>
<li><code>union</code></li>
<li><code>intersect</code></li>
<li><code>setdiff</code></li>
<li><code>punion</code></li>
<li><code>pintersect</code></li>
<li><code>psetdiff</code></li>
</ul>
<p>I am happy to add these methods if appropriate, so please contact me if you have suggestions for good definitions.</p>
</div>
</div>
<div id="additional-methods-unique-to-gtuples" class="section level3">
<h3 class="hasAnchor">
<a href="#additional-methods-unique-to-gtuples" class="anchor"></a>Additional methods unique to <em>GTuples</em>
</h3>
<p><em>GTuples</em> have a few specifically defined methods that do not exist for <em>GenomicRanges::GRanges</em>. These are <code>tuples</code>, <code>size</code> and <code>IPD</code>.</p>
<p>The <code>tuples</code> method we have already seen and is somewhat analogous to the <code>ranges</code> method for <em>GenomicRanges::GRanges</em>, although returning an integer matrix rather than an <em>IRanges::IRanges</em> object:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1"><span class="kw"><a href="../reference/GTuples-class.html">tuples</a></span>(gt3)</a>
<a class="sourceLine" id="cb18-2" data-line-number="2"><span class="co">#&gt;       pos1 pos2 pos3</span></a>
<a class="sourceLine" id="cb18-3" data-line-number="3"><span class="co">#&gt;  [1,]    1    2    3</span></a>
<a class="sourceLine" id="cb18-4" data-line-number="4"><span class="co">#&gt;  [2,]    2    3    4</span></a>
<a class="sourceLine" id="cb18-5" data-line-number="5"><span class="co">#&gt;  [3,]    3    4    5</span></a>
<a class="sourceLine" id="cb18-6" data-line-number="6"><span class="co">#&gt;  [4,]    4    5    6</span></a>
<a class="sourceLine" id="cb18-7" data-line-number="7"><span class="co">#&gt;  [5,]    5    6    7</span></a>
<a class="sourceLine" id="cb18-8" data-line-number="8"><span class="co">#&gt;  [6,]    6    7    8</span></a>
<a class="sourceLine" id="cb18-9" data-line-number="9"><span class="co">#&gt;  [7,]    7    8    9</span></a>
<a class="sourceLine" id="cb18-10" data-line-number="10"><span class="co">#&gt;  [8,]    8    9   10</span></a>
<a class="sourceLine" id="cb18-11" data-line-number="11"><span class="co">#&gt;  [9,]    9   10   11</span></a>
<a class="sourceLine" id="cb18-12" data-line-number="12"><span class="co">#&gt; [10,]   10   11   12</span></a></code></pre></div>
<p>The <code>size</code> method returns the size of the tuples stored in the object:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1"><span class="kw"><a href="../reference/GTuples-class.html">size</a></span>(gt3)</a>
<a class="sourceLine" id="cb19-2" data-line-number="2"><span class="co">#&gt; [1] 3</span></a></code></pre></div>
<p>Every m-tuple with <span class="math inline">\(m \geq 2\)</span> has an associated vector of intra-pair distances (<span class="math inline">\(IPD\)</span>). This is defined as <span class="math inline">\(IPD = (pos_{2} - pos_{1}, \ldots, pos_{m} - pos_{m - 1})\)</span>. The <code>IPD</code> method returns this as an integer matrix, where the <span class="math inline">\(i^{th}\)</span> row contains the <span class="math inline">\(IPD\)</span> for the <span class="math inline">\(i^{th}\)</span> tuple:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1"><span class="kw"><a href="../reference/GTuples-class.html">IPD</a></span>(gt3)</a>
<a class="sourceLine" id="cb20-2" data-line-number="2"><span class="co">#&gt;       [,1] [,2]</span></a>
<a class="sourceLine" id="cb20-3" data-line-number="3"><span class="co">#&gt;  [1,]    1    1</span></a>
<a class="sourceLine" id="cb20-4" data-line-number="4"><span class="co">#&gt;  [2,]    1    1</span></a>
<a class="sourceLine" id="cb20-5" data-line-number="5"><span class="co">#&gt;  [3,]    1    1</span></a>
<a class="sourceLine" id="cb20-6" data-line-number="6"><span class="co">#&gt;  [4,]    1    1</span></a>
<a class="sourceLine" id="cb20-7" data-line-number="7"><span class="co">#&gt;  [5,]    1    1</span></a>
<a class="sourceLine" id="cb20-8" data-line-number="8"><span class="co">#&gt;  [6,]    1    1</span></a>
<a class="sourceLine" id="cb20-9" data-line-number="9"><span class="co">#&gt;  [7,]    1    1</span></a>
<a class="sourceLine" id="cb20-10" data-line-number="10"><span class="co">#&gt;  [8,]    1    1</span></a>
<a class="sourceLine" id="cb20-11" data-line-number="11"><span class="co">#&gt;  [9,]    1    1</span></a>
<a class="sourceLine" id="cb20-12" data-line-number="12"><span class="co">#&gt; [10,]    1    1</span></a></code></pre></div>
</div>
</div>
<div id="implementation-details" class="section level2">
<h2 class="hasAnchor">
<a href="#implementation-details" class="anchor"></a>Implementation details</h2>
<p>While the <em>GTuples</em> class can be thought of as a matrix-link object, with the number of columns equal to the <code>size</code> of the tuples plus two (one for the <code>seqname</code> and one for the <code>strand</code>), internally, it extends the <em>GenomicRanges::GRanges</em> class. Specifically, the <code>ranges</code> slot stores an <em>IRanges::IRanges</em> object containing <span class="math inline">\(pos_{1}\)</span> and <span class="math inline">\(pos_{m}\)</span> and, if <code>size</code> <span class="math inline">\(&gt; 2\)</span>, a matrix is used to store the co-ordinates of the “internal positions”, <span class="math inline">\(pos_{2}, \ldots, pos_{m - 1}\)</span> in the <code>internalPos</code> slot. If <code>size</code> <span class="math inline">\(\leq 2\)</span> then the <code>internalPos</code> slot is set to <code>NULL</code>. The <code>size</code> is stored as an integer in the <code>size</code> slot.</p>
<p>While there are arguments for creating stand-alone <em>GTuples</em> and <em>GTuplesList</em> classes, by extending the <em>GenomicRanges::GRanges</em> and <em>GenomicRanges::GRangesList</em> classes we get a lot of very useful functionality “for free” via appropriately defined inheritance.</p>
</div>
</div>
<div id="gtupleslist" class="section level1">
<h1 class="hasAnchor">
<a href="#gtupleslist" class="anchor"></a><em>GTuplesList</em>
</h1>
<p>The <em>GTuplesList</em> class is a container to store a <em>S4Vectors::List</em> of <em>GTuples</em> objects. It extends the <em>GenomicRanges::GRangesList</em> class.</p>
<p>Currently, all <em>GTuples</em> in a <em>GTuplesList</em> must have the same <code>size</code><a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>. I expect that users will mostly use <em>GTuples</em> objects and have little need to directly use <em>GTuplesList</em> objects.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1">seqinfo &lt;-<span class="st"> </span><span class="kw">Seqinfo</span>(<span class="kw">paste0</span>(<span class="st">"chr"</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>), <span class="kw">c</span>(<span class="dv">1000</span>, <span class="dv">2000</span>, <span class="dv">1500</span>), <span class="ot">NA</span>, <span class="st">"mock1"</span>)</a>
<a class="sourceLine" id="cb21-2" data-line-number="2">gt3 &lt;-<span class="st"> </span><span class="kw">GTuples</span>(<span class="dt">seqnames =</span> <span class="kw">Rle</span>(<span class="kw">c</span>(<span class="st">"chr1"</span>, <span class="st">"chr2"</span>, <span class="st">"chr1"</span>, <span class="st">"chr3"</span>),</a>
<a class="sourceLine" id="cb21-3" data-line-number="3">                              <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">4</span>)),</a>
<a class="sourceLine" id="cb21-4" data-line-number="4">               <span class="dt">tuples =</span> <span class="kw">matrix</span>(<span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">2</span><span class="op">:</span><span class="dv">11</span>, <span class="dv">3</span><span class="op">:</span><span class="dv">12</span>), <span class="dt">ncol =</span> <span class="dv">3</span>),</a>
<a class="sourceLine" id="cb21-5" data-line-number="5">               <span class="dt">strand =</span> <span class="kw">Rle</span>(<span class="kw">strand</span>(<span class="kw">c</span>(<span class="st">"-"</span>, <span class="st">"+"</span>, <span class="st">"*"</span>, <span class="st">"+"</span>, <span class="st">"-"</span>)),</a>
<a class="sourceLine" id="cb21-6" data-line-number="6">                            <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">2</span>)),</a>
<a class="sourceLine" id="cb21-7" data-line-number="7">               <span class="dt">score =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">GC =</span> <span class="kw">seq</span>(<span class="dv">1</span>, <span class="dv">0</span>, <span class="dt">length =</span> <span class="dv">10</span>), <span class="dt">seqinfo =</span> seqinfo)</a>
<a class="sourceLine" id="cb21-8" data-line-number="8">gtl3 &lt;-<span class="st"> </span><span class="kw">GTuplesList</span>(<span class="dt">A =</span> gt3[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>], <span class="dt">B =</span> gt3[<span class="dv">6</span><span class="op">:</span><span class="dv">10</span>])</a>
<a class="sourceLine" id="cb21-9" data-line-number="9">gtl3</a>
<a class="sourceLine" id="cb21-10" data-line-number="10"><span class="co">#&gt; GTuplesList object of length 2:</span></a>
<a class="sourceLine" id="cb21-11" data-line-number="11"><span class="co">#&gt; $A </span></a>
<a class="sourceLine" id="cb21-12" data-line-number="12"><span class="co">#&gt; GTuples object with 5 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb21-13" data-line-number="13"><span class="co">#&gt;       seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb21-14" data-line-number="14"><span class="co">#&gt;   [1]     chr1    1    2    3      - |     1                 1</span></a>
<a class="sourceLine" id="cb21-15" data-line-number="15"><span class="co">#&gt;   [2]     chr2    2    3    4      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb21-16" data-line-number="16"><span class="co">#&gt;   [3]     chr2    3    4    5      + |     3 0.777777777777778</span></a>
<a class="sourceLine" id="cb21-17" data-line-number="17"><span class="co">#&gt;   [4]     chr2    4    5    6      * |     4 0.666666666666667</span></a>
<a class="sourceLine" id="cb21-18" data-line-number="18"><span class="co">#&gt;   [5]     chr1    5    6    7      * |     5 0.555555555555556</span></a>
<a class="sourceLine" id="cb21-19" data-line-number="19"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb21-20" data-line-number="20"><span class="co">#&gt; $B </span></a>
<a class="sourceLine" id="cb21-21" data-line-number="21"><span class="co">#&gt; GTuples object with 5 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb21-22" data-line-number="22"><span class="co">#&gt;       seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb21-23" data-line-number="23"><span class="co">#&gt;   [1]     chr1    6    7    8      + |     6 0.444444444444444</span></a>
<a class="sourceLine" id="cb21-24" data-line-number="24"><span class="co">#&gt;   [2]     chr3    7    8    9      + |     7 0.333333333333333</span></a>
<a class="sourceLine" id="cb21-25" data-line-number="25"><span class="co">#&gt;   [3]     chr3    8    9   10      + |     8 0.222222222222222</span></a>
<a class="sourceLine" id="cb21-26" data-line-number="26"><span class="co">#&gt;   [4]     chr3    9   10   11      - |     9 0.111111111111111</span></a>
<a class="sourceLine" id="cb21-27" data-line-number="27"><span class="co">#&gt;   [5]     chr3   10   11   12      - |    10                 0</span></a>
<a class="sourceLine" id="cb21-28" data-line-number="28"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb21-29" data-line-number="29"><span class="co">#&gt; -------</span></a>
<a class="sourceLine" id="cb21-30" data-line-number="30"><span class="co">#&gt; seqinfo: 3 sequences from mock1 genome</span></a></code></pre></div>
<p>For even more information on the <em>GTuplesList</em> class, be sure to consult the documentation:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1">?GTuplesList</a></code></pre></div>
<div id="gtupleslist-methods" class="section level2">
<h2 class="hasAnchor">
<a href="#gtupleslist-methods" class="anchor"></a><em>GTuplesList</em> methods</h2>
<p>Most methods defined for <em>GenomicRanges::GRangesList</em> are also applicable to <em>GTuplesList</em>. Those that are not yet defined, which are those that make sense for ranges but generally not for tuples, return error messages.</p>
<p>If a method that is not defined for <em>GTuplesList</em> but is defined for <em>GenomicRanges::GRangesList</em> is truly required, then this can be achieved by first coercing the <em>GTuplesList</em> object to a <em>GenomicRanges::GRangesList</em> object, noting that this is generally a destructive operation:</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1"><span class="kw">as</span>(gtl3, <span class="st">"GRangesList"</span>)</a>
<a class="sourceLine" id="cb23-2" data-line-number="2"><span class="co">#&gt; GRangesList object of length 2:</span></a>
<a class="sourceLine" id="cb23-3" data-line-number="3"><span class="co">#&gt; $A </span></a>
<a class="sourceLine" id="cb23-4" data-line-number="4"><span class="co">#&gt; GRanges object with 5 ranges and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb23-5" data-line-number="5"><span class="co">#&gt;       seqnames    ranges strand |     score                GC</span></a>
<a class="sourceLine" id="cb23-6" data-line-number="6"><span class="co">#&gt;          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;         &lt;numeric&gt;</span></a>
<a class="sourceLine" id="cb23-7" data-line-number="7"><span class="co">#&gt;   [1]     chr1       1-3      - |         1                 1</span></a>
<a class="sourceLine" id="cb23-8" data-line-number="8"><span class="co">#&gt;   [2]     chr2       2-4      + |         2 0.888888888888889</span></a>
<a class="sourceLine" id="cb23-9" data-line-number="9"><span class="co">#&gt;   [3]     chr2       3-5      + |         3 0.777777777777778</span></a>
<a class="sourceLine" id="cb23-10" data-line-number="10"><span class="co">#&gt;   [4]     chr2       4-6      * |         4 0.666666666666667</span></a>
<a class="sourceLine" id="cb23-11" data-line-number="11"><span class="co">#&gt;   [5]     chr1       5-7      * |         5 0.555555555555556</span></a>
<a class="sourceLine" id="cb23-12" data-line-number="12"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb23-13" data-line-number="13"><span class="co">#&gt; $B </span></a>
<a class="sourceLine" id="cb23-14" data-line-number="14"><span class="co">#&gt; GRanges object with 5 ranges and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb23-15" data-line-number="15"><span class="co">#&gt;       seqnames ranges strand | score                GC</span></a>
<a class="sourceLine" id="cb23-16" data-line-number="16"><span class="co">#&gt;   [1]     chr1    6-8      + |     6 0.444444444444444</span></a>
<a class="sourceLine" id="cb23-17" data-line-number="17"><span class="co">#&gt;   [2]     chr3    7-9      + |     7 0.333333333333333</span></a>
<a class="sourceLine" id="cb23-18" data-line-number="18"><span class="co">#&gt;   [3]     chr3   8-10      + |     8 0.222222222222222</span></a>
<a class="sourceLine" id="cb23-19" data-line-number="19"><span class="co">#&gt;   [4]     chr3   9-11      - |     9 0.111111111111111</span></a>
<a class="sourceLine" id="cb23-20" data-line-number="20"><span class="co">#&gt;   [5]     chr3  10-12      - |    10                 0</span></a>
<a class="sourceLine" id="cb23-21" data-line-number="21"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb23-22" data-line-number="22"><span class="co">#&gt; -------</span></a>
<a class="sourceLine" id="cb23-23" data-line-number="23"><span class="co">#&gt; seqinfo: 3 sequences from mock1 genome</span></a></code></pre></div>
<div id="basic-gtupleslist-accessors" class="section level3">
<h3 class="hasAnchor">
<a href="#basic-gtupleslist-accessors" class="anchor"></a>Basic <em>GTuplesList</em> accessors}</h3>
<p>These are very similar to those available for <em>GTuples</em> objects, except that they typically return a list since the input is now essentially a list of <em>GTuples</em> objects:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1"><span class="kw">seqnames</span>(gtl3)</a>
<a class="sourceLine" id="cb24-2" data-line-number="2"><span class="co">#&gt; RleList of length 2</span></a>
<a class="sourceLine" id="cb24-3" data-line-number="3"><span class="co">#&gt; $A</span></a>
<a class="sourceLine" id="cb24-4" data-line-number="4"><span class="co">#&gt; factor-Rle of length 5 with 3 runs</span></a>
<a class="sourceLine" id="cb24-5" data-line-number="5"><span class="co">#&gt;   Lengths:    1    3    1</span></a>
<a class="sourceLine" id="cb24-6" data-line-number="6"><span class="co">#&gt;   Values : chr1 chr2 chr1</span></a>
<a class="sourceLine" id="cb24-7" data-line-number="7"><span class="co">#&gt; Levels(3): chr1 chr2 chr3</span></a>
<a class="sourceLine" id="cb24-8" data-line-number="8"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb24-9" data-line-number="9"><span class="co">#&gt; $B</span></a>
<a class="sourceLine" id="cb24-10" data-line-number="10"><span class="co">#&gt; factor-Rle of length 5 with 2 runs</span></a>
<a class="sourceLine" id="cb24-11" data-line-number="11"><span class="co">#&gt;   Lengths:    1    4</span></a>
<a class="sourceLine" id="cb24-12" data-line-number="12"><span class="co">#&gt;   Values : chr1 chr3</span></a>
<a class="sourceLine" id="cb24-13" data-line-number="13"><span class="co">#&gt; Levels(3): chr1 chr2 chr3</span></a>
<a class="sourceLine" id="cb24-14" data-line-number="14"><span class="co"># Returns a list of integer matrices</span></a>
<a class="sourceLine" id="cb24-15" data-line-number="15"><span class="kw"><a href="../reference/GTuples-class.html">tuples</a></span>(gtl3)</a>
<a class="sourceLine" id="cb24-16" data-line-number="16"><span class="co">#&gt; List of length 2</span></a>
<a class="sourceLine" id="cb24-17" data-line-number="17"><span class="co">#&gt; names(2): A B</span></a>
<a class="sourceLine" id="cb24-18" data-line-number="18"><span class="kw"><a href="../reference/GTuples-class.html">tuples</a></span>(gtl3)[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb24-19" data-line-number="19"><span class="co">#&gt;      pos1 pos2 pos3</span></a>
<a class="sourceLine" id="cb24-20" data-line-number="20"><span class="co">#&gt; [1,]    1    2    3</span></a>
<a class="sourceLine" id="cb24-21" data-line-number="21"><span class="co">#&gt; [2,]    2    3    4</span></a>
<a class="sourceLine" id="cb24-22" data-line-number="22"><span class="co">#&gt; [3,]    3    4    5</span></a>
<a class="sourceLine" id="cb24-23" data-line-number="23"><span class="co">#&gt; [4,]    4    5    6</span></a>
<a class="sourceLine" id="cb24-24" data-line-number="24"><span class="co">#&gt; [5,]    5    6    7</span></a>
<a class="sourceLine" id="cb24-25" data-line-number="25"><span class="kw">strand</span>(gtl3)</a>
<a class="sourceLine" id="cb24-26" data-line-number="26"><span class="co">#&gt; RleList of length 2</span></a>
<a class="sourceLine" id="cb24-27" data-line-number="27"><span class="co">#&gt; $A</span></a>
<a class="sourceLine" id="cb24-28" data-line-number="28"><span class="co">#&gt; factor-Rle of length 5 with 3 runs</span></a>
<a class="sourceLine" id="cb24-29" data-line-number="29"><span class="co">#&gt;   Lengths: 1 2 2</span></a>
<a class="sourceLine" id="cb24-30" data-line-number="30"><span class="co">#&gt;   Values : - + *</span></a>
<a class="sourceLine" id="cb24-31" data-line-number="31"><span class="co">#&gt; Levels(3): + - *</span></a>
<a class="sourceLine" id="cb24-32" data-line-number="32"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb24-33" data-line-number="33"><span class="co">#&gt; $B</span></a>
<a class="sourceLine" id="cb24-34" data-line-number="34"><span class="co">#&gt; factor-Rle of length 5 with 2 runs</span></a>
<a class="sourceLine" id="cb24-35" data-line-number="35"><span class="co">#&gt;   Lengths: 3 2</span></a>
<a class="sourceLine" id="cb24-36" data-line-number="36"><span class="co">#&gt;   Values : + -</span></a>
<a class="sourceLine" id="cb24-37" data-line-number="37"><span class="co">#&gt; Levels(3): + - *</span></a></code></pre></div>
<p>The <code>length</code> and <code>names</code> methods will return the length and names of the list, respectively:</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" data-line-number="1"><span class="kw">length</span>(gtl3)</a>
<a class="sourceLine" id="cb25-2" data-line-number="2"><span class="co">#&gt; [1] 2</span></a>
<a class="sourceLine" id="cb25-3" data-line-number="3"><span class="kw">names</span>(gtl3)</a>
<a class="sourceLine" id="cb25-4" data-line-number="4"><span class="co">#&gt; [1] "A" "B"</span></a></code></pre></div>
<p><em>Seqinfo</em> can be extracted using the <code>seqinfo</code> accessor:</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" data-line-number="1"><span class="kw">seqinfo</span>(gtl3)</a>
<a class="sourceLine" id="cb26-2" data-line-number="2"><span class="co">#&gt; Seqinfo object with 3 sequences from mock1 genome:</span></a>
<a class="sourceLine" id="cb26-3" data-line-number="3"><span class="co">#&gt;   seqnames seqlengths isCircular genome</span></a>
<a class="sourceLine" id="cb26-4" data-line-number="4"><span class="co">#&gt;   chr1           1000         NA  mock1</span></a>
<a class="sourceLine" id="cb26-5" data-line-number="5"><span class="co">#&gt;   chr2           2000         NA  mock1</span></a>
<a class="sourceLine" id="cb26-6" data-line-number="6"><span class="co">#&gt;   chr3           1500         NA  mock1</span></a></code></pre></div>
<p>The <code>elementNROWS</code> method returns a list of integers corresponding to the result of calling <code>length</code> on each individual <em>GTuples</em> object contained by the <em>GTuplesList</em>. This is a faster alternative to calling <code>lapply</code> on the <em>GTuplesList</em>:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" data-line-number="1"><span class="kw">elementNROWS</span>(gtl3)</a>
<a class="sourceLine" id="cb27-2" data-line-number="2"><span class="co">#&gt; A B </span></a>
<a class="sourceLine" id="cb27-3" data-line-number="3"><span class="co">#&gt; 5 5</span></a></code></pre></div>
<p>You can also use <code>isEmpty</code> to test if a <em>GTuplesList</em> object contains anything:</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" data-line-number="1"><span class="kw">isEmpty</span>(gtl3)</a>
<a class="sourceLine" id="cb28-2" data-line-number="2"><span class="co">#&gt; [1] FALSE</span></a>
<a class="sourceLine" id="cb28-3" data-line-number="3"><span class="kw">isEmpty</span>(<span class="kw">GTuplesList</span>())</a>
<a class="sourceLine" id="cb28-4" data-line-number="4"><span class="co">#&gt; [1] TRUE</span></a></code></pre></div>
<p>Finally, in the context of a <em>GTuplesList</em> object, the <code>mcols</code> method performs a similar operation to what it does on a <em>GTuples</em> object. However, this metadata now refers to information at the list level instead of the level of the individual <em>GTuples</em> objects:</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" data-line-number="1"><span class="kw">mcols</span>(gtl3) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"Feature A"</span>, <span class="st">"Feature B"</span>)</a>
<a class="sourceLine" id="cb29-2" data-line-number="2"><span class="kw">mcols</span>(gtl3)</a>
<a class="sourceLine" id="cb29-3" data-line-number="3"><span class="co">#&gt; DataFrame with 2 rows and 1 column</span></a>
<a class="sourceLine" id="cb29-4" data-line-number="4"><span class="co">#&gt;         value</span></a>
<a class="sourceLine" id="cb29-5" data-line-number="5"><span class="co">#&gt;   &lt;character&gt;</span></a>
<a class="sourceLine" id="cb29-6" data-line-number="6"><span class="co">#&gt; 1   Feature A</span></a>
<a class="sourceLine" id="cb29-7" data-line-number="7"><span class="co">#&gt; 2   Feature B</span></a></code></pre></div>
</div>
<div id="combining-gtupleslist-objects" class="section level3">
<h3 class="hasAnchor">
<a href="#combining-gtupleslist-objects" class="anchor"></a>Combining <em>GTuplesList</em> objects</h3>
<p><em>GTuplesList</em> objects can be unlisted to combine the separate <em>GTuples</em> objects that they contain as an expanded <em>GTuples</em>:</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" data-line-number="1">ul &lt;-<span class="st"> </span><span class="kw">unlist</span>(gtl3)</a>
<a class="sourceLine" id="cb30-2" data-line-number="2">ul</a>
<a class="sourceLine" id="cb30-3" data-line-number="3"><span class="co">#&gt; GTuples object with 10 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb30-4" data-line-number="4"><span class="co">#&gt;     seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb30-5" data-line-number="5"><span class="co">#&gt;   A     chr1    1    2    3      - |     1                 1</span></a>
<a class="sourceLine" id="cb30-6" data-line-number="6"><span class="co">#&gt;   A     chr2    2    3    4      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb30-7" data-line-number="7"><span class="co">#&gt;   A     chr2    3    4    5      + |     3 0.777777777777778</span></a>
<a class="sourceLine" id="cb30-8" data-line-number="8"><span class="co">#&gt;   A     chr2    4    5    6      * |     4 0.666666666666667</span></a>
<a class="sourceLine" id="cb30-9" data-line-number="9"><span class="co">#&gt;   A     chr1    5    6    7      * |     5 0.555555555555556</span></a>
<a class="sourceLine" id="cb30-10" data-line-number="10"><span class="co">#&gt;   B     chr1    6    7    8      + |     6 0.444444444444444</span></a>
<a class="sourceLine" id="cb30-11" data-line-number="11"><span class="co">#&gt;   B     chr3    7    8    9      + |     7 0.333333333333333</span></a>
<a class="sourceLine" id="cb30-12" data-line-number="12"><span class="co">#&gt;   B     chr3    8    9   10      + |     8 0.222222222222222</span></a>
<a class="sourceLine" id="cb30-13" data-line-number="13"><span class="co">#&gt;   B     chr3    9   10   11      - |     9 0.111111111111111</span></a>
<a class="sourceLine" id="cb30-14" data-line-number="14"><span class="co">#&gt;   B     chr3   10   11   12      - |    10                 0</span></a>
<a class="sourceLine" id="cb30-15" data-line-number="15"><span class="co">#&gt;   ---</span></a>
<a class="sourceLine" id="cb30-16" data-line-number="16"><span class="co">#&gt;   seqinfo: 3 sequences from mock1 genome</span></a></code></pre></div>
<p>You can also combine <em>GTuplesList</em> objects together using <code>append</code> or <code>c</code>.</p>
</div>
<div id="subsetting-gtupleslist-objects" class="section level3">
<h3 class="hasAnchor">
<a href="#subsetting-gtupleslist-objects" class="anchor"></a>Subsetting <em>GTuplesList</em> objects</h3>
<p>Subsetting of <em>GTuplesList</em> objects is identical to subsetting of <em>GenomicRanges::GRangesList</em> objects:</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb31-1" data-line-number="1">gtl3[<span class="dv">1</span>]</a>
<a class="sourceLine" id="cb31-2" data-line-number="2"><span class="co">#&gt; GTuplesList object of length 1:</span></a>
<a class="sourceLine" id="cb31-3" data-line-number="3"><span class="co">#&gt; $A </span></a>
<a class="sourceLine" id="cb31-4" data-line-number="4"><span class="co">#&gt; GTuples object with 5 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb31-5" data-line-number="5"><span class="co">#&gt;       seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb31-6" data-line-number="6"><span class="co">#&gt;   [1]     chr1    1    2    3      - |     1                 1</span></a>
<a class="sourceLine" id="cb31-7" data-line-number="7"><span class="co">#&gt;   [2]     chr2    2    3    4      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb31-8" data-line-number="8"><span class="co">#&gt;   [3]     chr2    3    4    5      + |     3 0.777777777777778</span></a>
<a class="sourceLine" id="cb31-9" data-line-number="9"><span class="co">#&gt;   [4]     chr2    4    5    6      * |     4 0.666666666666667</span></a>
<a class="sourceLine" id="cb31-10" data-line-number="10"><span class="co">#&gt;   [5]     chr1    5    6    7      * |     5 0.555555555555556</span></a>
<a class="sourceLine" id="cb31-11" data-line-number="11"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb31-12" data-line-number="12"><span class="co">#&gt; -------</span></a>
<a class="sourceLine" id="cb31-13" data-line-number="13"><span class="co">#&gt; seqinfo: 3 sequences from mock1 genome</span></a>
<a class="sourceLine" id="cb31-14" data-line-number="14">gtl3[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb31-15" data-line-number="15"><span class="co">#&gt; GTuples object with 5 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb31-16" data-line-number="16"><span class="co">#&gt;       seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb31-17" data-line-number="17"><span class="co">#&gt;   [1]     chr1    1    2    3      - |     1                 1</span></a>
<a class="sourceLine" id="cb31-18" data-line-number="18"><span class="co">#&gt;   [2]     chr2    2    3    4      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb31-19" data-line-number="19"><span class="co">#&gt;   [3]     chr2    3    4    5      + |     3 0.777777777777778</span></a>
<a class="sourceLine" id="cb31-20" data-line-number="20"><span class="co">#&gt;   [4]     chr2    4    5    6      * |     4 0.666666666666667</span></a>
<a class="sourceLine" id="cb31-21" data-line-number="21"><span class="co">#&gt;   [5]     chr1    5    6    7      * |     5 0.555555555555556</span></a>
<a class="sourceLine" id="cb31-22" data-line-number="22"><span class="co">#&gt;   ---</span></a>
<a class="sourceLine" id="cb31-23" data-line-number="23"><span class="co">#&gt;   seqinfo: 3 sequences from mock1 genome</span></a>
<a class="sourceLine" id="cb31-24" data-line-number="24">gtl3[<span class="st">"A"</span>]</a>
<a class="sourceLine" id="cb31-25" data-line-number="25"><span class="co">#&gt; GTuplesList object of length 1:</span></a>
<a class="sourceLine" id="cb31-26" data-line-number="26"><span class="co">#&gt; $A </span></a>
<a class="sourceLine" id="cb31-27" data-line-number="27"><span class="co">#&gt; GTuples object with 5 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb31-28" data-line-number="28"><span class="co">#&gt;       seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb31-29" data-line-number="29"><span class="co">#&gt;   [1]     chr1    1    2    3      - |     1                 1</span></a>
<a class="sourceLine" id="cb31-30" data-line-number="30"><span class="co">#&gt;   [2]     chr2    2    3    4      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb31-31" data-line-number="31"><span class="co">#&gt;   [3]     chr2    3    4    5      + |     3 0.777777777777778</span></a>
<a class="sourceLine" id="cb31-32" data-line-number="32"><span class="co">#&gt;   [4]     chr2    4    5    6      * |     4 0.666666666666667</span></a>
<a class="sourceLine" id="cb31-33" data-line-number="33"><span class="co">#&gt;   [5]     chr1    5    6    7      * |     5 0.555555555555556</span></a>
<a class="sourceLine" id="cb31-34" data-line-number="34"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb31-35" data-line-number="35"><span class="co">#&gt; -------</span></a>
<a class="sourceLine" id="cb31-36" data-line-number="36"><span class="co">#&gt; seqinfo: 3 sequences from mock1 genome</span></a>
<a class="sourceLine" id="cb31-37" data-line-number="37">gtl3<span class="op">$</span>B</a>
<a class="sourceLine" id="cb31-38" data-line-number="38"><span class="co">#&gt; GTuples object with 5 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb31-39" data-line-number="39"><span class="co">#&gt;       seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb31-40" data-line-number="40"><span class="co">#&gt;   [1]     chr1    6    7    8      + |     6 0.444444444444444</span></a>
<a class="sourceLine" id="cb31-41" data-line-number="41"><span class="co">#&gt;   [2]     chr3    7    8    9      + |     7 0.333333333333333</span></a>
<a class="sourceLine" id="cb31-42" data-line-number="42"><span class="co">#&gt;   [3]     chr3    8    9   10      + |     8 0.222222222222222</span></a>
<a class="sourceLine" id="cb31-43" data-line-number="43"><span class="co">#&gt;   [4]     chr3    9   10   11      - |     9 0.111111111111111</span></a>
<a class="sourceLine" id="cb31-44" data-line-number="44"><span class="co">#&gt;   [5]     chr3   10   11   12      - |    10                 0</span></a>
<a class="sourceLine" id="cb31-45" data-line-number="45"><span class="co">#&gt;   ---</span></a>
<a class="sourceLine" id="cb31-46" data-line-number="46"><span class="co">#&gt;   seqinfo: 3 sequences from mock1 genome</span></a></code></pre></div>
<p>When subsetting a <em>GTuplesList</em>, you can also pass in a second parameter (as with a <em>GTuples</em> object) to again specify which of the metadata columns you wish to select:</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" data-line-number="1">gtl3[<span class="dv">1</span>, <span class="st">"score"</span>]</a>
<a class="sourceLine" id="cb32-2" data-line-number="2"><span class="co">#&gt; GTuplesList object of length 1:</span></a>
<a class="sourceLine" id="cb32-3" data-line-number="3"><span class="co">#&gt; $A </span></a>
<a class="sourceLine" id="cb32-4" data-line-number="4"><span class="co">#&gt; GTuples object with 5 x 3-tuples and 1 metadata column:</span></a>
<a class="sourceLine" id="cb32-5" data-line-number="5"><span class="co">#&gt;       seqnames pos1 pos2 pos3 strand | score</span></a>
<a class="sourceLine" id="cb32-6" data-line-number="6"><span class="co">#&gt;   [1]     chr1    1    2    3      - |     1</span></a>
<a class="sourceLine" id="cb32-7" data-line-number="7"><span class="co">#&gt;   [2]     chr2    2    3    4      + |     2</span></a>
<a class="sourceLine" id="cb32-8" data-line-number="8"><span class="co">#&gt;   [3]     chr2    3    4    5      + |     3</span></a>
<a class="sourceLine" id="cb32-9" data-line-number="9"><span class="co">#&gt;   [4]     chr2    4    5    6      * |     4</span></a>
<a class="sourceLine" id="cb32-10" data-line-number="10"><span class="co">#&gt;   [5]     chr1    5    6    7      * |     5</span></a>
<a class="sourceLine" id="cb32-11" data-line-number="11"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb32-12" data-line-number="12"><span class="co">#&gt; -------</span></a>
<a class="sourceLine" id="cb32-13" data-line-number="13"><span class="co">#&gt; seqinfo: 3 sequences from mock1 genome</span></a>
<a class="sourceLine" id="cb32-14" data-line-number="14">gtl3[<span class="st">"B"</span>, <span class="st">"GC"</span>]</a>
<a class="sourceLine" id="cb32-15" data-line-number="15"><span class="co">#&gt; GTuplesList object of length 1:</span></a>
<a class="sourceLine" id="cb32-16" data-line-number="16"><span class="co">#&gt; $B </span></a>
<a class="sourceLine" id="cb32-17" data-line-number="17"><span class="co">#&gt; GTuples object with 5 x 3-tuples and 1 metadata column:</span></a>
<a class="sourceLine" id="cb32-18" data-line-number="18"><span class="co">#&gt;       seqnames pos1 pos2 pos3 strand |                GC</span></a>
<a class="sourceLine" id="cb32-19" data-line-number="19"><span class="co">#&gt;   [1]     chr1    6    7    8      + | 0.444444444444444</span></a>
<a class="sourceLine" id="cb32-20" data-line-number="20"><span class="co">#&gt;   [2]     chr3    7    8    9      + | 0.333333333333333</span></a>
<a class="sourceLine" id="cb32-21" data-line-number="21"><span class="co">#&gt;   [3]     chr3    8    9   10      + | 0.222222222222222</span></a>
<a class="sourceLine" id="cb32-22" data-line-number="22"><span class="co">#&gt;   [4]     chr3    9   10   11      - | 0.111111111111111</span></a>
<a class="sourceLine" id="cb32-23" data-line-number="23"><span class="co">#&gt;   [5]     chr3   10   11   12      - |                 0</span></a>
<a class="sourceLine" id="cb32-24" data-line-number="24"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb32-25" data-line-number="25"><span class="co">#&gt; -------</span></a>
<a class="sourceLine" id="cb32-26" data-line-number="26"><span class="co">#&gt; seqinfo: 3 sequences from mock1 genome</span></a></code></pre></div>
<p>The <code>head</code>, <code>tail</code>, <code>rep</code>, <code>rev</code>, and <code>window</code> methods all behave as you would expect them to for a <em>List</em> object. For example, the elements referred to by <code>window</code> are now list elements instead of <em>GTuples</em> elements:</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb33-1" data-line-number="1"><span class="kw">rep</span>(gtl3[[<span class="dv">1</span>]], <span class="dt">times =</span> <span class="dv">3</span>)</a>
<a class="sourceLine" id="cb33-2" data-line-number="2"><span class="co">#&gt; GTuples object with 15 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb33-3" data-line-number="3"><span class="co">#&gt;        seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb33-4" data-line-number="4"><span class="co">#&gt;    [1]     chr1    1    2    3      - |     1                 1</span></a>
<a class="sourceLine" id="cb33-5" data-line-number="5"><span class="co">#&gt;    [2]     chr2    2    3    4      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb33-6" data-line-number="6"><span class="co">#&gt;    [3]     chr2    3    4    5      + |     3 0.777777777777778</span></a>
<a class="sourceLine" id="cb33-7" data-line-number="7"><span class="co">#&gt;    [4]     chr2    4    5    6      * |     4 0.666666666666667</span></a>
<a class="sourceLine" id="cb33-8" data-line-number="8"><span class="co">#&gt;    [5]     chr1    5    6    7      * |     5 0.555555555555556</span></a>
<a class="sourceLine" id="cb33-9" data-line-number="9"><span class="co">#&gt;    ...      ...  ...  ...  ...    ... .   ...               ...</span></a>
<a class="sourceLine" id="cb33-10" data-line-number="10"><span class="co">#&gt;   [11]     chr1    1    2    3      - |     1                 1</span></a>
<a class="sourceLine" id="cb33-11" data-line-number="11"><span class="co">#&gt;   [12]     chr2    2    3    4      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb33-12" data-line-number="12"><span class="co">#&gt;   [13]     chr2    3    4    5      + |     3 0.777777777777778</span></a>
<a class="sourceLine" id="cb33-13" data-line-number="13"><span class="co">#&gt;   [14]     chr2    4    5    6      * |     4 0.666666666666667</span></a>
<a class="sourceLine" id="cb33-14" data-line-number="14"><span class="co">#&gt;   [15]     chr1    5    6    7      * |     5 0.555555555555556</span></a>
<a class="sourceLine" id="cb33-15" data-line-number="15"><span class="co">#&gt;   ---</span></a>
<a class="sourceLine" id="cb33-16" data-line-number="16"><span class="co">#&gt;   seqinfo: 3 sequences from mock1 genome</span></a>
<a class="sourceLine" id="cb33-17" data-line-number="17"><span class="kw">rev</span>(gtl3)</a>
<a class="sourceLine" id="cb33-18" data-line-number="18"><span class="co">#&gt; GTuplesList object of length 2:</span></a>
<a class="sourceLine" id="cb33-19" data-line-number="19"><span class="co">#&gt; $B </span></a>
<a class="sourceLine" id="cb33-20" data-line-number="20"><span class="co">#&gt; GTuples object with 5 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb33-21" data-line-number="21"><span class="co">#&gt;       seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb33-22" data-line-number="22"><span class="co">#&gt;   [1]     chr1    6    7    8      + |     6 0.444444444444444</span></a>
<a class="sourceLine" id="cb33-23" data-line-number="23"><span class="co">#&gt;   [2]     chr3    7    8    9      + |     7 0.333333333333333</span></a>
<a class="sourceLine" id="cb33-24" data-line-number="24"><span class="co">#&gt;   [3]     chr3    8    9   10      + |     8 0.222222222222222</span></a>
<a class="sourceLine" id="cb33-25" data-line-number="25"><span class="co">#&gt;   [4]     chr3    9   10   11      - |     9 0.111111111111111</span></a>
<a class="sourceLine" id="cb33-26" data-line-number="26"><span class="co">#&gt;   [5]     chr3   10   11   12      - |    10                 0</span></a>
<a class="sourceLine" id="cb33-27" data-line-number="27"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb33-28" data-line-number="28"><span class="co">#&gt; $A </span></a>
<a class="sourceLine" id="cb33-29" data-line-number="29"><span class="co">#&gt; GTuples object with 5 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb33-30" data-line-number="30"><span class="co">#&gt;       seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb33-31" data-line-number="31"><span class="co">#&gt;   [1]     chr1    1    2    3      - |     1                 1</span></a>
<a class="sourceLine" id="cb33-32" data-line-number="32"><span class="co">#&gt;   [2]     chr2    2    3    4      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb33-33" data-line-number="33"><span class="co">#&gt;   [3]     chr2    3    4    5      + |     3 0.777777777777778</span></a>
<a class="sourceLine" id="cb33-34" data-line-number="34"><span class="co">#&gt;   [4]     chr2    4    5    6      * |     4 0.666666666666667</span></a>
<a class="sourceLine" id="cb33-35" data-line-number="35"><span class="co">#&gt;   [5]     chr1    5    6    7      * |     5 0.555555555555556</span></a>
<a class="sourceLine" id="cb33-36" data-line-number="36"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb33-37" data-line-number="37"><span class="co">#&gt; -------</span></a>
<a class="sourceLine" id="cb33-38" data-line-number="38"><span class="co">#&gt; seqinfo: 3 sequences from mock1 genome</span></a>
<a class="sourceLine" id="cb33-39" data-line-number="39"><span class="kw">head</span>(gtl3, <span class="dt">n =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb33-40" data-line-number="40"><span class="co">#&gt; GTuplesList object of length 1:</span></a>
<a class="sourceLine" id="cb33-41" data-line-number="41"><span class="co">#&gt; $A </span></a>
<a class="sourceLine" id="cb33-42" data-line-number="42"><span class="co">#&gt; GTuples object with 5 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb33-43" data-line-number="43"><span class="co">#&gt;       seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb33-44" data-line-number="44"><span class="co">#&gt;   [1]     chr1    1    2    3      - |     1                 1</span></a>
<a class="sourceLine" id="cb33-45" data-line-number="45"><span class="co">#&gt;   [2]     chr2    2    3    4      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb33-46" data-line-number="46"><span class="co">#&gt;   [3]     chr2    3    4    5      + |     3 0.777777777777778</span></a>
<a class="sourceLine" id="cb33-47" data-line-number="47"><span class="co">#&gt;   [4]     chr2    4    5    6      * |     4 0.666666666666667</span></a>
<a class="sourceLine" id="cb33-48" data-line-number="48"><span class="co">#&gt;   [5]     chr1    5    6    7      * |     5 0.555555555555556</span></a>
<a class="sourceLine" id="cb33-49" data-line-number="49"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb33-50" data-line-number="50"><span class="co">#&gt; -------</span></a>
<a class="sourceLine" id="cb33-51" data-line-number="51"><span class="co">#&gt; seqinfo: 3 sequences from mock1 genome</span></a>
<a class="sourceLine" id="cb33-52" data-line-number="52"><span class="kw">tail</span>(gtl3, <span class="dt">n =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb33-53" data-line-number="53"><span class="co">#&gt; GTuplesList object of length 1:</span></a>
<a class="sourceLine" id="cb33-54" data-line-number="54"><span class="co">#&gt; $B </span></a>
<a class="sourceLine" id="cb33-55" data-line-number="55"><span class="co">#&gt; GTuples object with 5 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb33-56" data-line-number="56"><span class="co">#&gt;       seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb33-57" data-line-number="57"><span class="co">#&gt;   [1]     chr1    6    7    8      + |     6 0.444444444444444</span></a>
<a class="sourceLine" id="cb33-58" data-line-number="58"><span class="co">#&gt;   [2]     chr3    7    8    9      + |     7 0.333333333333333</span></a>
<a class="sourceLine" id="cb33-59" data-line-number="59"><span class="co">#&gt;   [3]     chr3    8    9   10      + |     8 0.222222222222222</span></a>
<a class="sourceLine" id="cb33-60" data-line-number="60"><span class="co">#&gt;   [4]     chr3    9   10   11      - |     9 0.111111111111111</span></a>
<a class="sourceLine" id="cb33-61" data-line-number="61"><span class="co">#&gt;   [5]     chr3   10   11   12      - |    10                 0</span></a>
<a class="sourceLine" id="cb33-62" data-line-number="62"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb33-63" data-line-number="63"><span class="co">#&gt; -------</span></a>
<a class="sourceLine" id="cb33-64" data-line-number="64"><span class="co">#&gt; seqinfo: 3 sequences from mock1 genome</span></a>
<a class="sourceLine" id="cb33-65" data-line-number="65"><span class="kw">window</span>(gtl3, <span class="dt">start =</span> <span class="dv">1</span>, <span class="dt">end =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb33-66" data-line-number="66"><span class="co">#&gt; GTuplesList object of length 1:</span></a>
<a class="sourceLine" id="cb33-67" data-line-number="67"><span class="co">#&gt; $A </span></a>
<a class="sourceLine" id="cb33-68" data-line-number="68"><span class="co">#&gt; GTuples object with 5 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb33-69" data-line-number="69"><span class="co">#&gt;       seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb33-70" data-line-number="70"><span class="co">#&gt;   [1]     chr1    1    2    3      - |     1                 1</span></a>
<a class="sourceLine" id="cb33-71" data-line-number="71"><span class="co">#&gt;   [2]     chr2    2    3    4      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb33-72" data-line-number="72"><span class="co">#&gt;   [3]     chr2    3    4    5      + |     3 0.777777777777778</span></a>
<a class="sourceLine" id="cb33-73" data-line-number="73"><span class="co">#&gt;   [4]     chr2    4    5    6      * |     4 0.666666666666667</span></a>
<a class="sourceLine" id="cb33-74" data-line-number="74"><span class="co">#&gt;   [5]     chr1    5    6    7      * |     5 0.555555555555556</span></a>
<a class="sourceLine" id="cb33-75" data-line-number="75"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb33-76" data-line-number="76"><span class="co">#&gt; -------</span></a>
<a class="sourceLine" id="cb33-77" data-line-number="77"><span class="co">#&gt; seqinfo: 3 sequences from mock1 genome</span></a></code></pre></div>
</div>
<div id="basic-tuple-operations-for-gtupleslist-objects" class="section level3">
<h3 class="hasAnchor">
<a href="#basic-tuple-operations-for-gtupleslist-objects" class="anchor"></a>Basic tuple operations for <em>GTuplesList</em> objects</h3>
<p>Basic tuple characteristics of <em>GTuplesList</em> objects can be extracted using the <code>start</code>, <code>end</code>, and <code>tuples</code> methods. These are very similar to those available for <em>GTuples</em> objects, except that they typically return a list since the input is now essentially a list of <em>GTuples</em> objects.</p>
<p><strong>WARNING: While the <code>width</code> method is well-defined, namely it returns an <em>IntegerList</em> of <span class="math inline">\(pos_{m} - pos_{1} + 1\)</span>, this is not generally what is desired or required. Instead, please see the <code>IPD</code> method that is discussed later.</strong></p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb34-1" data-line-number="1"><span class="kw">start</span>(gtl3)</a>
<a class="sourceLine" id="cb34-2" data-line-number="2"><span class="co">#&gt; [1] 1 1</span></a>
<a class="sourceLine" id="cb34-3" data-line-number="3"><span class="kw">end</span>(gtl3)</a>
<a class="sourceLine" id="cb34-4" data-line-number="4"><span class="co">#&gt; [1] 2 1</span></a>
<a class="sourceLine" id="cb34-5" data-line-number="5"><span class="kw"><a href="../reference/GTuples-class.html">tuples</a></span>(gtl3)</a>
<a class="sourceLine" id="cb34-6" data-line-number="6"><span class="co">#&gt; List of length 2</span></a>
<a class="sourceLine" id="cb34-7" data-line-number="7"><span class="co">#&gt; names(2): A B</span></a></code></pre></div>
<div id="intra-tuple-operations-1" class="section level4">
<h4 class="hasAnchor">
<a href="#intra-tuple-operations-1" class="anchor"></a>Intra-tuple operations</h4>
<p>Most of the intra-range methods defined for <em>GenomicRanges::GRangesList</em> objects are not currently defined via extension for <em>GTuples</em> objects due to the differences between ranges and tuples. Those not currently defined, and which return an error message, are:</p>
<ul>
<li><code>flank</code></li>
<li><code>promoters</code></li>
<li><code>resize</code></li>
<li><code>restrict</code></li>
</ul>
<p>I am happy to add these methods if appropriate, so please contact me if you have suggestions for good definitions.</p>
<p>The <code>shift</code> method is well-defined:</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb35-1" data-line-number="1"><span class="kw"><a href="../reference/intra-tuple-methods.html">shift</a></span>(gtl3, <span class="dv">500</span>)</a>
<a class="sourceLine" id="cb35-2" data-line-number="2"><span class="co">#&gt; GTuplesList object of length 2:</span></a>
<a class="sourceLine" id="cb35-3" data-line-number="3"><span class="co">#&gt; $A </span></a>
<a class="sourceLine" id="cb35-4" data-line-number="4"><span class="co">#&gt; GTuples object with 5 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb35-5" data-line-number="5"><span class="co">#&gt;       seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb35-6" data-line-number="6"><span class="co">#&gt;   [1]     chr1  501  502  503      - |     1                 1</span></a>
<a class="sourceLine" id="cb35-7" data-line-number="7"><span class="co">#&gt;   [2]     chr2  502  503  504      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb35-8" data-line-number="8"><span class="co">#&gt;   [3]     chr2  503  504  505      + |     3 0.777777777777778</span></a>
<a class="sourceLine" id="cb35-9" data-line-number="9"><span class="co">#&gt;   [4]     chr2  504  505  506      * |     4 0.666666666666667</span></a>
<a class="sourceLine" id="cb35-10" data-line-number="10"><span class="co">#&gt;   [5]     chr1  505  506  507      * |     5 0.555555555555556</span></a>
<a class="sourceLine" id="cb35-11" data-line-number="11"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb35-12" data-line-number="12"><span class="co">#&gt; $B </span></a>
<a class="sourceLine" id="cb35-13" data-line-number="13"><span class="co">#&gt; GTuples object with 5 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb35-14" data-line-number="14"><span class="co">#&gt;       seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb35-15" data-line-number="15"><span class="co">#&gt;   [1]     chr1  506  507  508      + |     6 0.444444444444444</span></a>
<a class="sourceLine" id="cb35-16" data-line-number="16"><span class="co">#&gt;   [2]     chr3  507  508  509      + |     7 0.333333333333333</span></a>
<a class="sourceLine" id="cb35-17" data-line-number="17"><span class="co">#&gt;   [3]     chr3  508  509  510      + |     8 0.222222222222222</span></a>
<a class="sourceLine" id="cb35-18" data-line-number="18"><span class="co">#&gt;   [4]     chr3  509  510  511      - |     9 0.111111111111111</span></a>
<a class="sourceLine" id="cb35-19" data-line-number="19"><span class="co">#&gt;   [5]     chr3  510  511  512      - |    10                 0</span></a>
<a class="sourceLine" id="cb35-20" data-line-number="20"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb35-21" data-line-number="21"><span class="co">#&gt; -------</span></a>
<a class="sourceLine" id="cb35-22" data-line-number="22"><span class="co">#&gt; seqinfo: 3 sequences from mock1 genome</span></a>
<a class="sourceLine" id="cb35-23" data-line-number="23"><span class="kw"><a href="../reference/intra-tuple-methods.html">shift</a></span>(gtl3, <span class="kw">IntegerList</span>(<span class="dt">A =</span> 300L, <span class="dt">B =</span> 500L))</a>
<a class="sourceLine" id="cb35-24" data-line-number="24"><span class="co">#&gt; GTuplesList object of length 2:</span></a>
<a class="sourceLine" id="cb35-25" data-line-number="25"><span class="co">#&gt; $A </span></a>
<a class="sourceLine" id="cb35-26" data-line-number="26"><span class="co">#&gt; GTuples object with 5 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb35-27" data-line-number="27"><span class="co">#&gt;       seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb35-28" data-line-number="28"><span class="co">#&gt;   [1]     chr1  301  302  303      - |     1                 1</span></a>
<a class="sourceLine" id="cb35-29" data-line-number="29"><span class="co">#&gt;   [2]     chr2  302  303  304      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb35-30" data-line-number="30"><span class="co">#&gt;   [3]     chr2  303  304  305      + |     3 0.777777777777778</span></a>
<a class="sourceLine" id="cb35-31" data-line-number="31"><span class="co">#&gt;   [4]     chr2  304  305  306      * |     4 0.666666666666667</span></a>
<a class="sourceLine" id="cb35-32" data-line-number="32"><span class="co">#&gt;   [5]     chr1  305  306  307      * |     5 0.555555555555556</span></a>
<a class="sourceLine" id="cb35-33" data-line-number="33"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb35-34" data-line-number="34"><span class="co">#&gt; $B </span></a>
<a class="sourceLine" id="cb35-35" data-line-number="35"><span class="co">#&gt; GTuples object with 5 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb35-36" data-line-number="36"><span class="co">#&gt;       seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb35-37" data-line-number="37"><span class="co">#&gt;   [1]     chr1  506  507  508      + |     6 0.444444444444444</span></a>
<a class="sourceLine" id="cb35-38" data-line-number="38"><span class="co">#&gt;   [2]     chr3  507  508  509      + |     7 0.333333333333333</span></a>
<a class="sourceLine" id="cb35-39" data-line-number="39"><span class="co">#&gt;   [3]     chr3  508  509  510      + |     8 0.222222222222222</span></a>
<a class="sourceLine" id="cb35-40" data-line-number="40"><span class="co">#&gt;   [4]     chr3  509  510  511      - |     9 0.111111111111111</span></a>
<a class="sourceLine" id="cb35-41" data-line-number="41"><span class="co">#&gt;   [5]     chr3  510  511  512      - |    10                 0</span></a>
<a class="sourceLine" id="cb35-42" data-line-number="42"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb35-43" data-line-number="43"><span class="co">#&gt; -------</span></a>
<a class="sourceLine" id="cb35-44" data-line-number="44"><span class="co">#&gt; seqinfo: 3 sequences from mock1 genome</span></a></code></pre></div>
</div>
<div id="inter-tuple-operations-1" class="section level4">
<h4 class="hasAnchor">
<a href="#inter-tuple-operations-1" class="anchor"></a>Inter-tuple operations</h4>
<p>None of the inter-range methods defined for <em>GenomicRanges::GRangesList</em> objects are currently defined via extension for <em>GTuplesList</em> objects due to the differences between ranges and tuples. Those not currently defined, and which return an error message, are:</p>
<ul>
<li><code>range</code></li>
<li><code>reduce</code></li>
<li><code>disjoin</code></li>
<li><code>isDisjoint</code></li>
</ul>
<p>I am happy to add these methods if appropriate, so please contact me if you have suggestions for good definitions.</p>
</div>
<div id="interval-set-operations-for-gtupleslist-objects" class="section level4">
<h4 class="hasAnchor">
<a href="#interval-set-operations-for-gtupleslist-objects" class="anchor"></a>Interval set operations for <em>GTuplesList</em> objects</h4>
<p>None of the interval set operations defined for <em>GenomicRanges::GRangesList</em> objects are currently defined via extension for <em>GTuplesList</em> objects due to the differences between ranges and tuples. Those not currently defined, and which return an error message, are:</p>
<ul>
<li><code>punion</code></li>
<li><code>pintersect</code></li>
<li><code>psetdiff</code></li>
</ul>
<p>I am happy to add these methods if appropriate, so please contact me if you have suggestions for good definitions.</p>
</div>
</div>
<div id="looping-over-gtupleslist-objects" class="section level3">
<h3 class="hasAnchor">
<a href="#looping-over-gtupleslist-objects" class="anchor"></a>Looping over <em>GTuplesList</em> objects</h3>
<p>Like for <em>GenomicRanges::GRangesList</em> objects, for <em>GTuplesList</em> objects there is a family of apply methods. These include <code>lapply</code>, <code>sapply</code>, <code>mapply</code>, <code>endoapply</code>, <code>mendoapply</code>, <code>Map</code>, and <code>Reduce</code>. The different looping methods defined for <em>GTuplesList</em> objects are useful for returning different kinds of results. The standard <code>lapply</code> and <code>sapply</code> behave according to convention, with the <code>lapply</code> method returning a list and <code>sapply</code> returning a more simplified output:</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb36-1" data-line-number="1"><span class="kw">lapply</span>(gtl3, length)</a>
<a class="sourceLine" id="cb36-2" data-line-number="2"><span class="co">#&gt; $A</span></a>
<a class="sourceLine" id="cb36-3" data-line-number="3"><span class="co">#&gt; [1] 5</span></a>
<a class="sourceLine" id="cb36-4" data-line-number="4"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb36-5" data-line-number="5"><span class="co">#&gt; $B</span></a>
<a class="sourceLine" id="cb36-6" data-line-number="6"><span class="co">#&gt; [1] 5</span></a>
<a class="sourceLine" id="cb36-7" data-line-number="7"><span class="kw">sapply</span>(gtl3, length)</a>
<a class="sourceLine" id="cb36-8" data-line-number="8"><span class="co">#&gt; A B </span></a>
<a class="sourceLine" id="cb36-9" data-line-number="9"><span class="co">#&gt; 5 5</span></a></code></pre></div>
<p>As with <em>GenomicRanges::GRangesList</em> objects, there is also a multivariate version of <code>sapply</code>, called <code>mapply</code>, defined for <em>GTuplesList</em> objects. And, if you don’t want the results simplified, you can call the <code>Map</code> method, which does the same things as <code>mapply</code> but without simplifying the output:</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb37-1" data-line-number="1">gtl3_shift &lt;-<span class="st"> </span><span class="kw"><a href="../reference/intra-tuple-methods.html">shift</a></span>(gtl3, <span class="dv">10</span>)</a>
<a class="sourceLine" id="cb37-2" data-line-number="2"><span class="kw">names</span>(gtl3) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"shiftA"</span>, <span class="st">"shiftB"</span>)</a>
<a class="sourceLine" id="cb37-3" data-line-number="3"><span class="kw">mapply</span>(c, gtl3, gtl3_shift)</a>
<a class="sourceLine" id="cb37-4" data-line-number="4"><span class="co">#&gt; $shiftA</span></a>
<a class="sourceLine" id="cb37-5" data-line-number="5"><span class="co">#&gt; GTuples object with 10 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb37-6" data-line-number="6"><span class="co">#&gt;        seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb37-7" data-line-number="7"><span class="co">#&gt;    [1]     chr1    1    2    3      - |     1                 1</span></a>
<a class="sourceLine" id="cb37-8" data-line-number="8"><span class="co">#&gt;    [2]     chr2    2    3    4      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb37-9" data-line-number="9"><span class="co">#&gt;    [3]     chr2    3    4    5      + |     3 0.777777777777778</span></a>
<a class="sourceLine" id="cb37-10" data-line-number="10"><span class="co">#&gt;    [4]     chr2    4    5    6      * |     4 0.666666666666667</span></a>
<a class="sourceLine" id="cb37-11" data-line-number="11"><span class="co">#&gt;    [5]     chr1    5    6    7      * |     5 0.555555555555556</span></a>
<a class="sourceLine" id="cb37-12" data-line-number="12"><span class="co">#&gt;    [6]     chr1   11   12   13      - |     1                 1</span></a>
<a class="sourceLine" id="cb37-13" data-line-number="13"><span class="co">#&gt;    [7]     chr2   12   13   14      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb37-14" data-line-number="14"><span class="co">#&gt;    [8]     chr2   13   14   15      + |     3 0.777777777777778</span></a>
<a class="sourceLine" id="cb37-15" data-line-number="15"><span class="co">#&gt;    [9]     chr2   14   15   16      * |     4 0.666666666666667</span></a>
<a class="sourceLine" id="cb37-16" data-line-number="16"><span class="co">#&gt;   [10]     chr1   15   16   17      * |     5 0.555555555555556</span></a>
<a class="sourceLine" id="cb37-17" data-line-number="17"><span class="co">#&gt;   ---</span></a>
<a class="sourceLine" id="cb37-18" data-line-number="18"><span class="co">#&gt;   seqinfo: 3 sequences from mock1 genome</span></a>
<a class="sourceLine" id="cb37-19" data-line-number="19"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb37-20" data-line-number="20"><span class="co">#&gt; $shiftB</span></a>
<a class="sourceLine" id="cb37-21" data-line-number="21"><span class="co">#&gt; GTuples object with 10 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb37-22" data-line-number="22"><span class="co">#&gt;        seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb37-23" data-line-number="23"><span class="co">#&gt;    [1]     chr1    6    7    8      + |     6 0.444444444444444</span></a>
<a class="sourceLine" id="cb37-24" data-line-number="24"><span class="co">#&gt;    [2]     chr3    7    8    9      + |     7 0.333333333333333</span></a>
<a class="sourceLine" id="cb37-25" data-line-number="25"><span class="co">#&gt;    [3]     chr3    8    9   10      + |     8 0.222222222222222</span></a>
<a class="sourceLine" id="cb37-26" data-line-number="26"><span class="co">#&gt;    [4]     chr3    9   10   11      - |     9 0.111111111111111</span></a>
<a class="sourceLine" id="cb37-27" data-line-number="27"><span class="co">#&gt;    [5]     chr3   10   11   12      - |    10                 0</span></a>
<a class="sourceLine" id="cb37-28" data-line-number="28"><span class="co">#&gt;    [6]     chr1   16   17   18      + |     6 0.444444444444444</span></a>
<a class="sourceLine" id="cb37-29" data-line-number="29"><span class="co">#&gt;    [7]     chr3   17   18   19      + |     7 0.333333333333333</span></a>
<a class="sourceLine" id="cb37-30" data-line-number="30"><span class="co">#&gt;    [8]     chr3   18   19   20      + |     8 0.222222222222222</span></a>
<a class="sourceLine" id="cb37-31" data-line-number="31"><span class="co">#&gt;    [9]     chr3   19   20   21      - |     9 0.111111111111111</span></a>
<a class="sourceLine" id="cb37-32" data-line-number="32"><span class="co">#&gt;   [10]     chr3   20   21   22      - |    10                 0</span></a>
<a class="sourceLine" id="cb37-33" data-line-number="33"><span class="co">#&gt;   ---</span></a>
<a class="sourceLine" id="cb37-34" data-line-number="34"><span class="co">#&gt;   seqinfo: 3 sequences from mock1 genome</span></a>
<a class="sourceLine" id="cb37-35" data-line-number="35"><span class="kw">Map</span>(c, gtl3, gtl3_shift)</a>
<a class="sourceLine" id="cb37-36" data-line-number="36"><span class="co">#&gt; $shiftA</span></a>
<a class="sourceLine" id="cb37-37" data-line-number="37"><span class="co">#&gt; GTuples object with 10 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb37-38" data-line-number="38"><span class="co">#&gt;        seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb37-39" data-line-number="39"><span class="co">#&gt;    [1]     chr1    1    2    3      - |     1                 1</span></a>
<a class="sourceLine" id="cb37-40" data-line-number="40"><span class="co">#&gt;    [2]     chr2    2    3    4      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb37-41" data-line-number="41"><span class="co">#&gt;    [3]     chr2    3    4    5      + |     3 0.777777777777778</span></a>
<a class="sourceLine" id="cb37-42" data-line-number="42"><span class="co">#&gt;    [4]     chr2    4    5    6      * |     4 0.666666666666667</span></a>
<a class="sourceLine" id="cb37-43" data-line-number="43"><span class="co">#&gt;    [5]     chr1    5    6    7      * |     5 0.555555555555556</span></a>
<a class="sourceLine" id="cb37-44" data-line-number="44"><span class="co">#&gt;    [6]     chr1   11   12   13      - |     1                 1</span></a>
<a class="sourceLine" id="cb37-45" data-line-number="45"><span class="co">#&gt;    [7]     chr2   12   13   14      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb37-46" data-line-number="46"><span class="co">#&gt;    [8]     chr2   13   14   15      + |     3 0.777777777777778</span></a>
<a class="sourceLine" id="cb37-47" data-line-number="47"><span class="co">#&gt;    [9]     chr2   14   15   16      * |     4 0.666666666666667</span></a>
<a class="sourceLine" id="cb37-48" data-line-number="48"><span class="co">#&gt;   [10]     chr1   15   16   17      * |     5 0.555555555555556</span></a>
<a class="sourceLine" id="cb37-49" data-line-number="49"><span class="co">#&gt;   ---</span></a>
<a class="sourceLine" id="cb37-50" data-line-number="50"><span class="co">#&gt;   seqinfo: 3 sequences from mock1 genome</span></a>
<a class="sourceLine" id="cb37-51" data-line-number="51"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb37-52" data-line-number="52"><span class="co">#&gt; $shiftB</span></a>
<a class="sourceLine" id="cb37-53" data-line-number="53"><span class="co">#&gt; GTuples object with 10 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb37-54" data-line-number="54"><span class="co">#&gt;        seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb37-55" data-line-number="55"><span class="co">#&gt;    [1]     chr1    6    7    8      + |     6 0.444444444444444</span></a>
<a class="sourceLine" id="cb37-56" data-line-number="56"><span class="co">#&gt;    [2]     chr3    7    8    9      + |     7 0.333333333333333</span></a>
<a class="sourceLine" id="cb37-57" data-line-number="57"><span class="co">#&gt;    [3]     chr3    8    9   10      + |     8 0.222222222222222</span></a>
<a class="sourceLine" id="cb37-58" data-line-number="58"><span class="co">#&gt;    [4]     chr3    9   10   11      - |     9 0.111111111111111</span></a>
<a class="sourceLine" id="cb37-59" data-line-number="59"><span class="co">#&gt;    [5]     chr3   10   11   12      - |    10                 0</span></a>
<a class="sourceLine" id="cb37-60" data-line-number="60"><span class="co">#&gt;    [6]     chr1   16   17   18      + |     6 0.444444444444444</span></a>
<a class="sourceLine" id="cb37-61" data-line-number="61"><span class="co">#&gt;    [7]     chr3   17   18   19      + |     7 0.333333333333333</span></a>
<a class="sourceLine" id="cb37-62" data-line-number="62"><span class="co">#&gt;    [8]     chr3   18   19   20      + |     8 0.222222222222222</span></a>
<a class="sourceLine" id="cb37-63" data-line-number="63"><span class="co">#&gt;    [9]     chr3   19   20   21      - |     9 0.111111111111111</span></a>
<a class="sourceLine" id="cb37-64" data-line-number="64"><span class="co">#&gt;   [10]     chr3   20   21   22      - |    10                 0</span></a>
<a class="sourceLine" id="cb37-65" data-line-number="65"><span class="co">#&gt;   ---</span></a>
<a class="sourceLine" id="cb37-66" data-line-number="66"><span class="co">#&gt;   seqinfo: 3 sequences from mock1 genome</span></a></code></pre></div>
<p>The <code>endoapply</code> method will return the results as a <em>GTuplesList</em> object rather than as a list:</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb38-1" data-line-number="1"><span class="kw">endoapply</span>(gtl3, rev)</a>
<a class="sourceLine" id="cb38-2" data-line-number="2"><span class="co">#&gt; GTuplesList object of length 2:</span></a>
<a class="sourceLine" id="cb38-3" data-line-number="3"><span class="co">#&gt; $shiftA </span></a>
<a class="sourceLine" id="cb38-4" data-line-number="4"><span class="co">#&gt; GTuples object with 5 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb38-5" data-line-number="5"><span class="co">#&gt;       seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb38-6" data-line-number="6"><span class="co">#&gt;   [1]     chr1    5    6    7      * |     5 0.555555555555556</span></a>
<a class="sourceLine" id="cb38-7" data-line-number="7"><span class="co">#&gt;   [2]     chr2    4    5    6      * |     4 0.666666666666667</span></a>
<a class="sourceLine" id="cb38-8" data-line-number="8"><span class="co">#&gt;   [3]     chr2    3    4    5      + |     3 0.777777777777778</span></a>
<a class="sourceLine" id="cb38-9" data-line-number="9"><span class="co">#&gt;   [4]     chr2    2    3    4      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb38-10" data-line-number="10"><span class="co">#&gt;   [5]     chr1    1    2    3      - |     1                 1</span></a>
<a class="sourceLine" id="cb38-11" data-line-number="11"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb38-12" data-line-number="12"><span class="co">#&gt; $shiftB </span></a>
<a class="sourceLine" id="cb38-13" data-line-number="13"><span class="co">#&gt; GTuples object with 5 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb38-14" data-line-number="14"><span class="co">#&gt;       seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb38-15" data-line-number="15"><span class="co">#&gt;   [1]     chr3   10   11   12      - |    10                 0</span></a>
<a class="sourceLine" id="cb38-16" data-line-number="16"><span class="co">#&gt;   [2]     chr3    9   10   11      - |     9 0.111111111111111</span></a>
<a class="sourceLine" id="cb38-17" data-line-number="17"><span class="co">#&gt;   [3]     chr3    8    9   10      + |     8 0.222222222222222</span></a>
<a class="sourceLine" id="cb38-18" data-line-number="18"><span class="co">#&gt;   [4]     chr3    7    8    9      + |     7 0.333333333333333</span></a>
<a class="sourceLine" id="cb38-19" data-line-number="19"><span class="co">#&gt;   [5]     chr1    6    7    8      + |     6 0.444444444444444</span></a>
<a class="sourceLine" id="cb38-20" data-line-number="20"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb38-21" data-line-number="21"><span class="co">#&gt; -------</span></a>
<a class="sourceLine" id="cb38-22" data-line-number="22"><span class="co">#&gt; seqinfo: 3 sequences from mock1 genome</span></a></code></pre></div>
<p>There is also a multivariate version of the <code>endoapply</code> method in the form of the <code>mendoapply</code> method:</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb39-1" data-line-number="1"><span class="kw">mendoapply</span>(c, gtl3, gtl3_shift)</a>
<a class="sourceLine" id="cb39-2" data-line-number="2"><span class="co">#&gt; GTuplesList object of length 2:</span></a>
<a class="sourceLine" id="cb39-3" data-line-number="3"><span class="co">#&gt; $shiftA </span></a>
<a class="sourceLine" id="cb39-4" data-line-number="4"><span class="co">#&gt; GTuples object with 10 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb39-5" data-line-number="5"><span class="co">#&gt;        seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb39-6" data-line-number="6"><span class="co">#&gt;    [1]     chr1    1    2    3      - |     1                 1</span></a>
<a class="sourceLine" id="cb39-7" data-line-number="7"><span class="co">#&gt;    [2]     chr2    2    3    4      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb39-8" data-line-number="8"><span class="co">#&gt;    [3]     chr2    3    4    5      + |     3 0.777777777777778</span></a>
<a class="sourceLine" id="cb39-9" data-line-number="9"><span class="co">#&gt;    [4]     chr2    4    5    6      * |     4 0.666666666666667</span></a>
<a class="sourceLine" id="cb39-10" data-line-number="10"><span class="co">#&gt;    [5]     chr1    5    6    7      * |     5 0.555555555555556</span></a>
<a class="sourceLine" id="cb39-11" data-line-number="11"><span class="co">#&gt;    [6]     chr1   11   12   13      - |     1                 1</span></a>
<a class="sourceLine" id="cb39-12" data-line-number="12"><span class="co">#&gt;    [7]     chr2   12   13   14      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb39-13" data-line-number="13"><span class="co">#&gt;    [8]     chr2   13   14   15      + |     3 0.777777777777778</span></a>
<a class="sourceLine" id="cb39-14" data-line-number="14"><span class="co">#&gt;    [9]     chr2   14   15   16      * |     4 0.666666666666667</span></a>
<a class="sourceLine" id="cb39-15" data-line-number="15"><span class="co">#&gt;   [10]     chr1   15   16   17      * |     5 0.555555555555556</span></a>
<a class="sourceLine" id="cb39-16" data-line-number="16"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb39-17" data-line-number="17"><span class="co">#&gt; $shiftB </span></a>
<a class="sourceLine" id="cb39-18" data-line-number="18"><span class="co">#&gt; GTuples object with 10 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb39-19" data-line-number="19"><span class="co">#&gt;        seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb39-20" data-line-number="20"><span class="co">#&gt;    [1]     chr1    6    7    8      + |     6 0.444444444444444</span></a>
<a class="sourceLine" id="cb39-21" data-line-number="21"><span class="co">#&gt;    [2]     chr3    7    8    9      + |     7 0.333333333333333</span></a>
<a class="sourceLine" id="cb39-22" data-line-number="22"><span class="co">#&gt;    [3]     chr3    8    9   10      + |     8 0.222222222222222</span></a>
<a class="sourceLine" id="cb39-23" data-line-number="23"><span class="co">#&gt;    [4]     chr3    9   10   11      - |     9 0.111111111111111</span></a>
<a class="sourceLine" id="cb39-24" data-line-number="24"><span class="co">#&gt;    [5]     chr3   10   11   12      - |    10                 0</span></a>
<a class="sourceLine" id="cb39-25" data-line-number="25"><span class="co">#&gt;    [6]     chr1   16   17   18      + |     6 0.444444444444444</span></a>
<a class="sourceLine" id="cb39-26" data-line-number="26"><span class="co">#&gt;    [7]     chr3   17   18   19      + |     7 0.333333333333333</span></a>
<a class="sourceLine" id="cb39-27" data-line-number="27"><span class="co">#&gt;    [8]     chr3   18   19   20      + |     8 0.222222222222222</span></a>
<a class="sourceLine" id="cb39-28" data-line-number="28"><span class="co">#&gt;    [9]     chr3   19   20   21      - |     9 0.111111111111111</span></a>
<a class="sourceLine" id="cb39-29" data-line-number="29"><span class="co">#&gt;   [10]     chr3   20   21   22      - |    10                 0</span></a>
<a class="sourceLine" id="cb39-30" data-line-number="30"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb39-31" data-line-number="31"><span class="co">#&gt; -------</span></a>
<a class="sourceLine" id="cb39-32" data-line-number="32"><span class="co">#&gt; seqinfo: 3 sequences from mock1 genome</span></a></code></pre></div>
<p>Finally, the <code>Reduce</code> method will allow the <em>GTuples</em> objects to be collapsed across the whole of the <em>GTuplesList</em> object:</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb40-1" data-line-number="1"><span class="kw">Reduce</span>(c, gtl3)</a>
<a class="sourceLine" id="cb40-2" data-line-number="2"><span class="co">#&gt; GTuples object with 10 x 3-tuples and 2 metadata columns:</span></a>
<a class="sourceLine" id="cb40-3" data-line-number="3"><span class="co">#&gt;        seqnames pos1 pos2 pos3 strand | score                GC</span></a>
<a class="sourceLine" id="cb40-4" data-line-number="4"><span class="co">#&gt;    [1]     chr1    1    2    3      - |     1                 1</span></a>
<a class="sourceLine" id="cb40-5" data-line-number="5"><span class="co">#&gt;    [2]     chr2    2    3    4      + |     2 0.888888888888889</span></a>
<a class="sourceLine" id="cb40-6" data-line-number="6"><span class="co">#&gt;    [3]     chr2    3    4    5      + |     3 0.777777777777778</span></a>
<a class="sourceLine" id="cb40-7" data-line-number="7"><span class="co">#&gt;    [4]     chr2    4    5    6      * |     4 0.666666666666667</span></a>
<a class="sourceLine" id="cb40-8" data-line-number="8"><span class="co">#&gt;    [5]     chr1    5    6    7      * |     5 0.555555555555556</span></a>
<a class="sourceLine" id="cb40-9" data-line-number="9"><span class="co">#&gt;    [6]     chr1    6    7    8      + |     6 0.444444444444444</span></a>
<a class="sourceLine" id="cb40-10" data-line-number="10"><span class="co">#&gt;    [7]     chr3    7    8    9      + |     7 0.333333333333333</span></a>
<a class="sourceLine" id="cb40-11" data-line-number="11"><span class="co">#&gt;    [8]     chr3    8    9   10      + |     8 0.222222222222222</span></a>
<a class="sourceLine" id="cb40-12" data-line-number="12"><span class="co">#&gt;    [9]     chr3    9   10   11      - |     9 0.111111111111111</span></a>
<a class="sourceLine" id="cb40-13" data-line-number="13"><span class="co">#&gt;   [10]     chr3   10   11   12      - |    10                 0</span></a>
<a class="sourceLine" id="cb40-14" data-line-number="14"><span class="co">#&gt;   ---</span></a>
<a class="sourceLine" id="cb40-15" data-line-number="15"><span class="co">#&gt;   seqinfo: 3 sequences from mock1 genome</span></a></code></pre></div>
</div>
<div id="additional-methods-unique-to-gtupleslist" class="section level3">
<h3 class="hasAnchor">
<a href="#additional-methods-unique-to-gtupleslist" class="anchor"></a>Additional methods unique to <em>GTuplesList</em>
</h3>
<p>Like <em>GTuples</em>, <em>GTuplesList</em> have a few specifically defined methods that do not exist for <em>GenomicRanges::GRangesList</em>. These are <code>tuples</code>, <code>size</code> and <code>IPD</code>. These are identical to the methods for <em>GTuples</em>, except that they typically return a list since the input is now essentially a <em>List</em> of <em>GTuples</em> objects.</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb41-1" data-line-number="1"><span class="kw"><a href="../reference/GTuples-class.html">tuples</a></span>(gtl3)</a>
<a class="sourceLine" id="cb41-2" data-line-number="2"><span class="co">#&gt; List of length 2</span></a>
<a class="sourceLine" id="cb41-3" data-line-number="3"><span class="co">#&gt; names(2): shiftA shiftB</span></a>
<a class="sourceLine" id="cb41-4" data-line-number="4"><span class="kw"><a href="../reference/GTuples-class.html">tuples</a></span>(gtl3)[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb41-5" data-line-number="5"><span class="co">#&gt;      pos1 pos2 pos3</span></a>
<a class="sourceLine" id="cb41-6" data-line-number="6"><span class="co">#&gt; [1,]    1    2    3</span></a>
<a class="sourceLine" id="cb41-7" data-line-number="7"><span class="co">#&gt; [2,]    2    3    4</span></a>
<a class="sourceLine" id="cb41-8" data-line-number="8"><span class="co">#&gt; [3,]    3    4    5</span></a>
<a class="sourceLine" id="cb41-9" data-line-number="9"><span class="co">#&gt; [4,]    4    5    6</span></a>
<a class="sourceLine" id="cb41-10" data-line-number="10"><span class="co">#&gt; [5,]    5    6    7</span></a>
<a class="sourceLine" id="cb41-11" data-line-number="11"><span class="kw"><a href="../reference/GTuples-class.html">size</a></span>(gtl3)</a>
<a class="sourceLine" id="cb41-12" data-line-number="12"><span class="co">#&gt; [1] 3</span></a>
<a class="sourceLine" id="cb41-13" data-line-number="13"><span class="kw"><a href="../reference/GTuples-class.html">IPD</a></span>(gtl3)</a>
<a class="sourceLine" id="cb41-14" data-line-number="14"><span class="co">#&gt; List of length 2</span></a>
<a class="sourceLine" id="cb41-15" data-line-number="15"><span class="co">#&gt; names(2): shiftA shiftB</span></a>
<a class="sourceLine" id="cb41-16" data-line-number="16"><span class="kw"><a href="../reference/GTuples-class.html">IPD</a></span>(gtl3)[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb41-17" data-line-number="17"><span class="co">#&gt;      [,1] [,2]</span></a>
<a class="sourceLine" id="cb41-18" data-line-number="18"><span class="co">#&gt; [1,]    1    1</span></a>
<a class="sourceLine" id="cb41-19" data-line-number="19"><span class="co">#&gt; [2,]    1    1</span></a>
<a class="sourceLine" id="cb41-20" data-line-number="20"><span class="co">#&gt; [3,]    1    1</span></a>
<a class="sourceLine" id="cb41-21" data-line-number="21"><span class="co">#&gt; [4,]    1    1</span></a>
<a class="sourceLine" id="cb41-22" data-line-number="22"><span class="co">#&gt; [5,]    1    1</span></a></code></pre></div>
</div>
</div>
<div id="implementation-details-1" class="section level2">
<h2 class="hasAnchor">
<a href="#implementation-details-1" class="anchor"></a>Implementation details</h2>
<p>The <em>GTuplesList</em> class extends the <em>GenomicRanges::GRangesList</em> class.</p>
</div>
</div>
<div id="findoverlaps-based-methods" class="section level1">
<h1 class="hasAnchor">
<a href="#findoverlaps-based-methods" class="anchor"></a><code>findOverlaps</code>-based methods</h1>
<p>The definition of what constitutes an “overlap” between genomic tuples, or between genomic tuples and genomic ranges, lies at the heart of all <code>findOverlaps</code>-based methods<a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a> for <em>GTuples</em> and <em>GTuplesList</em> objects.</p>
<p>I have chosen a definition that matches my intuition of what constitutes an “overlap” between genomic tuples or between genomic tuples and genomic ranges. However, I am open to suggestions on amending or extending this behaviour in future versions of <em><a href="http://bioconductor.org/packages/GenomicTuples">GenomicTuples</a></em>.</p>
<div id="definition-of-overlapping-genomic-tuples" class="section level2">
<h2 class="hasAnchor">
<a href="#definition-of-overlapping-genomic-tuples" class="anchor"></a>Definition of overlapping genomic tuples</h2>
<p>I consider two genomic tuples to be <em>equal</em> (<code>type = "equal"</code>) if they have identical sequence names (<code>seqnames</code>), strands (<code>strand</code>) and tuples (<code>tuples</code>). For 1-tuples and 2-tuples, this means we can simply defer to the <code>findOverlaps</code>-based methods for <em>GenomicRanges::GRanges</em> and <em>GenomicRanges::GRangesList</em> objects via inheritance. However, we cannot do the same for m-tuples with <span class="math inline">\(m &gt; 2\)</span> since this would ignore the “internal positions”. Therefore, I have implemented a special case of the <code>findOverlaps</code> method for when <code>size</code> <span class="math inline">\(&gt; 2\)</span> and <code>type = "equal"</code>, which ensures that the “internal positions” are also checked for equality.</p>
<p><strong>In all other cases genomic tuples are treated as genomic ranges.</strong> This means that when <code>type = "any"</code>, <code>type = "start"</code>, <code>type = "end"</code> or <code>type = "within"</code> then the genomic tuples are treated as if they were genomic ranges. Specifically, <em>GTuples</em> (resp. <em>GTuplesList</em>) are treated as though they were <em>GenomicRanges::GRanges</em> (resp. <em>GenomicRanges::GRangesList</em>) with <code>pos1</code> = <code>start</code> and <code>posm</code> = <code>end</code>.</p>
</div>
<div id="definition-of-overlapping-genomic-tuples-and-ranges" class="section level2">
<h2 class="hasAnchor">
<a href="#definition-of-overlapping-genomic-tuples-and-ranges" class="anchor"></a>Definition of overlapping genomic tuples and ranges</h2>
<p>Genomic tuples are <strong>always</strong> treated as genomic ranges when searching for overlaps between genomic tuples and genomic ranges.</p>
</div>
<div id="examples" class="section level2">
<h2 class="hasAnchor">
<a href="#examples" class="anchor"></a>Examples</h2>
<p>It is easiest to understand the above definitions by studying a few examples.</p>
<p>Firstly, for 1-tuples where the <em>GTuples</em> methods use the <em>GenomicRanges::GRanges</em> methods:</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb42-1" data-line-number="1"><span class="co"># Construct example 1-tuples</span></a>
<a class="sourceLine" id="cb42-2" data-line-number="2">gt1 &lt;-<span class="st"> </span><span class="kw">GTuples</span>(<span class="dt">seqnames =</span> <span class="kw">c</span>(<span class="st">'chr1'</span>, <span class="st">'chr1'</span>, <span class="st">'chr1'</span>, <span class="st">'chr2'</span>), </a>
<a class="sourceLine" id="cb42-3" data-line-number="3">               <span class="dt">tuples =</span> <span class="kw">matrix</span>(<span class="kw">c</span>(10L, 10L, 10L, 10L), <span class="dt">ncol =</span> <span class="dv">1</span>), </a>
<a class="sourceLine" id="cb42-4" data-line-number="4">               <span class="dt">strand =</span> <span class="kw">c</span>(<span class="st">'+'</span>, <span class="st">'-'</span>, <span class="st">'*'</span>, <span class="st">'+'</span>))</a>
<a class="sourceLine" id="cb42-5" data-line-number="5"><span class="co"># GRanges version of gt1</span></a>
<a class="sourceLine" id="cb42-6" data-line-number="6">gr1 &lt;-<span class="st"> </span><span class="kw">as</span>(gt1, <span class="st">"GRanges"</span>)</a>
<a class="sourceLine" id="cb42-7" data-line-number="7"><span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt1, gt1, <span class="dt">type =</span> <span class="st">'any'</span>)</a>
<a class="sourceLine" id="cb42-8" data-line-number="8"><span class="co">#&gt; Hits object with 8 hits and 0 metadata columns:</span></a>
<a class="sourceLine" id="cb42-9" data-line-number="9"><span class="co">#&gt;       queryHits subjectHits</span></a>
<a class="sourceLine" id="cb42-10" data-line-number="10"><span class="co">#&gt;       &lt;integer&gt;   &lt;integer&gt;</span></a>
<a class="sourceLine" id="cb42-11" data-line-number="11"><span class="co">#&gt;   [1]         1           1</span></a>
<a class="sourceLine" id="cb42-12" data-line-number="12"><span class="co">#&gt;   [2]         1           3</span></a>
<a class="sourceLine" id="cb42-13" data-line-number="13"><span class="co">#&gt;   [3]         2           2</span></a>
<a class="sourceLine" id="cb42-14" data-line-number="14"><span class="co">#&gt;   [4]         2           3</span></a>
<a class="sourceLine" id="cb42-15" data-line-number="15"><span class="co">#&gt;   [5]         3           1</span></a>
<a class="sourceLine" id="cb42-16" data-line-number="16"><span class="co">#&gt;   [6]         3           2</span></a>
<a class="sourceLine" id="cb42-17" data-line-number="17"><span class="co">#&gt;   [7]         3           3</span></a>
<a class="sourceLine" id="cb42-18" data-line-number="18"><span class="co">#&gt;   [8]         4           4</span></a>
<a class="sourceLine" id="cb42-19" data-line-number="19"><span class="co">#&gt;   -------</span></a>
<a class="sourceLine" id="cb42-20" data-line-number="20"><span class="co">#&gt;   queryLength: 4 / subjectLength: 4</span></a>
<a class="sourceLine" id="cb42-21" data-line-number="21"><span class="co"># GTuples and GRanges methods identical</span></a>
<a class="sourceLine" id="cb42-22" data-line-number="22"><span class="kw">identical</span>(<span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt1, gt1, <span class="dt">type =</span> <span class="st">'any'</span>), </a>
<a class="sourceLine" id="cb42-23" data-line-number="23">          <span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gr1, gr1, <span class="dt">type =</span> <span class="st">'any'</span>))</a>
<a class="sourceLine" id="cb42-24" data-line-number="24"><span class="co">#&gt; [1] TRUE</span></a>
<a class="sourceLine" id="cb42-25" data-line-number="25"><span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt1, gt1, <span class="dt">type =</span> <span class="st">'start'</span>)</a>
<a class="sourceLine" id="cb42-26" data-line-number="26"><span class="co">#&gt; Hits object with 8 hits and 0 metadata columns:</span></a>
<a class="sourceLine" id="cb42-27" data-line-number="27"><span class="co">#&gt;       queryHits subjectHits</span></a>
<a class="sourceLine" id="cb42-28" data-line-number="28"><span class="co">#&gt;       &lt;integer&gt;   &lt;integer&gt;</span></a>
<a class="sourceLine" id="cb42-29" data-line-number="29"><span class="co">#&gt;   [1]         1           1</span></a>
<a class="sourceLine" id="cb42-30" data-line-number="30"><span class="co">#&gt;   [2]         1           3</span></a>
<a class="sourceLine" id="cb42-31" data-line-number="31"><span class="co">#&gt;   [3]         2           2</span></a>
<a class="sourceLine" id="cb42-32" data-line-number="32"><span class="co">#&gt;   [4]         2           3</span></a>
<a class="sourceLine" id="cb42-33" data-line-number="33"><span class="co">#&gt;   [5]         3           1</span></a>
<a class="sourceLine" id="cb42-34" data-line-number="34"><span class="co">#&gt;   [6]         3           2</span></a>
<a class="sourceLine" id="cb42-35" data-line-number="35"><span class="co">#&gt;   [7]         3           3</span></a>
<a class="sourceLine" id="cb42-36" data-line-number="36"><span class="co">#&gt;   [8]         4           4</span></a>
<a class="sourceLine" id="cb42-37" data-line-number="37"><span class="co">#&gt;   -------</span></a>
<a class="sourceLine" id="cb42-38" data-line-number="38"><span class="co">#&gt;   queryLength: 4 / subjectLength: 4</span></a>
<a class="sourceLine" id="cb42-39" data-line-number="39"><span class="co"># GTuples and GRanges methods identical</span></a>
<a class="sourceLine" id="cb42-40" data-line-number="40"><span class="kw">identical</span>(<span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt1, gt1, <span class="dt">type =</span> <span class="st">'start'</span>), </a>
<a class="sourceLine" id="cb42-41" data-line-number="41">          <span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gr1, gr1, <span class="dt">type =</span> <span class="st">'start'</span>))</a>
<a class="sourceLine" id="cb42-42" data-line-number="42"><span class="co">#&gt; [1] TRUE</span></a>
<a class="sourceLine" id="cb42-43" data-line-number="43"><span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt1, gt1, <span class="dt">type =</span> <span class="st">'end'</span>)</a>
<a class="sourceLine" id="cb42-44" data-line-number="44"><span class="co">#&gt; Hits object with 8 hits and 0 metadata columns:</span></a>
<a class="sourceLine" id="cb42-45" data-line-number="45"><span class="co">#&gt;       queryHits subjectHits</span></a>
<a class="sourceLine" id="cb42-46" data-line-number="46"><span class="co">#&gt;       &lt;integer&gt;   &lt;integer&gt;</span></a>
<a class="sourceLine" id="cb42-47" data-line-number="47"><span class="co">#&gt;   [1]         1           1</span></a>
<a class="sourceLine" id="cb42-48" data-line-number="48"><span class="co">#&gt;   [2]         1           3</span></a>
<a class="sourceLine" id="cb42-49" data-line-number="49"><span class="co">#&gt;   [3]         2           2</span></a>
<a class="sourceLine" id="cb42-50" data-line-number="50"><span class="co">#&gt;   [4]         2           3</span></a>
<a class="sourceLine" id="cb42-51" data-line-number="51"><span class="co">#&gt;   [5]         3           1</span></a>
<a class="sourceLine" id="cb42-52" data-line-number="52"><span class="co">#&gt;   [6]         3           2</span></a>
<a class="sourceLine" id="cb42-53" data-line-number="53"><span class="co">#&gt;   [7]         3           3</span></a>
<a class="sourceLine" id="cb42-54" data-line-number="54"><span class="co">#&gt;   [8]         4           4</span></a>
<a class="sourceLine" id="cb42-55" data-line-number="55"><span class="co">#&gt;   -------</span></a>
<a class="sourceLine" id="cb42-56" data-line-number="56"><span class="co">#&gt;   queryLength: 4 / subjectLength: 4</span></a>
<a class="sourceLine" id="cb42-57" data-line-number="57"><span class="co"># GTuples and GRanges methods identical</span></a>
<a class="sourceLine" id="cb42-58" data-line-number="58"><span class="kw">identical</span>(<span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt1, gt1, <span class="dt">type =</span> <span class="st">'end'</span>), </a>
<a class="sourceLine" id="cb42-59" data-line-number="59">          <span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gr1, gr1, <span class="dt">type =</span> <span class="st">'end'</span>))</a>
<a class="sourceLine" id="cb42-60" data-line-number="60"><span class="co">#&gt; [1] TRUE</span></a>
<a class="sourceLine" id="cb42-61" data-line-number="61"><span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt1, gt1, <span class="dt">type =</span> <span class="st">'within'</span>)</a>
<a class="sourceLine" id="cb42-62" data-line-number="62"><span class="co">#&gt; Hits object with 8 hits and 0 metadata columns:</span></a>
<a class="sourceLine" id="cb42-63" data-line-number="63"><span class="co">#&gt;       queryHits subjectHits</span></a>
<a class="sourceLine" id="cb42-64" data-line-number="64"><span class="co">#&gt;       &lt;integer&gt;   &lt;integer&gt;</span></a>
<a class="sourceLine" id="cb42-65" data-line-number="65"><span class="co">#&gt;   [1]         1           1</span></a>
<a class="sourceLine" id="cb42-66" data-line-number="66"><span class="co">#&gt;   [2]         1           3</span></a>
<a class="sourceLine" id="cb42-67" data-line-number="67"><span class="co">#&gt;   [3]         2           2</span></a>
<a class="sourceLine" id="cb42-68" data-line-number="68"><span class="co">#&gt;   [4]         2           3</span></a>
<a class="sourceLine" id="cb42-69" data-line-number="69"><span class="co">#&gt;   [5]         3           1</span></a>
<a class="sourceLine" id="cb42-70" data-line-number="70"><span class="co">#&gt;   [6]         3           2</span></a>
<a class="sourceLine" id="cb42-71" data-line-number="71"><span class="co">#&gt;   [7]         3           3</span></a>
<a class="sourceLine" id="cb42-72" data-line-number="72"><span class="co">#&gt;   [8]         4           4</span></a>
<a class="sourceLine" id="cb42-73" data-line-number="73"><span class="co">#&gt;   -------</span></a>
<a class="sourceLine" id="cb42-74" data-line-number="74"><span class="co">#&gt;   queryLength: 4 / subjectLength: 4</span></a>
<a class="sourceLine" id="cb42-75" data-line-number="75"><span class="co"># GTuples and GRanges methods identical</span></a>
<a class="sourceLine" id="cb42-76" data-line-number="76"><span class="kw">identical</span>(<span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt1, gt1, <span class="dt">type =</span> <span class="st">'within'</span>), </a>
<a class="sourceLine" id="cb42-77" data-line-number="77">          <span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gr1, gr1, <span class="dt">type =</span> <span class="st">'within'</span>))</a>
<a class="sourceLine" id="cb42-78" data-line-number="78"><span class="co">#&gt; [1] TRUE</span></a>
<a class="sourceLine" id="cb42-79" data-line-number="79"><span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt1, gt1, <span class="dt">type =</span> <span class="st">'equal'</span>)</a>
<a class="sourceLine" id="cb42-80" data-line-number="80"><span class="co">#&gt; Hits object with 8 hits and 0 metadata columns:</span></a>
<a class="sourceLine" id="cb42-81" data-line-number="81"><span class="co">#&gt;       queryHits subjectHits</span></a>
<a class="sourceLine" id="cb42-82" data-line-number="82"><span class="co">#&gt;       &lt;integer&gt;   &lt;integer&gt;</span></a>
<a class="sourceLine" id="cb42-83" data-line-number="83"><span class="co">#&gt;   [1]         1           1</span></a>
<a class="sourceLine" id="cb42-84" data-line-number="84"><span class="co">#&gt;   [2]         1           3</span></a>
<a class="sourceLine" id="cb42-85" data-line-number="85"><span class="co">#&gt;   [3]         2           2</span></a>
<a class="sourceLine" id="cb42-86" data-line-number="86"><span class="co">#&gt;   [4]         2           3</span></a>
<a class="sourceLine" id="cb42-87" data-line-number="87"><span class="co">#&gt;   [5]         3           1</span></a>
<a class="sourceLine" id="cb42-88" data-line-number="88"><span class="co">#&gt;   [6]         3           2</span></a>
<a class="sourceLine" id="cb42-89" data-line-number="89"><span class="co">#&gt;   [7]         3           3</span></a>
<a class="sourceLine" id="cb42-90" data-line-number="90"><span class="co">#&gt;   [8]         4           4</span></a>
<a class="sourceLine" id="cb42-91" data-line-number="91"><span class="co">#&gt;   -------</span></a>
<a class="sourceLine" id="cb42-92" data-line-number="92"><span class="co">#&gt;   queryLength: 4 / subjectLength: 4</span></a>
<a class="sourceLine" id="cb42-93" data-line-number="93"><span class="co"># GTuples and GRanges methods identical</span></a>
<a class="sourceLine" id="cb42-94" data-line-number="94"><span class="kw">identical</span>(<span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt1, gt1, <span class="dt">type =</span> <span class="st">'equal'</span>), </a>
<a class="sourceLine" id="cb42-95" data-line-number="95">          <span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gr1, gr1, <span class="dt">type =</span> <span class="st">'equal'</span>))</a>
<a class="sourceLine" id="cb42-96" data-line-number="96"><span class="co">#&gt; [1] TRUE</span></a>
<a class="sourceLine" id="cb42-97" data-line-number="97"><span class="co"># Can pass other arguments, such as select and ignore.strand</span></a>
<a class="sourceLine" id="cb42-98" data-line-number="98"><span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt1, gt1, <span class="dt">type =</span> <span class="st">'equal'</span>, <span class="dt">ignore.strand =</span> <span class="ot">TRUE</span>, <span class="dt">select =</span> <span class="st">'last'</span>)</a>
<a class="sourceLine" id="cb42-99" data-line-number="99"><span class="co">#&gt; [1] 3 3 3 4</span></a></code></pre></div>
<p>Next, for 2-tuples where the <em>GTuples</em> methods use the <em>GenomicRanges::GRanges</em> methods:</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb43-1" data-line-number="1"><span class="co"># Construct example 2-tuples</span></a>
<a class="sourceLine" id="cb43-2" data-line-number="2">gt2 &lt;-<span class="st"> </span><span class="kw">GTuples</span>(<span class="dt">seqnames =</span> <span class="kw">c</span>(<span class="st">'chr1'</span>, <span class="st">'chr1'</span>, <span class="st">'chr1'</span>, <span class="st">'chr1'</span>, <span class="st">'chr2'</span>), </a>
<a class="sourceLine" id="cb43-3" data-line-number="3">               <span class="dt">tuples =</span> <span class="kw">matrix</span>(<span class="kw">c</span>(10L, 10L, 10L, 10L, 10L, 20L, 20L, 20L, 25L, </a>
<a class="sourceLine" id="cb43-4" data-line-number="4">                                 20L), <span class="dt">ncol =</span> <span class="dv">2</span>), </a>
<a class="sourceLine" id="cb43-5" data-line-number="5">               <span class="dt">strand =</span> <span class="kw">c</span>(<span class="st">'+'</span>, <span class="st">'-'</span>, <span class="st">'*'</span>, <span class="st">'+'</span>, <span class="st">'+'</span>))</a>
<a class="sourceLine" id="cb43-6" data-line-number="6"><span class="co"># GRanges version of gt2</span></a>
<a class="sourceLine" id="cb43-7" data-line-number="7">gr2 &lt;-<span class="st"> </span><span class="kw">as</span>(gt2, <span class="st">"GRanges"</span>)</a>
<a class="sourceLine" id="cb43-8" data-line-number="8"><span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt2, gt2, <span class="dt">type =</span> <span class="st">'any'</span>)</a>
<a class="sourceLine" id="cb43-9" data-line-number="9"><span class="co">#&gt; Hits object with 13 hits and 0 metadata columns:</span></a>
<a class="sourceLine" id="cb43-10" data-line-number="10"><span class="co">#&gt;        queryHits subjectHits</span></a>
<a class="sourceLine" id="cb43-11" data-line-number="11"><span class="co">#&gt;        &lt;integer&gt;   &lt;integer&gt;</span></a>
<a class="sourceLine" id="cb43-12" data-line-number="12"><span class="co">#&gt;    [1]         1           4</span></a>
<a class="sourceLine" id="cb43-13" data-line-number="13"><span class="co">#&gt;    [2]         1           1</span></a>
<a class="sourceLine" id="cb43-14" data-line-number="14"><span class="co">#&gt;    [3]         1           3</span></a>
<a class="sourceLine" id="cb43-15" data-line-number="15"><span class="co">#&gt;    [4]         2           2</span></a>
<a class="sourceLine" id="cb43-16" data-line-number="16"><span class="co">#&gt;    [5]         2           3</span></a>
<a class="sourceLine" id="cb43-17" data-line-number="17"><span class="co">#&gt;    ...       ...         ...</span></a>
<a class="sourceLine" id="cb43-18" data-line-number="18"><span class="co">#&gt;    [9]         3           3</span></a>
<a class="sourceLine" id="cb43-19" data-line-number="19"><span class="co">#&gt;   [10]         4           4</span></a>
<a class="sourceLine" id="cb43-20" data-line-number="20"><span class="co">#&gt;   [11]         4           1</span></a>
<a class="sourceLine" id="cb43-21" data-line-number="21"><span class="co">#&gt;   [12]         4           3</span></a>
<a class="sourceLine" id="cb43-22" data-line-number="22"><span class="co">#&gt;   [13]         5           5</span></a>
<a class="sourceLine" id="cb43-23" data-line-number="23"><span class="co">#&gt;   -------</span></a>
<a class="sourceLine" id="cb43-24" data-line-number="24"><span class="co">#&gt;   queryLength: 5 / subjectLength: 5</span></a>
<a class="sourceLine" id="cb43-25" data-line-number="25"><span class="co"># GTuples and GRanges methods identical</span></a>
<a class="sourceLine" id="cb43-26" data-line-number="26"><span class="kw">identical</span>(<span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt2, gt2, <span class="dt">type =</span> <span class="st">'any'</span>), </a>
<a class="sourceLine" id="cb43-27" data-line-number="27">          <span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gr2, gr2, <span class="dt">type =</span> <span class="st">'any'</span>))</a>
<a class="sourceLine" id="cb43-28" data-line-number="28"><span class="co">#&gt; [1] TRUE</span></a>
<a class="sourceLine" id="cb43-29" data-line-number="29"><span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt2, gt2, <span class="dt">type =</span> <span class="st">'start'</span>)</a>
<a class="sourceLine" id="cb43-30" data-line-number="30"><span class="co">#&gt; Hits object with 13 hits and 0 metadata columns:</span></a>
<a class="sourceLine" id="cb43-31" data-line-number="31"><span class="co">#&gt;        queryHits subjectHits</span></a>
<a class="sourceLine" id="cb43-32" data-line-number="32"><span class="co">#&gt;        &lt;integer&gt;   &lt;integer&gt;</span></a>
<a class="sourceLine" id="cb43-33" data-line-number="33"><span class="co">#&gt;    [1]         1           4</span></a>
<a class="sourceLine" id="cb43-34" data-line-number="34"><span class="co">#&gt;    [2]         1           1</span></a>
<a class="sourceLine" id="cb43-35" data-line-number="35"><span class="co">#&gt;    [3]         1           3</span></a>
<a class="sourceLine" id="cb43-36" data-line-number="36"><span class="co">#&gt;    [4]         2           2</span></a>
<a class="sourceLine" id="cb43-37" data-line-number="37"><span class="co">#&gt;    [5]         2           3</span></a>
<a class="sourceLine" id="cb43-38" data-line-number="38"><span class="co">#&gt;    ...       ...         ...</span></a>
<a class="sourceLine" id="cb43-39" data-line-number="39"><span class="co">#&gt;    [9]         3           3</span></a>
<a class="sourceLine" id="cb43-40" data-line-number="40"><span class="co">#&gt;   [10]         4           4</span></a>
<a class="sourceLine" id="cb43-41" data-line-number="41"><span class="co">#&gt;   [11]         4           1</span></a>
<a class="sourceLine" id="cb43-42" data-line-number="42"><span class="co">#&gt;   [12]         4           3</span></a>
<a class="sourceLine" id="cb43-43" data-line-number="43"><span class="co">#&gt;   [13]         5           5</span></a>
<a class="sourceLine" id="cb43-44" data-line-number="44"><span class="co">#&gt;   -------</span></a>
<a class="sourceLine" id="cb43-45" data-line-number="45"><span class="co">#&gt;   queryLength: 5 / subjectLength: 5</span></a>
<a class="sourceLine" id="cb43-46" data-line-number="46"><span class="co"># GTuples and GRanges methods identical</span></a>
<a class="sourceLine" id="cb43-47" data-line-number="47"><span class="kw">identical</span>(<span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt2, gt2, <span class="dt">type =</span> <span class="st">'start'</span>), </a>
<a class="sourceLine" id="cb43-48" data-line-number="48">          <span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gr2, gr2, <span class="dt">type =</span> <span class="st">'start'</span>))</a>
<a class="sourceLine" id="cb43-49" data-line-number="49"><span class="co">#&gt; [1] TRUE</span></a>
<a class="sourceLine" id="cb43-50" data-line-number="50"><span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt2, gt2, <span class="dt">type =</span> <span class="st">'end'</span>)</a>
<a class="sourceLine" id="cb43-51" data-line-number="51"><span class="co">#&gt; Hits object with 9 hits and 0 metadata columns:</span></a>
<a class="sourceLine" id="cb43-52" data-line-number="52"><span class="co">#&gt;       queryHits subjectHits</span></a>
<a class="sourceLine" id="cb43-53" data-line-number="53"><span class="co">#&gt;       &lt;integer&gt;   &lt;integer&gt;</span></a>
<a class="sourceLine" id="cb43-54" data-line-number="54"><span class="co">#&gt;   [1]         1           1</span></a>
<a class="sourceLine" id="cb43-55" data-line-number="55"><span class="co">#&gt;   [2]         1           3</span></a>
<a class="sourceLine" id="cb43-56" data-line-number="56"><span class="co">#&gt;   [3]         2           2</span></a>
<a class="sourceLine" id="cb43-57" data-line-number="57"><span class="co">#&gt;   [4]         2           3</span></a>
<a class="sourceLine" id="cb43-58" data-line-number="58"><span class="co">#&gt;   [5]         3           1</span></a>
<a class="sourceLine" id="cb43-59" data-line-number="59"><span class="co">#&gt;   [6]         3           2</span></a>
<a class="sourceLine" id="cb43-60" data-line-number="60"><span class="co">#&gt;   [7]         3           3</span></a>
<a class="sourceLine" id="cb43-61" data-line-number="61"><span class="co">#&gt;   [8]         4           4</span></a>
<a class="sourceLine" id="cb43-62" data-line-number="62"><span class="co">#&gt;   [9]         5           5</span></a>
<a class="sourceLine" id="cb43-63" data-line-number="63"><span class="co">#&gt;   -------</span></a>
<a class="sourceLine" id="cb43-64" data-line-number="64"><span class="co">#&gt;   queryLength: 5 / subjectLength: 5</span></a>
<a class="sourceLine" id="cb43-65" data-line-number="65"><span class="co"># GTuples and GRanges methods identical</span></a>
<a class="sourceLine" id="cb43-66" data-line-number="66"><span class="kw">identical</span>(<span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt2, gt2, <span class="dt">type =</span> <span class="st">'end'</span>), </a>
<a class="sourceLine" id="cb43-67" data-line-number="67">          <span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gr2, gr2, <span class="dt">type =</span> <span class="st">'end'</span>))</a>
<a class="sourceLine" id="cb43-68" data-line-number="68"><span class="co">#&gt; [1] TRUE</span></a>
<a class="sourceLine" id="cb43-69" data-line-number="69"><span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt2, gt2, <span class="dt">type =</span> <span class="st">'within'</span>)</a>
<a class="sourceLine" id="cb43-70" data-line-number="70"><span class="co">#&gt; Hits object with 11 hits and 0 metadata columns:</span></a>
<a class="sourceLine" id="cb43-71" data-line-number="71"><span class="co">#&gt;        queryHits subjectHits</span></a>
<a class="sourceLine" id="cb43-72" data-line-number="72"><span class="co">#&gt;        &lt;integer&gt;   &lt;integer&gt;</span></a>
<a class="sourceLine" id="cb43-73" data-line-number="73"><span class="co">#&gt;    [1]         1           4</span></a>
<a class="sourceLine" id="cb43-74" data-line-number="74"><span class="co">#&gt;    [2]         1           1</span></a>
<a class="sourceLine" id="cb43-75" data-line-number="75"><span class="co">#&gt;    [3]         1           3</span></a>
<a class="sourceLine" id="cb43-76" data-line-number="76"><span class="co">#&gt;    [4]         2           2</span></a>
<a class="sourceLine" id="cb43-77" data-line-number="77"><span class="co">#&gt;    [5]         2           3</span></a>
<a class="sourceLine" id="cb43-78" data-line-number="78"><span class="co">#&gt;    [6]         3           4</span></a>
<a class="sourceLine" id="cb43-79" data-line-number="79"><span class="co">#&gt;    [7]         3           1</span></a>
<a class="sourceLine" id="cb43-80" data-line-number="80"><span class="co">#&gt;    [8]         3           2</span></a>
<a class="sourceLine" id="cb43-81" data-line-number="81"><span class="co">#&gt;    [9]         3           3</span></a>
<a class="sourceLine" id="cb43-82" data-line-number="82"><span class="co">#&gt;   [10]         4           4</span></a>
<a class="sourceLine" id="cb43-83" data-line-number="83"><span class="co">#&gt;   [11]         5           5</span></a>
<a class="sourceLine" id="cb43-84" data-line-number="84"><span class="co">#&gt;   -------</span></a>
<a class="sourceLine" id="cb43-85" data-line-number="85"><span class="co">#&gt;   queryLength: 5 / subjectLength: 5</span></a>
<a class="sourceLine" id="cb43-86" data-line-number="86"><span class="co"># GTuples and GRanges methods identical</span></a>
<a class="sourceLine" id="cb43-87" data-line-number="87"><span class="kw">identical</span>(<span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt2, gt2, <span class="dt">type =</span> <span class="st">'within'</span>), </a>
<a class="sourceLine" id="cb43-88" data-line-number="88">          <span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gr2, gr2, <span class="dt">type =</span> <span class="st">'within'</span>))</a>
<a class="sourceLine" id="cb43-89" data-line-number="89"><span class="co">#&gt; [1] TRUE</span></a>
<a class="sourceLine" id="cb43-90" data-line-number="90"><span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt2, gt2, <span class="dt">type =</span> <span class="st">'equal'</span>)</a>
<a class="sourceLine" id="cb43-91" data-line-number="91"><span class="co">#&gt; Hits object with 9 hits and 0 metadata columns:</span></a>
<a class="sourceLine" id="cb43-92" data-line-number="92"><span class="co">#&gt;       queryHits subjectHits</span></a>
<a class="sourceLine" id="cb43-93" data-line-number="93"><span class="co">#&gt;       &lt;integer&gt;   &lt;integer&gt;</span></a>
<a class="sourceLine" id="cb43-94" data-line-number="94"><span class="co">#&gt;   [1]         1           1</span></a>
<a class="sourceLine" id="cb43-95" data-line-number="95"><span class="co">#&gt;   [2]         1           3</span></a>
<a class="sourceLine" id="cb43-96" data-line-number="96"><span class="co">#&gt;   [3]         2           2</span></a>
<a class="sourceLine" id="cb43-97" data-line-number="97"><span class="co">#&gt;   [4]         2           3</span></a>
<a class="sourceLine" id="cb43-98" data-line-number="98"><span class="co">#&gt;   [5]         3           1</span></a>
<a class="sourceLine" id="cb43-99" data-line-number="99"><span class="co">#&gt;   [6]         3           2</span></a>
<a class="sourceLine" id="cb43-100" data-line-number="100"><span class="co">#&gt;   [7]         3           3</span></a>
<a class="sourceLine" id="cb43-101" data-line-number="101"><span class="co">#&gt;   [8]         4           4</span></a>
<a class="sourceLine" id="cb43-102" data-line-number="102"><span class="co">#&gt;   [9]         5           5</span></a>
<a class="sourceLine" id="cb43-103" data-line-number="103"><span class="co">#&gt;   -------</span></a>
<a class="sourceLine" id="cb43-104" data-line-number="104"><span class="co">#&gt;   queryLength: 5 / subjectLength: 5</span></a>
<a class="sourceLine" id="cb43-105" data-line-number="105"><span class="co"># GTuples and GRanges methods identical</span></a>
<a class="sourceLine" id="cb43-106" data-line-number="106"><span class="kw">identical</span>(<span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt2, gt2, <span class="dt">type =</span> <span class="st">'equal'</span>), </a>
<a class="sourceLine" id="cb43-107" data-line-number="107">          <span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gr2, gr2, <span class="dt">type =</span> <span class="st">'equal'</span>))</a>
<a class="sourceLine" id="cb43-108" data-line-number="108"><span class="co">#&gt; [1] TRUE</span></a>
<a class="sourceLine" id="cb43-109" data-line-number="109"><span class="co"># Can pass other arguments, such as select and ignore.strand</span></a>
<a class="sourceLine" id="cb43-110" data-line-number="110"><span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt2, gt2, <span class="dt">type =</span> <span class="st">'equal'</span>, <span class="dt">ignore.strand =</span> <span class="ot">TRUE</span>, <span class="dt">select =</span> <span class="st">'last'</span>)</a>
<a class="sourceLine" id="cb43-111" data-line-number="111"><span class="co">#&gt; [1] 3 3 3 4 5</span></a></code></pre></div>
<p>Finally, for m-tuples with <span class="math inline">\(m &gt; 2\)</span> where <em>GTuples</em> methods use the <em>GenomicRanges::GRanges</em> methods <strong>unless <code>type = "equal"</code></strong>:</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb44-1" data-line-number="1"><span class="co"># Construct example 3-tuples</span></a>
<a class="sourceLine" id="cb44-2" data-line-number="2">gt3 &lt;-<span class="st"> </span><span class="kw">GTuples</span>(<span class="dt">seqnames =</span> <span class="kw">c</span>(<span class="st">'chr1'</span>, <span class="st">'chr1'</span>, <span class="st">'chr1'</span>, <span class="st">'chr1'</span>, <span class="st">'chr2'</span>), </a>
<a class="sourceLine" id="cb44-3" data-line-number="3">               <span class="dt">tuples =</span> <span class="kw">matrix</span>(<span class="kw">c</span>(10L, 10L, 10L, 10L, 10L, 20L, 20L, 20L, 25L, </a>
<a class="sourceLine" id="cb44-4" data-line-number="4">                                 20L, 30L, 30L, 35L, 30L, 30L), <span class="dt">ncol =</span> <span class="dv">3</span>), </a>
<a class="sourceLine" id="cb44-5" data-line-number="5">               <span class="dt">strand =</span> <span class="kw">c</span>(<span class="st">'+'</span>, <span class="st">'-'</span>, <span class="st">'*'</span>, <span class="st">'+'</span>, <span class="st">'+'</span>))</a>
<a class="sourceLine" id="cb44-6" data-line-number="6"><span class="co"># GRanges version of gt3</span></a>
<a class="sourceLine" id="cb44-7" data-line-number="7">gr3 &lt;-<span class="st"> </span><span class="kw">as</span>(gt3, <span class="st">"GRanges"</span>)</a>
<a class="sourceLine" id="cb44-8" data-line-number="8"><span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt3, gt3, <span class="dt">type =</span> <span class="st">'any'</span>)</a>
<a class="sourceLine" id="cb44-9" data-line-number="9"><span class="co">#&gt; Warning in .local(query, subject, maxgap, minoverlap, type, select, ...):</span></a>
<a class="sourceLine" id="cb44-10" data-line-number="10"><span class="co">#&gt; 'type' is not 'equal' so coercing 'query' and 'subject' to 'GRanges'</span></a>
<a class="sourceLine" id="cb44-11" data-line-number="11"><span class="co">#&gt; objects (see docs for details)</span></a>
<a class="sourceLine" id="cb44-12" data-line-number="12"><span class="co">#&gt; Hits object with 13 hits and 0 metadata columns:</span></a>
<a class="sourceLine" id="cb44-13" data-line-number="13"><span class="co">#&gt;        queryHits subjectHits</span></a>
<a class="sourceLine" id="cb44-14" data-line-number="14"><span class="co">#&gt;        &lt;integer&gt;   &lt;integer&gt;</span></a>
<a class="sourceLine" id="cb44-15" data-line-number="15"><span class="co">#&gt;    [1]         1           3</span></a>
<a class="sourceLine" id="cb44-16" data-line-number="16"><span class="co">#&gt;    [2]         1           1</span></a>
<a class="sourceLine" id="cb44-17" data-line-number="17"><span class="co">#&gt;    [3]         1           4</span></a>
<a class="sourceLine" id="cb44-18" data-line-number="18"><span class="co">#&gt;    [4]         2           3</span></a>
<a class="sourceLine" id="cb44-19" data-line-number="19"><span class="co">#&gt;    [5]         2           2</span></a>
<a class="sourceLine" id="cb44-20" data-line-number="20"><span class="co">#&gt;    ...       ...         ...</span></a>
<a class="sourceLine" id="cb44-21" data-line-number="21"><span class="co">#&gt;    [9]         3           4</span></a>
<a class="sourceLine" id="cb44-22" data-line-number="22"><span class="co">#&gt;   [10]         4           3</span></a>
<a class="sourceLine" id="cb44-23" data-line-number="23"><span class="co">#&gt;   [11]         4           1</span></a>
<a class="sourceLine" id="cb44-24" data-line-number="24"><span class="co">#&gt;   [12]         4           4</span></a>
<a class="sourceLine" id="cb44-25" data-line-number="25"><span class="co">#&gt;   [13]         5           5</span></a>
<a class="sourceLine" id="cb44-26" data-line-number="26"><span class="co">#&gt;   -------</span></a>
<a class="sourceLine" id="cb44-27" data-line-number="27"><span class="co">#&gt;   queryLength: 5 / subjectLength: 5</span></a>
<a class="sourceLine" id="cb44-28" data-line-number="28"><span class="co"># GTuples and GRanges methods identical</span></a>
<a class="sourceLine" id="cb44-29" data-line-number="29"><span class="kw">identical</span>(<span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt3, gt3, <span class="dt">type =</span> <span class="st">'any'</span>), </a>
<a class="sourceLine" id="cb44-30" data-line-number="30">          <span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gr3, gr3, <span class="dt">type =</span> <span class="st">'any'</span>)) <span class="co"># TRUE</span></a>
<a class="sourceLine" id="cb44-31" data-line-number="31"><span class="co">#&gt; Warning in .local(query, subject, maxgap, minoverlap, type, select, ...):</span></a>
<a class="sourceLine" id="cb44-32" data-line-number="32"><span class="co">#&gt; 'type' is not 'equal' so coercing 'query' and 'subject' to 'GRanges'</span></a>
<a class="sourceLine" id="cb44-33" data-line-number="33"><span class="co">#&gt; objects (see docs for details)</span></a>
<a class="sourceLine" id="cb44-34" data-line-number="34"><span class="co">#&gt; [1] TRUE</span></a>
<a class="sourceLine" id="cb44-35" data-line-number="35"></a>
<a class="sourceLine" id="cb44-36" data-line-number="36"><span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt3, gt3, <span class="dt">type =</span> <span class="st">'start'</span>)</a>
<a class="sourceLine" id="cb44-37" data-line-number="37"><span class="co">#&gt; Warning in .local(query, subject, maxgap, minoverlap, type, select, ...):</span></a>
<a class="sourceLine" id="cb44-38" data-line-number="38"><span class="co">#&gt; 'type' is not 'equal' so coercing 'query' and 'subject' to 'GRanges'</span></a>
<a class="sourceLine" id="cb44-39" data-line-number="39"><span class="co">#&gt; objects (see docs for details)</span></a>
<a class="sourceLine" id="cb44-40" data-line-number="40"><span class="co">#&gt; Hits object with 13 hits and 0 metadata columns:</span></a>
<a class="sourceLine" id="cb44-41" data-line-number="41"><span class="co">#&gt;        queryHits subjectHits</span></a>
<a class="sourceLine" id="cb44-42" data-line-number="42"><span class="co">#&gt;        &lt;integer&gt;   &lt;integer&gt;</span></a>
<a class="sourceLine" id="cb44-43" data-line-number="43"><span class="co">#&gt;    [1]         1           3</span></a>
<a class="sourceLine" id="cb44-44" data-line-number="44"><span class="co">#&gt;    [2]         1           1</span></a>
<a class="sourceLine" id="cb44-45" data-line-number="45"><span class="co">#&gt;    [3]         1           4</span></a>
<a class="sourceLine" id="cb44-46" data-line-number="46"><span class="co">#&gt;    [4]         2           3</span></a>
<a class="sourceLine" id="cb44-47" data-line-number="47"><span class="co">#&gt;    [5]         2           2</span></a>
<a class="sourceLine" id="cb44-48" data-line-number="48"><span class="co">#&gt;    ...       ...         ...</span></a>
<a class="sourceLine" id="cb44-49" data-line-number="49"><span class="co">#&gt;    [9]         3           4</span></a>
<a class="sourceLine" id="cb44-50" data-line-number="50"><span class="co">#&gt;   [10]         4           3</span></a>
<a class="sourceLine" id="cb44-51" data-line-number="51"><span class="co">#&gt;   [11]         4           1</span></a>
<a class="sourceLine" id="cb44-52" data-line-number="52"><span class="co">#&gt;   [12]         4           4</span></a>
<a class="sourceLine" id="cb44-53" data-line-number="53"><span class="co">#&gt;   [13]         5           5</span></a>
<a class="sourceLine" id="cb44-54" data-line-number="54"><span class="co">#&gt;   -------</span></a>
<a class="sourceLine" id="cb44-55" data-line-number="55"><span class="co">#&gt;   queryLength: 5 / subjectLength: 5</span></a>
<a class="sourceLine" id="cb44-56" data-line-number="56"><span class="co"># GTuples and GRanges methods identical</span></a>
<a class="sourceLine" id="cb44-57" data-line-number="57"><span class="kw">identical</span>(<span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt3, gt3, <span class="dt">type =</span> <span class="st">'start'</span>), </a>
<a class="sourceLine" id="cb44-58" data-line-number="58">          <span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gr3, gr3, <span class="dt">type =</span> <span class="st">'start'</span>)) <span class="co"># TRUE</span></a>
<a class="sourceLine" id="cb44-59" data-line-number="59"><span class="co">#&gt; Warning in .local(query, subject, maxgap, minoverlap, type, select, ...):</span></a>
<a class="sourceLine" id="cb44-60" data-line-number="60"><span class="co">#&gt; 'type' is not 'equal' so coercing 'query' and 'subject' to 'GRanges'</span></a>
<a class="sourceLine" id="cb44-61" data-line-number="61"><span class="co">#&gt; objects (see docs for details)</span></a>
<a class="sourceLine" id="cb44-62" data-line-number="62"><span class="co">#&gt; [1] TRUE</span></a>
<a class="sourceLine" id="cb44-63" data-line-number="63"></a>
<a class="sourceLine" id="cb44-64" data-line-number="64"><span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt3, gt3, <span class="dt">type =</span> <span class="st">'end'</span>)</a>
<a class="sourceLine" id="cb44-65" data-line-number="65"><span class="co">#&gt; Warning in .local(query, subject, maxgap, minoverlap, type, select, ...):</span></a>
<a class="sourceLine" id="cb44-66" data-line-number="66"><span class="co">#&gt; 'type' is not 'equal' so coercing 'query' and 'subject' to 'GRanges'</span></a>
<a class="sourceLine" id="cb44-67" data-line-number="67"><span class="co">#&gt; objects (see docs for details)</span></a>
<a class="sourceLine" id="cb44-68" data-line-number="68"><span class="co">#&gt; Hits object with 7 hits and 0 metadata columns:</span></a>
<a class="sourceLine" id="cb44-69" data-line-number="69"><span class="co">#&gt;       queryHits subjectHits</span></a>
<a class="sourceLine" id="cb44-70" data-line-number="70"><span class="co">#&gt;       &lt;integer&gt;   &lt;integer&gt;</span></a>
<a class="sourceLine" id="cb44-71" data-line-number="71"><span class="co">#&gt;   [1]         1           1</span></a>
<a class="sourceLine" id="cb44-72" data-line-number="72"><span class="co">#&gt;   [2]         1           4</span></a>
<a class="sourceLine" id="cb44-73" data-line-number="73"><span class="co">#&gt;   [3]         2           2</span></a>
<a class="sourceLine" id="cb44-74" data-line-number="74"><span class="co">#&gt;   [4]         3           3</span></a>
<a class="sourceLine" id="cb44-75" data-line-number="75"><span class="co">#&gt;   [5]         4           1</span></a>
<a class="sourceLine" id="cb44-76" data-line-number="76"><span class="co">#&gt;   [6]         4           4</span></a>
<a class="sourceLine" id="cb44-77" data-line-number="77"><span class="co">#&gt;   [7]         5           5</span></a>
<a class="sourceLine" id="cb44-78" data-line-number="78"><span class="co">#&gt;   -------</span></a>
<a class="sourceLine" id="cb44-79" data-line-number="79"><span class="co">#&gt;   queryLength: 5 / subjectLength: 5</span></a>
<a class="sourceLine" id="cb44-80" data-line-number="80"><span class="co"># GTuples and GRanges methods identical</span></a>
<a class="sourceLine" id="cb44-81" data-line-number="81"><span class="kw">identical</span>(<span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt3, gt3, <span class="dt">type =</span> <span class="st">'end'</span>), </a>
<a class="sourceLine" id="cb44-82" data-line-number="82">          <span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gr3, gr3, <span class="dt">type =</span> <span class="st">'end'</span>)) <span class="co"># TRUE</span></a>
<a class="sourceLine" id="cb44-83" data-line-number="83"><span class="co">#&gt; Warning in .local(query, subject, maxgap, minoverlap, type, select, ...):</span></a>
<a class="sourceLine" id="cb44-84" data-line-number="84"><span class="co">#&gt; 'type' is not 'equal' so coercing 'query' and 'subject' to 'GRanges'</span></a>
<a class="sourceLine" id="cb44-85" data-line-number="85"><span class="co">#&gt; objects (see docs for details)</span></a>
<a class="sourceLine" id="cb44-86" data-line-number="86"><span class="co">#&gt; [1] TRUE</span></a>
<a class="sourceLine" id="cb44-87" data-line-number="87"></a>
<a class="sourceLine" id="cb44-88" data-line-number="88"><span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt3, gt3, <span class="dt">type =</span> <span class="st">'within'</span>)</a>
<a class="sourceLine" id="cb44-89" data-line-number="89"><span class="co">#&gt; Warning in .local(query, subject, maxgap, minoverlap, type, select, ...):</span></a>
<a class="sourceLine" id="cb44-90" data-line-number="90"><span class="co">#&gt; 'type' is not 'equal' so coercing 'query' and 'subject' to 'GRanges'</span></a>
<a class="sourceLine" id="cb44-91" data-line-number="91"><span class="co">#&gt; objects (see docs for details)</span></a>
<a class="sourceLine" id="cb44-92" data-line-number="92"><span class="co">#&gt; Hits object with 10 hits and 0 metadata columns:</span></a>
<a class="sourceLine" id="cb44-93" data-line-number="93"><span class="co">#&gt;        queryHits subjectHits</span></a>
<a class="sourceLine" id="cb44-94" data-line-number="94"><span class="co">#&gt;        &lt;integer&gt;   &lt;integer&gt;</span></a>
<a class="sourceLine" id="cb44-95" data-line-number="95"><span class="co">#&gt;    [1]         1           3</span></a>
<a class="sourceLine" id="cb44-96" data-line-number="96"><span class="co">#&gt;    [2]         1           1</span></a>
<a class="sourceLine" id="cb44-97" data-line-number="97"><span class="co">#&gt;    [3]         1           4</span></a>
<a class="sourceLine" id="cb44-98" data-line-number="98"><span class="co">#&gt;    [4]         2           3</span></a>
<a class="sourceLine" id="cb44-99" data-line-number="99"><span class="co">#&gt;    [5]         2           2</span></a>
<a class="sourceLine" id="cb44-100" data-line-number="100"><span class="co">#&gt;    [6]         3           3</span></a>
<a class="sourceLine" id="cb44-101" data-line-number="101"><span class="co">#&gt;    [7]         4           3</span></a>
<a class="sourceLine" id="cb44-102" data-line-number="102"><span class="co">#&gt;    [8]         4           1</span></a>
<a class="sourceLine" id="cb44-103" data-line-number="103"><span class="co">#&gt;    [9]         4           4</span></a>
<a class="sourceLine" id="cb44-104" data-line-number="104"><span class="co">#&gt;   [10]         5           5</span></a>
<a class="sourceLine" id="cb44-105" data-line-number="105"><span class="co">#&gt;   -------</span></a>
<a class="sourceLine" id="cb44-106" data-line-number="106"><span class="co">#&gt;   queryLength: 5 / subjectLength: 5</span></a>
<a class="sourceLine" id="cb44-107" data-line-number="107"><span class="co"># GTuples and GRanges methods identical</span></a>
<a class="sourceLine" id="cb44-108" data-line-number="108"><span class="kw">identical</span>(<span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt3, gt3, <span class="dt">type =</span> <span class="st">'within'</span>), </a>
<a class="sourceLine" id="cb44-109" data-line-number="109">          <span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gr3, gr3, <span class="dt">type =</span> <span class="st">'within'</span>)) <span class="co"># TRUE</span></a>
<a class="sourceLine" id="cb44-110" data-line-number="110"><span class="co">#&gt; Warning in .local(query, subject, maxgap, minoverlap, type, select, ...):</span></a>
<a class="sourceLine" id="cb44-111" data-line-number="111"><span class="co">#&gt; 'type' is not 'equal' so coercing 'query' and 'subject' to 'GRanges'</span></a>
<a class="sourceLine" id="cb44-112" data-line-number="112"><span class="co">#&gt; objects (see docs for details)</span></a>
<a class="sourceLine" id="cb44-113" data-line-number="113"><span class="co">#&gt; [1] TRUE</span></a>
<a class="sourceLine" id="cb44-114" data-line-number="114"></a>
<a class="sourceLine" id="cb44-115" data-line-number="115"><span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt3, gt3, <span class="dt">type =</span> <span class="st">'equal'</span>)</a>
<a class="sourceLine" id="cb44-116" data-line-number="116"><span class="co">#&gt; Hits object with 5 hits and 0 metadata columns:</span></a>
<a class="sourceLine" id="cb44-117" data-line-number="117"><span class="co">#&gt;       queryHits subjectHits</span></a>
<a class="sourceLine" id="cb44-118" data-line-number="118"><span class="co">#&gt;       &lt;integer&gt;   &lt;integer&gt;</span></a>
<a class="sourceLine" id="cb44-119" data-line-number="119"><span class="co">#&gt;   [1]         1           1</span></a>
<a class="sourceLine" id="cb44-120" data-line-number="120"><span class="co">#&gt;   [2]         2           2</span></a>
<a class="sourceLine" id="cb44-121" data-line-number="121"><span class="co">#&gt;   [3]         3           3</span></a>
<a class="sourceLine" id="cb44-122" data-line-number="122"><span class="co">#&gt;   [4]         4           4</span></a>
<a class="sourceLine" id="cb44-123" data-line-number="123"><span class="co">#&gt;   [5]         5           5</span></a>
<a class="sourceLine" id="cb44-124" data-line-number="124"><span class="co">#&gt;   -------</span></a>
<a class="sourceLine" id="cb44-125" data-line-number="125"><span class="co">#&gt;   queryLength: 5 / subjectLength: 5</span></a>
<a class="sourceLine" id="cb44-126" data-line-number="126"><span class="co"># GTuples and GRanges methods **not** identical because  GRanges method ignores </span></a>
<a class="sourceLine" id="cb44-127" data-line-number="127"><span class="co"># "internal positions".</span></a>
<a class="sourceLine" id="cb44-128" data-line-number="128"><span class="kw">identical</span>(<span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt3, gt3, <span class="dt">type =</span> <span class="st">'equal'</span>), </a>
<a class="sourceLine" id="cb44-129" data-line-number="129">          <span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gr3, gr3, <span class="dt">type =</span> <span class="st">'equal'</span>)) <span class="co"># FALSE</span></a>
<a class="sourceLine" id="cb44-130" data-line-number="130"><span class="co">#&gt; [1] FALSE</span></a>
<a class="sourceLine" id="cb44-131" data-line-number="131"><span class="co"># Can pass other arguments, such as select and ignore.strand</span></a>
<a class="sourceLine" id="cb44-132" data-line-number="132"><span class="kw"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span>(gt3, gt3, <span class="dt">type =</span> <span class="st">'equal'</span>, <span class="dt">ignore.strand =</span> <span class="ot">TRUE</span>, <span class="dt">select =</span> <span class="st">'last'</span>)</a>
<a class="sourceLine" id="cb44-133" data-line-number="133"><span class="co">#&gt; [1] 2 2 3 4 5</span></a></code></pre></div>
</div>
</div>
<div id="comparison-of-genomic-tuples" class="section level1">
<h1 class="hasAnchor">
<a href="#comparison-of-genomic-tuples" class="anchor"></a>Comparison of genomic tuples</h1>
<p>I have chosen a definition that matches my intuition of what constitutes a<br>
comparison between genomic tuples. However, I am open to suggestions on amending or extending this behaviour in future versions of <em><a href="http://bioconductor.org/packages/GenomicTuples">GenomicTuples</a></em>.</p>
<div id="definition-of-comparison-methods-for-genomic-tuples" class="section level2">
<h2 class="hasAnchor">
<a href="#definition-of-comparison-methods-for-genomic-tuples" class="anchor"></a>Definition of comparison methods for genomic tuples</h2>
<p>The comparison of two genomic tuples, <code>x</code> and <code>y</code>, is done by first comparing the <code>seqnames(x)</code> to <code>seqnames(y)</code>, then <code>strand(x)</code> to <code>strand(y)</code> and finally <code><a href="../reference/GTuples-class.html">tuples(x)</a></code> to <code><a href="../reference/GTuples-class.html">tuples(y)</a></code>.</p>
<p>Ordering of <code>seqnames</code> and <code>strand</code> is as implemented <em>GenomicRanges::GRanges</em>. Ordering of <code>tuples</code> is element-wise, i.e. <span class="math inline">\(pos_{1}, \ldots, pos_{m}\)</span> are compared in turn. For example, <code>chr1:+:10, 20, 30</code> is considered less than <code>chr1:+:10, 20, 40</code>. This defines what I will refer to as the “natural order” of genomic tuples.</p>
<p>The above is implemented in the <code>pcompare</code> method for <em>GTuples</em>, which performs “generalized range-wise comparison” of two <em>GTuples</em> objects, <code>x</code> and <code>y</code>. That is, <code>pcompare(x, y)</code> returns an integer vector where the <span class="math inline">\(i^{th}\)</span> element is a code describing how the <span class="math inline">\(i^{th}\)</span> element in <code>x</code> is qualitatively positioned relatively to the <span class="math inline">\(i^{th}\)</span> element in <code>y</code>. A code that is <code>&lt; 0</code>, <code>= 0</code>, or <code>&gt; 0</code>, corresponds to <code>x[i] &lt; y[i]</code>, <code>x[i] == y[i]</code>, or <code>x[i] &gt; y[i]</code>, respectively.</p>
<p>The 6 traditional binary comparison operators (<code>==</code>, <code>!=</code>, <code>&lt;=</code>, <code>&gt;=</code>, <code>&lt;</code>, and <code>&gt;</code>), other comparison operators (<code>match</code>, <code>order</code>, <code>sort</code>, and <code>rank</code>) and duplicate-based methods (<code>duplicated</code> and <code>unique</code>) all use this “natural order”.</p>
</div>
<div id="examples-1" class="section level2">
<h2 class="hasAnchor">
<a href="#examples-1" class="anchor"></a>Examples</h2>
<p>It is easiest to understand the above definitions by studying a few examples, here using 3-tuples:</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb45-1" data-line-number="1"><span class="co"># Construct example 3-tuples</span></a>
<a class="sourceLine" id="cb45-2" data-line-number="2">gt3 &lt;-<span class="st"> </span><span class="kw">GTuples</span>(<span class="dt">seqnames =</span> <span class="kw">c</span>(<span class="st">'chr1'</span>, <span class="st">'chr1'</span>, <span class="st">'chr1'</span>, <span class="st">'chr1'</span>, <span class="st">'chr2'</span>, <span class="st">'chr1'</span>, </a>
<a class="sourceLine" id="cb45-3" data-line-number="3">                            <span class="st">'chr1'</span>), </a>
<a class="sourceLine" id="cb45-4" data-line-number="4">               <span class="dt">tuples =</span> <span class="kw">matrix</span>(<span class="kw">c</span>(10L, 10L, 10L, 10L, 10L, 5L, 10L, 20L, 20L, </a>
<a class="sourceLine" id="cb45-5" data-line-number="5">                                 20L, 25L, 20L, 20L, 20L, 30L, 30L, 35L, 30L, </a>
<a class="sourceLine" id="cb45-6" data-line-number="6">                                 30L, 30L, 35L), </a>
<a class="sourceLine" id="cb45-7" data-line-number="7">                               <span class="dt">ncol =</span> <span class="dv">3</span>), </a>
<a class="sourceLine" id="cb45-8" data-line-number="8">               <span class="dt">strand =</span> <span class="kw">c</span>(<span class="st">'+'</span>, <span class="st">'-'</span>, <span class="st">'*'</span>, <span class="st">'+'</span>, <span class="st">'+'</span>, <span class="st">'+'</span>, <span class="st">'+'</span>))</a>
<a class="sourceLine" id="cb45-9" data-line-number="9">gt3</a>
<a class="sourceLine" id="cb45-10" data-line-number="10"><span class="co">#&gt; GTuples object with 7 x 3-tuples and 0 metadata columns:</span></a>
<a class="sourceLine" id="cb45-11" data-line-number="11"><span class="co">#&gt;       seqnames pos1 pos2 pos3 strand</span></a>
<a class="sourceLine" id="cb45-12" data-line-number="12"><span class="co">#&gt;   [1]     chr1   10   20   30      +</span></a>
<a class="sourceLine" id="cb45-13" data-line-number="13"><span class="co">#&gt;   [2]     chr1   10   20   30      -</span></a>
<a class="sourceLine" id="cb45-14" data-line-number="14"><span class="co">#&gt;   [3]     chr1   10   20   35      *</span></a>
<a class="sourceLine" id="cb45-15" data-line-number="15"><span class="co">#&gt;   [4]     chr1   10   25   30      +</span></a>
<a class="sourceLine" id="cb45-16" data-line-number="16"><span class="co">#&gt;   [5]     chr2   10   20   30      +</span></a>
<a class="sourceLine" id="cb45-17" data-line-number="17"><span class="co">#&gt;   [6]     chr1    5   20   30      +</span></a>
<a class="sourceLine" id="cb45-18" data-line-number="18"><span class="co">#&gt;   [7]     chr1   10   20   35      +</span></a>
<a class="sourceLine" id="cb45-19" data-line-number="19"><span class="co">#&gt;   ---</span></a>
<a class="sourceLine" id="cb45-20" data-line-number="20"><span class="co">#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></a>
<a class="sourceLine" id="cb45-21" data-line-number="21"></a>
<a class="sourceLine" id="cb45-22" data-line-number="22"><span class="co"># pcompare each tuple to itself</span></a>
<a class="sourceLine" id="cb45-23" data-line-number="23"><span class="kw">pcompare</span>(gt3, gt3)</a>
<a class="sourceLine" id="cb45-24" data-line-number="24"><span class="co">#&gt; [1] 0 0 0 0 0 0 0</span></a>
<a class="sourceLine" id="cb45-25" data-line-number="25">gt3 <span class="op">&lt;</span><span class="st"> </span>gt3</a>
<a class="sourceLine" id="cb45-26" data-line-number="26"><span class="co">#&gt; [1] FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span></a>
<a class="sourceLine" id="cb45-27" data-line-number="27">gt3 <span class="op">&gt;</span><span class="st"> </span>gt3</a>
<a class="sourceLine" id="cb45-28" data-line-number="28"><span class="co">#&gt; [1] FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span></a>
<a class="sourceLine" id="cb45-29" data-line-number="29">gt3 <span class="op">==</span><span class="st"> </span>gt3</a>
<a class="sourceLine" id="cb45-30" data-line-number="30"><span class="co">#&gt; [1] TRUE TRUE TRUE TRUE TRUE TRUE TRUE</span></a>
<a class="sourceLine" id="cb45-31" data-line-number="31"></a>
<a class="sourceLine" id="cb45-32" data-line-number="32"><span class="co"># pcompare the third tuple to all tuples</span></a>
<a class="sourceLine" id="cb45-33" data-line-number="33"><span class="kw">pcompare</span>(gt3[<span class="dv">3</span>], gt3)</a>
<a class="sourceLine" id="cb45-34" data-line-number="34"><span class="co">#&gt; [1]  2  1  0  2 -1  2  2</span></a>
<a class="sourceLine" id="cb45-35" data-line-number="35">gt3[<span class="dv">3</span>] <span class="op">&lt;</span><span class="st"> </span>gt3</a>
<a class="sourceLine" id="cb45-36" data-line-number="36"><span class="co">#&gt; [1] FALSE FALSE FALSE FALSE  TRUE FALSE FALSE</span></a>
<a class="sourceLine" id="cb45-37" data-line-number="37">gt3[<span class="dv">3</span>] <span class="op">&gt;</span><span class="st"> </span>gt3</a>
<a class="sourceLine" id="cb45-38" data-line-number="38"><span class="co">#&gt; [1]  TRUE  TRUE FALSE  TRUE FALSE  TRUE  TRUE</span></a>
<a class="sourceLine" id="cb45-39" data-line-number="39">gt3[<span class="dv">3</span>] <span class="op">==</span><span class="st"> </span>gt3</a>
<a class="sourceLine" id="cb45-40" data-line-number="40"><span class="co">#&gt; [1] FALSE FALSE  TRUE FALSE FALSE FALSE FALSE</span></a>
<a class="sourceLine" id="cb45-41" data-line-number="41"></a>
<a class="sourceLine" id="cb45-42" data-line-number="42">## Some comparisons where tuples differ only in one coordinate</a>
<a class="sourceLine" id="cb45-43" data-line-number="43"></a>
<a class="sourceLine" id="cb45-44" data-line-number="44"><span class="co"># Ordering of seqnames </span></a>
<a class="sourceLine" id="cb45-45" data-line-number="45"><span class="co"># 'chr1' &lt; 'chr2' for tuples with otherwise identical coordinates</span></a>
<a class="sourceLine" id="cb45-46" data-line-number="46">gt3[<span class="dv">1</span>] <span class="op">&lt;</span><span class="st"> </span>gt3[<span class="dv">5</span>] <span class="co"># TRUE</span></a>
<a class="sourceLine" id="cb45-47" data-line-number="47"><span class="co">#&gt; [1] TRUE</span></a>
<a class="sourceLine" id="cb45-48" data-line-number="48"></a>
<a class="sourceLine" id="cb45-49" data-line-number="49"><span class="co"># Ordering of strands</span></a>
<a class="sourceLine" id="cb45-50" data-line-number="50"><span class="co"># '+' &lt; '-' &lt; '*' for tuples with otherwise identical coordiantes</span></a>
<a class="sourceLine" id="cb45-51" data-line-number="51">gt3[<span class="dv">1</span>] <span class="op">&lt;</span><span class="st"> </span>gt3[<span class="dv">2</span>] <span class="co"># TRUE</span></a>
<a class="sourceLine" id="cb45-52" data-line-number="52"><span class="co">#&gt; [1] TRUE</span></a>
<a class="sourceLine" id="cb45-53" data-line-number="53">gt3[<span class="dv">1</span>] <span class="op">&lt;</span><span class="st"> </span>gt3[<span class="dv">2</span>] <span class="co"># TRUE</span></a>
<a class="sourceLine" id="cb45-54" data-line-number="54"><span class="co">#&gt; [1] TRUE</span></a>
<a class="sourceLine" id="cb45-55" data-line-number="55">gt3[<span class="dv">1</span>] <span class="op">&lt;</span><span class="st"> </span><span class="kw">unstrand</span>(gt3[<span class="dv">2</span>]) <span class="co"># TRUE</span></a>
<a class="sourceLine" id="cb45-56" data-line-number="56"><span class="co">#&gt; [1] TRUE</span></a>
<a class="sourceLine" id="cb45-57" data-line-number="57">gt3[<span class="dv">2</span>] <span class="op">&lt;</span><span class="st"> </span><span class="kw">unstrand</span>(gt3[<span class="dv">2</span>]) <span class="co"># TRUE</span></a>
<a class="sourceLine" id="cb45-58" data-line-number="58"><span class="co">#&gt; [1] TRUE</span></a>
<a class="sourceLine" id="cb45-59" data-line-number="59"></a>
<a class="sourceLine" id="cb45-60" data-line-number="60"><span class="co"># Ordering of tuples</span></a>
<a class="sourceLine" id="cb45-61" data-line-number="61"><span class="co"># Tuples checked sequentially from pos1, ..., posm for tuples with otherwise</span></a>
<a class="sourceLine" id="cb45-62" data-line-number="62"><span class="co"># identical coordinates</span></a>
<a class="sourceLine" id="cb45-63" data-line-number="63">gt3[<span class="dv">6</span>] <span class="op">&lt;</span><span class="st"> </span>gt3[<span class="dv">1</span>] <span class="co"># TRUE due to pos1</span></a>
<a class="sourceLine" id="cb45-64" data-line-number="64"><span class="co">#&gt; [1] TRUE</span></a>
<a class="sourceLine" id="cb45-65" data-line-number="65">gt3[<span class="dv">2</span>] <span class="op">&lt;</span><span class="st"> </span>gt3[<span class="dv">4</span>] <span class="co"># TRUE due to pos2</span></a>
<a class="sourceLine" id="cb45-66" data-line-number="66"><span class="co">#&gt; [1] FALSE</span></a>
<a class="sourceLine" id="cb45-67" data-line-number="67">gt3[<span class="dv">1</span>] <span class="op">&lt;</span><span class="st"> </span>gt3[<span class="dv">7</span>] <span class="co"># TRUE due to pos3</span></a>
<a class="sourceLine" id="cb45-68" data-line-number="68"><span class="co">#&gt; [1] TRUE</span></a>
<a class="sourceLine" id="cb45-69" data-line-number="69"></a>
<a class="sourceLine" id="cb45-70" data-line-number="70"><span class="co"># Sorting of tuples</span></a>
<a class="sourceLine" id="cb45-71" data-line-number="71"><span class="co"># Sorted first by seqnames, then by strand, then by tuples</span></a>
<a class="sourceLine" id="cb45-72" data-line-number="72"><span class="kw">sort</span>(gt3)</a>
<a class="sourceLine" id="cb45-73" data-line-number="73"><span class="co">#&gt; GTuples object with 7 x 3-tuples and 0 metadata columns:</span></a>
<a class="sourceLine" id="cb45-74" data-line-number="74"><span class="co">#&gt;       seqnames pos1 pos2 pos3 strand</span></a>
<a class="sourceLine" id="cb45-75" data-line-number="75"><span class="co">#&gt;   [1]     chr1    5   20   30      +</span></a>
<a class="sourceLine" id="cb45-76" data-line-number="76"><span class="co">#&gt;   [2]     chr1   10   20   30      +</span></a>
<a class="sourceLine" id="cb45-77" data-line-number="77"><span class="co">#&gt;   [3]     chr1   10   20   35      +</span></a>
<a class="sourceLine" id="cb45-78" data-line-number="78"><span class="co">#&gt;   [4]     chr1   10   25   30      +</span></a>
<a class="sourceLine" id="cb45-79" data-line-number="79"><span class="co">#&gt;   [5]     chr1   10   20   30      -</span></a>
<a class="sourceLine" id="cb45-80" data-line-number="80"><span class="co">#&gt;   [6]     chr1   10   20   35      *</span></a>
<a class="sourceLine" id="cb45-81" data-line-number="81"><span class="co">#&gt;   [7]     chr2   10   20   30      +</span></a>
<a class="sourceLine" id="cb45-82" data-line-number="82"><span class="co">#&gt;   ---</span></a>
<a class="sourceLine" id="cb45-83" data-line-number="83"><span class="co">#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></a>
<a class="sourceLine" id="cb45-84" data-line-number="84"></a>
<a class="sourceLine" id="cb45-85" data-line-number="85"><span class="co"># Duplicate tuples</span></a>
<a class="sourceLine" id="cb45-86" data-line-number="86"><span class="co"># Duplicate tuples must have identical seqnames, strand and positions (tuples)</span></a>
<a class="sourceLine" id="cb45-87" data-line-number="87"><span class="kw">duplicated</span>(<span class="kw">c</span>(gt3, gt3[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]))</a>
<a class="sourceLine" id="cb45-88" data-line-number="88"><span class="co">#&gt;  [1] FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE</span></a>
<a class="sourceLine" id="cb45-89" data-line-number="89"><span class="kw">unique</span>(<span class="kw">c</span>(gt3, gt3[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]))</a>
<a class="sourceLine" id="cb45-90" data-line-number="90"><span class="co">#&gt; GTuples object with 7 x 3-tuples and 0 metadata columns:</span></a>
<a class="sourceLine" id="cb45-91" data-line-number="91"><span class="co">#&gt;       seqnames pos1 pos2 pos3 strand</span></a>
<a class="sourceLine" id="cb45-92" data-line-number="92"><span class="co">#&gt;   [1]     chr1   10   20   30      +</span></a>
<a class="sourceLine" id="cb45-93" data-line-number="93"><span class="co">#&gt;   [2]     chr1   10   20   30      -</span></a>
<a class="sourceLine" id="cb45-94" data-line-number="94"><span class="co">#&gt;   [3]     chr1   10   20   35      *</span></a>
<a class="sourceLine" id="cb45-95" data-line-number="95"><span class="co">#&gt;   [4]     chr1   10   25   30      +</span></a>
<a class="sourceLine" id="cb45-96" data-line-number="96"><span class="co">#&gt;   [5]     chr2   10   20   30      +</span></a>
<a class="sourceLine" id="cb45-97" data-line-number="97"><span class="co">#&gt;   [6]     chr1    5   20   30      +</span></a>
<a class="sourceLine" id="cb45-98" data-line-number="98"><span class="co">#&gt;   [7]     chr1   10   20   35      +</span></a>
<a class="sourceLine" id="cb45-99" data-line-number="99"><span class="co">#&gt;   ---</span></a>
<a class="sourceLine" id="cb45-100" data-line-number="100"><span class="co">#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></a></code></pre></div>
</div>
</div>
<div id="acknowledgements" class="section level1">
<h1 class="hasAnchor">
<a href="#acknowledgements" class="anchor"></a>Acknowledgements</h1>
<p>I am very grateful to all the Bioconductor developers but particularly wish to thank the developers of <em><a href="http://bioconductor.org/packages/GenomicRanges">GenomicRanges</a></em> (<a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003118">Lawrence, M. et al. Software for computing and annotating genomic ranges. PLoS Comput. Biol. 9, e1003118 (2013).</a>), which <em><a href="http://bioconductor.org/packages/GenomicTuples">GenomicTuples</a></em> uses heavily and is based upon. A special thanks to Hervé Pagès for his assistance and fixes when making upstream changes to <em><a href="http://bioconductor.org/packages/GenomicRanges">GenomicRanges</a></em>.</p>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h1>
<p>Here is the output of <code>sessionInfo</code> on the system on which this document was compiled:</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb46-1" data-line-number="1"><span class="kw">sessionInfo</span>()</a>
<a class="sourceLine" id="cb46-2" data-line-number="2"><span class="co">#&gt; R Under development (unstable) (2018-03-05 r74359)</span></a>
<a class="sourceLine" id="cb46-3" data-line-number="3"><span class="co">#&gt; Platform: x86_64-apple-darwin15.6.0 (64-bit)</span></a>
<a class="sourceLine" id="cb46-4" data-line-number="4"><span class="co">#&gt; Running under: macOS Sierra 10.12.6</span></a>
<a class="sourceLine" id="cb46-5" data-line-number="5"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb46-6" data-line-number="6"><span class="co">#&gt; Matrix products: default</span></a>
<a class="sourceLine" id="cb46-7" data-line-number="7"><span class="co">#&gt; BLAS: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRblas.0.dylib</span></a>
<a class="sourceLine" id="cb46-8" data-line-number="8"><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib</span></a>
<a class="sourceLine" id="cb46-9" data-line-number="9"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb46-10" data-line-number="10"><span class="co">#&gt; locale:</span></a>
<a class="sourceLine" id="cb46-11" data-line-number="11"><span class="co">#&gt; [1] en_AU.UTF-8/en_AU.UTF-8/en_AU.UTF-8/C/en_AU.UTF-8/en_AU.UTF-8</span></a>
<a class="sourceLine" id="cb46-12" data-line-number="12"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb46-13" data-line-number="13"><span class="co">#&gt; attached base packages:</span></a>
<a class="sourceLine" id="cb46-14" data-line-number="14"><span class="co">#&gt; [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span></a>
<a class="sourceLine" id="cb46-15" data-line-number="15"><span class="co">#&gt; [8] methods   base     </span></a>
<a class="sourceLine" id="cb46-16" data-line-number="16"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb46-17" data-line-number="17"><span class="co">#&gt; other attached packages:</span></a>
<a class="sourceLine" id="cb46-18" data-line-number="18"><span class="co">#&gt; [1] GenomicTuples_1.13.4  GenomicRanges_1.31.23 GenomeInfoDb_1.15.5  </span></a>
<a class="sourceLine" id="cb46-19" data-line-number="19"><span class="co">#&gt; [4] IRanges_2.13.28       S4Vectors_0.17.38     BiocGenerics_0.25.3  </span></a>
<a class="sourceLine" id="cb46-20" data-line-number="20"><span class="co">#&gt; [7] BiocStyle_2.7.8      </span></a>
<a class="sourceLine" id="cb46-21" data-line-number="21"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb46-22" data-line-number="22"><span class="co">#&gt; loaded via a namespace (and not attached):</span></a>
<a class="sourceLine" id="cb46-23" data-line-number="23"><span class="co">#&gt;  [1] Rcpp_0.12.16           digest_0.6.15          rprojroot_1.3-2       </span></a>
<a class="sourceLine" id="cb46-24" data-line-number="24"><span class="co">#&gt;  [4] bitops_1.0-6           backports_1.1.2        magrittr_1.5          </span></a>
<a class="sourceLine" id="cb46-25" data-line-number="25"><span class="co">#&gt;  [7] evaluate_0.10.1        zlibbioc_1.25.0        stringi_1.1.7         </span></a>
<a class="sourceLine" id="cb46-26" data-line-number="26"><span class="co">#&gt; [10] XVector_0.19.9         data.table_1.10.4-3    rmarkdown_1.9         </span></a>
<a class="sourceLine" id="cb46-27" data-line-number="27"><span class="co">#&gt; [13] tools_3.5.0            stringr_1.3.0          RCurl_1.95-4.10       </span></a>
<a class="sourceLine" id="cb46-28" data-line-number="28"><span class="co">#&gt; [16] yaml_2.1.18            compiler_3.5.0         htmltools_0.3.6       </span></a>
<a class="sourceLine" id="cb46-29" data-line-number="29"><span class="co">#&gt; [19] knitr_1.20             GenomeInfoDbData_1.1.0</span></a></code></pre></div>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<ul>
<li><a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003118">Lawrence, M. et al. Software for computing and annotating genomic ranges. PLoS Comput. Biol. 9, e1003118 (2013).</a></li>
</ul>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>The <em><a href="http://bioconductor.org/packages/GenomicRanges">GenomicRanges</a></em> vignette can be accessed by typing<code><a href="https://cran.rstudio.com/web/packages/GenomicRanges/vignettes/GenomicRangesIntroduction.pdf">vignette("GenomicRangesIntroduction", package = "GenomicRanges")</a></code> at the R console.<a href="#fnref1" class="footnote-back">↩</a></p></li>
<li id="fn2"><p>This may be changed in future versions of <em><a href="http://bioconductor.org/packages/GenomicTuples">GenomicTuples</a></em>.<a href="#fnref2" class="footnote-back">↩</a></p></li>
<li id="fn3"><p>The <code>findOverlaps</code>-based methods are <code>findOverlaps</code>, <code>countOverlaps</code>, <code>overlapsAny</code> and <code>subsetByOverlaps</code>.<a href="#fnref3" class="footnote-back">↩</a></p></li>
</ol>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#introduction">Introduction</a><ul class="nav nav-pills nav-stacked">
<li><a href="#what-is-a-genomic-tuple">What is a genomic tuple?</a></li>
      <li><a href="#when-might-you-need-a-genomic-tuple">When might you need a genomic tuple?</a></li>
      </ul>
</li>
      <li>
<a href="#gtuples"><em>GTuples</em></a><ul class="nav nav-pills nav-stacked">
<li><a href="#gtuples-methods"><em>GTuples</em> methods</a></li>
      <li><a href="#implementation-details">Implementation details</a></li>
      </ul>
</li>
      <li>
<a href="#gtupleslist"><em>GTuplesList</em></a><ul class="nav nav-pills nav-stacked">
<li><a href="#gtupleslist-methods"><em>GTuplesList</em> methods</a></li>
      <li><a href="#implementation-details-1">Implementation details</a></li>
      </ul>
</li>
      <li>
<a href="#findoverlaps-based-methods"><code>findOverlaps</code>-based methods</a><ul class="nav nav-pills nav-stacked">
<li><a href="#definition-of-overlapping-genomic-tuples">Definition of overlapping genomic tuples</a></li>
      <li><a href="#definition-of-overlapping-genomic-tuples-and-ranges">Definition of overlapping genomic tuples and ranges</a></li>
      <li><a href="#examples">Examples</a></li>
      </ul>
</li>
      <li>
<a href="#comparison-of-genomic-tuples">Comparison of genomic tuples</a><ul class="nav nav-pills nav-stacked">
<li><a href="#definition-of-comparison-methods-for-genomic-tuples">Definition of comparison methods for genomic tuples</a></li>
      <li><a href="#examples-1">Examples</a></li>
      </ul>
</li>
      <li><a href="#acknowledgements">Acknowledgements</a></li>
      <li><a href="#session-info">Session info</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Peter Hickey.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
